<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="UTF-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="generator" content="Asciidoctor 2.0.12">
<meta name="author" content="Simon Heumos, Andrea Guarracino, Erik Garrison">
<title>Reference Documentation for ODGI v0.6.0</title>
<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Open+Sans:300,300italic,400,400italic,600,600italic%7CNoto+Serif:400,400italic,700,700italic%7CDroid+Sans+Mono:400,700">
<style>
/* Asciidoctor default stylesheet | MIT License | https://asciidoctor.org */
/* Uncomment @import statement to use as custom stylesheet */
/*@import "https://fonts.googleapis.com/css?family=Open+Sans:300,300italic,400,400italic,600,600italic%7CNoto+Serif:400,400italic,700,700italic%7CDroid+Sans+Mono:400,700";*/
article,aside,details,figcaption,figure,footer,header,hgroup,main,nav,section{display:block}
audio,video{display:inline-block}
audio:not([controls]){display:none;height:0}
html{font-family:sans-serif;-ms-text-size-adjust:100%;-webkit-text-size-adjust:100%}
a{background:none}
a:focus{outline:thin dotted}
a:active,a:hover{outline:0}
h1{font-size:2em;margin:.67em 0}
abbr[title]{border-bottom:1px dotted}
b,strong{font-weight:bold}
dfn{font-style:italic}
hr{-moz-box-sizing:content-box;box-sizing:content-box;height:0}
mark{background:#ff0;color:#000}
code,kbd,pre,samp{font-family:monospace;font-size:1em}
pre{white-space:pre-wrap}
q{quotes:"\201C" "\201D" "\2018" "\2019"}
small{font-size:80%}
sub,sup{font-size:75%;line-height:0;position:relative;vertical-align:baseline}
sup{top:-.5em}
sub{bottom:-.25em}
img{border:0}
svg:not(:root){overflow:hidden}
figure{margin:0}
fieldset{border:1px solid silver;margin:0 2px;padding:.35em .625em .75em}
legend{border:0;padding:0}
button,input,select,textarea{font-family:inherit;font-size:100%;margin:0}
button,input{line-height:normal}
button,select{text-transform:none}
button,html input[type="button"],input[type="reset"],input[type="submit"]{-webkit-appearance:button;cursor:pointer}
button[disabled],html input[disabled]{cursor:default}
input[type="checkbox"],input[type="radio"]{box-sizing:border-box;padding:0}
button::-moz-focus-inner,input::-moz-focus-inner{border:0;padding:0}
textarea{overflow:auto;vertical-align:top}
table{border-collapse:collapse;border-spacing:0}
*,*::before,*::after{-moz-box-sizing:border-box;-webkit-box-sizing:border-box;box-sizing:border-box}
html,body{font-size:100%}
body{background:#fff;color:rgba(0,0,0,.8);padding:0;margin:0;font-family:"Noto Serif","DejaVu Serif",serif;font-weight:400;font-style:normal;line-height:1;position:relative;cursor:auto;tab-size:4;word-wrap:anywhere;-moz-osx-font-smoothing:grayscale;-webkit-font-smoothing:antialiased}
a:hover{cursor:pointer}
img,object,embed{max-width:100%;height:auto}
object,embed{height:100%}
img{-ms-interpolation-mode:bicubic}
.left{float:left!important}
.right{float:right!important}
.text-left{text-align:left!important}
.text-right{text-align:right!important}
.text-center{text-align:center!important}
.text-justify{text-align:justify!important}
.hide{display:none}
img,object,svg{display:inline-block;vertical-align:middle}
textarea{height:auto;min-height:50px}
select{width:100%}
.subheader,.admonitionblock td.content>.title,.audioblock>.title,.exampleblock>.title,.imageblock>.title,.listingblock>.title,.literalblock>.title,.stemblock>.title,.openblock>.title,.paragraph>.title,.quoteblock>.title,table.tableblock>.title,.verseblock>.title,.videoblock>.title,.dlist>.title,.olist>.title,.ulist>.title,.qlist>.title,.hdlist>.title{line-height:1.45;color:#7a2518;font-weight:400;margin-top:0;margin-bottom:.25em}
div,dl,dt,dd,ul,ol,li,h1,h2,h3,#toctitle,.sidebarblock>.content>.title,h4,h5,h6,pre,form,p,blockquote,th,td{margin:0;padding:0}
a{color:#2156a5;text-decoration:underline;line-height:inherit}
a:hover,a:focus{color:#1d4b8f}
a img{border:0}
p{font-family:inherit;font-weight:400;font-size:1em;line-height:1.6;margin-bottom:1.25em;text-rendering:optimizeLegibility}
p aside{font-size:.875em;line-height:1.35;font-style:italic}
h1,h2,h3,#toctitle,.sidebarblock>.content>.title,h4,h5,h6{font-family:"Open Sans","DejaVu Sans",sans-serif;font-weight:300;font-style:normal;color:#ba3925;text-rendering:optimizeLegibility;margin-top:1em;margin-bottom:.5em;line-height:1.0125em}
h1 small,h2 small,h3 small,#toctitle small,.sidebarblock>.content>.title small,h4 small,h5 small,h6 small{font-size:60%;color:#e99b8f;line-height:0}
h1{font-size:2.125em}
h2{font-size:1.6875em}
h3,#toctitle,.sidebarblock>.content>.title{font-size:1.375em}
h4,h5{font-size:1.125em}
h6{font-size:1em}
hr{border:solid #dddddf;border-width:1px 0 0;clear:both;margin:1.25em 0 1.1875em;height:0}
em,i{font-style:italic;line-height:inherit}
strong,b{font-weight:bold;line-height:inherit}
small{font-size:60%;line-height:inherit}
code{font-family:"Droid Sans Mono","DejaVu Sans Mono",monospace;font-weight:400;color:rgba(0,0,0,.9)}
ul,ol,dl{font-size:1em;line-height:1.6;margin-bottom:1.25em;list-style-position:outside;font-family:inherit}
ul,ol{margin-left:1.5em}
ul li ul,ul li ol{margin-left:1.25em;margin-bottom:0;font-size:1em}
ul.square li ul,ul.circle li ul,ul.disc li ul{list-style:inherit}
ul.square{list-style-type:square}
ul.circle{list-style-type:circle}
ul.disc{list-style-type:disc}
ol li ul,ol li ol{margin-left:1.25em;margin-bottom:0}
dl dt{margin-bottom:.3125em;font-weight:bold}
dl dd{margin-bottom:1.25em}
abbr,acronym{text-transform:uppercase;font-size:90%;color:rgba(0,0,0,.8);border-bottom:1px dotted #ddd;cursor:help}
abbr{text-transform:none}
blockquote{margin:0 0 1.25em;padding:.5625em 1.25em 0 1.1875em;border-left:1px solid #ddd}
blockquote cite{display:block;font-size:.9375em;color:rgba(0,0,0,.6)}
blockquote cite::before{content:"\2014 \0020"}
blockquote cite a,blockquote cite a:visited{color:rgba(0,0,0,.6)}
blockquote,blockquote p{line-height:1.6;color:rgba(0,0,0,.85)}
@media screen and (min-width:768px){h1,h2,h3,#toctitle,.sidebarblock>.content>.title,h4,h5,h6{line-height:1.2}
h1{font-size:2.75em}
h2{font-size:2.3125em}
h3,#toctitle,.sidebarblock>.content>.title{font-size:1.6875em}
h4{font-size:1.4375em}}
table{background:#fff;margin-bottom:1.25em;border:solid 1px #dedede;word-wrap:normal}
table thead,table tfoot{background:#f7f8f7}
table thead tr th,table thead tr td,table tfoot tr th,table tfoot tr td{padding:.5em .625em .625em;font-size:inherit;color:rgba(0,0,0,.8);text-align:left}
table tr th,table tr td{padding:.5625em .625em;font-size:inherit;color:rgba(0,0,0,.8)}
table tr.even,table tr.alt{background:#f8f8f7}
table thead tr th,table tfoot tr th,table tbody tr td,table tr td,table tfoot tr td{line-height:1.6}
h1,h2,h3,#toctitle,.sidebarblock>.content>.title,h4,h5,h6{line-height:1.2;word-spacing:-.05em}
h1 strong,h2 strong,h3 strong,#toctitle strong,.sidebarblock>.content>.title strong,h4 strong,h5 strong,h6 strong{font-weight:400}
.center{margin-left:auto;margin-right:auto}
.stretch{width:100%}
.clearfix::before,.clearfix::after,.float-group::before,.float-group::after{content:" ";display:table}
.clearfix::after,.float-group::after{clear:both}
:not(pre).nobreak{word-wrap:normal}
:not(pre).nowrap{white-space:nowrap}
:not(pre).pre-wrap{white-space:pre-wrap}
:not(pre):not([class^=L])>code{font-size:.9375em;font-style:normal!important;letter-spacing:0;padding:.1em .5ex;word-spacing:-.15em;background:#f7f7f8;-webkit-border-radius:4px;border-radius:4px;line-height:1.45;text-rendering:optimizeSpeed}
pre{color:rgba(0,0,0,.9);font-family:"Droid Sans Mono","DejaVu Sans Mono",monospace;line-height:1.45;text-rendering:optimizeSpeed}
pre code,pre pre{color:inherit;font-size:inherit;line-height:inherit}
pre>code{display:block}
pre.nowrap,pre.nowrap pre{white-space:pre;word-wrap:normal}
em em{font-style:normal}
strong strong{font-weight:400}
.keyseq{color:rgba(51,51,51,.8)}
kbd{font-family:"Droid Sans Mono","DejaVu Sans Mono",monospace;display:inline-block;color:rgba(0,0,0,.8);font-size:.65em;line-height:1.45;background:#f7f7f7;border:1px solid #ccc;-webkit-border-radius:3px;border-radius:3px;-webkit-box-shadow:0 1px 0 rgba(0,0,0,.2),0 0 0 .1em white inset;box-shadow:0 1px 0 rgba(0,0,0,.2),0 0 0 .1em #fff inset;margin:0 .15em;padding:.2em .5em;vertical-align:middle;position:relative;top:-.1em;white-space:nowrap}
.keyseq kbd:first-child{margin-left:0}
.keyseq kbd:last-child{margin-right:0}
.menuseq,.menuref{color:#000}
.menuseq b:not(.caret),.menuref{font-weight:inherit}
.menuseq{word-spacing:-.02em}
.menuseq b.caret{font-size:1.25em;line-height:.8}
.menuseq i.caret{font-weight:bold;text-align:center;width:.45em}
b.button::before,b.button::after{position:relative;top:-1px;font-weight:400}
b.button::before{content:"[";padding:0 3px 0 2px}
b.button::after{content:"]";padding:0 2px 0 3px}
p a>code:hover{color:rgba(0,0,0,.9)}
#header,#content,#footnotes,#footer{width:100%;margin-left:auto;margin-right:auto;margin-top:0;margin-bottom:0;max-width:62.5em;*zoom:1;position:relative;padding-left:.9375em;padding-right:.9375em}
#header::before,#header::after,#content::before,#content::after,#footnotes::before,#footnotes::after,#footer::before,#footer::after{content:" ";display:table}
#header::after,#content::after,#footnotes::after,#footer::after{clear:both}
#content{margin-top:1.25em}
#content::before{content:none}
#header>h1:first-child{color:rgba(0,0,0,.85);margin-top:2.25rem;margin-bottom:0}
#header>h1:first-child+#toc{margin-top:8px;border-top:1px solid #dddddf}
#header>h1:only-child,body.toc2 #header>h1:nth-last-child(2){border-bottom:1px solid #dddddf;padding-bottom:8px}
#header .details{border-bottom:1px solid #dddddf;line-height:1.45;padding-top:.25em;padding-bottom:.25em;padding-left:.25em;color:rgba(0,0,0,.6);display:-ms-flexbox;display:-webkit-flex;display:flex;-ms-flex-flow:row wrap;-webkit-flex-flow:row wrap;flex-flow:row wrap}
#header .details span:first-child{margin-left:-.125em}
#header .details span.email a{color:rgba(0,0,0,.85)}
#header .details br{display:none}
#header .details br+span::before{content:"\00a0\2013\00a0"}
#header .details br+span.author::before{content:"\00a0\22c5\00a0";color:rgba(0,0,0,.85)}
#header .details br+span#revremark::before{content:"\00a0|\00a0"}
#header #revnumber{text-transform:capitalize}
#header #revnumber::after{content:"\00a0"}
#content>h1:first-child:not([class]){color:rgba(0,0,0,.85);border-bottom:1px solid #dddddf;padding-bottom:8px;margin-top:0;padding-top:1rem;margin-bottom:1.25rem}
#toc{border-bottom:1px solid #e7e7e9;padding-bottom:.5em}
#toc>ul{margin-left:.125em}
#toc ul.sectlevel0>li>a{font-style:italic}
#toc ul.sectlevel0 ul.sectlevel1{margin:.5em 0}
#toc ul{font-family:"Open Sans","DejaVu Sans",sans-serif;list-style-type:none}
#toc li{line-height:1.3334;margin-top:.3334em}
#toc a{text-decoration:none}
#toc a:active{text-decoration:underline}
#toctitle{color:#7a2518;font-size:1.2em}
@media screen and (min-width:768px){#toctitle{font-size:1.375em}
body.toc2{padding-left:15em;padding-right:0}
#toc.toc2{margin-top:0!important;background:#f8f8f7;position:fixed;width:15em;left:0;top:0;border-right:1px solid #e7e7e9;border-top-width:0!important;border-bottom-width:0!important;z-index:1000;padding:1.25em 1em;height:100%;overflow:auto}
#toc.toc2 #toctitle{margin-top:0;margin-bottom:.8rem;font-size:1.2em}
#toc.toc2>ul{font-size:.9em;margin-bottom:0}
#toc.toc2 ul ul{margin-left:0;padding-left:1em}
#toc.toc2 ul.sectlevel0 ul.sectlevel1{padding-left:0;margin-top:.5em;margin-bottom:.5em}
body.toc2.toc-right{padding-left:0;padding-right:15em}
body.toc2.toc-right #toc.toc2{border-right-width:0;border-left:1px solid #e7e7e9;left:auto;right:0}}
@media screen and (min-width:1280px){body.toc2{padding-left:20em;padding-right:0}
#toc.toc2{width:20em}
#toc.toc2 #toctitle{font-size:1.375em}
#toc.toc2>ul{font-size:.95em}
#toc.toc2 ul ul{padding-left:1.25em}
body.toc2.toc-right{padding-left:0;padding-right:20em}}
#content #toc{border-style:solid;border-width:1px;border-color:#e0e0dc;margin-bottom:1.25em;padding:1.25em;background:#f8f8f7;-webkit-border-radius:4px;border-radius:4px}
#content #toc>:first-child{margin-top:0}
#content #toc>:last-child{margin-bottom:0}
#footer{max-width:none;background:rgba(0,0,0,.8);padding:1.25em}
#footer-text{color:rgba(255,255,255,.8);line-height:1.44}
#content{margin-bottom:.625em}
.sect1{padding-bottom:.625em}
@media screen and (min-width:768px){#content{margin-bottom:1.25em}
.sect1{padding-bottom:1.25em}}
.sect1:last-child{padding-bottom:0}
.sect1+.sect1{border-top:1px solid #e7e7e9}
#content h1>a.anchor,h2>a.anchor,h3>a.anchor,#toctitle>a.anchor,.sidebarblock>.content>.title>a.anchor,h4>a.anchor,h5>a.anchor,h6>a.anchor{position:absolute;z-index:1001;width:1.5ex;margin-left:-1.5ex;display:block;text-decoration:none!important;visibility:hidden;text-align:center;font-weight:400}
#content h1>a.anchor::before,h2>a.anchor::before,h3>a.anchor::before,#toctitle>a.anchor::before,.sidebarblock>.content>.title>a.anchor::before,h4>a.anchor::before,h5>a.anchor::before,h6>a.anchor::before{content:"\00A7";font-size:.85em;display:block;padding-top:.1em}
#content h1:hover>a.anchor,#content h1>a.anchor:hover,h2:hover>a.anchor,h2>a.anchor:hover,h3:hover>a.anchor,#toctitle:hover>a.anchor,.sidebarblock>.content>.title:hover>a.anchor,h3>a.anchor:hover,#toctitle>a.anchor:hover,.sidebarblock>.content>.title>a.anchor:hover,h4:hover>a.anchor,h4>a.anchor:hover,h5:hover>a.anchor,h5>a.anchor:hover,h6:hover>a.anchor,h6>a.anchor:hover{visibility:visible}
#content h1>a.link,h2>a.link,h3>a.link,#toctitle>a.link,.sidebarblock>.content>.title>a.link,h4>a.link,h5>a.link,h6>a.link{color:#ba3925;text-decoration:none}
#content h1>a.link:hover,h2>a.link:hover,h3>a.link:hover,#toctitle>a.link:hover,.sidebarblock>.content>.title>a.link:hover,h4>a.link:hover,h5>a.link:hover,h6>a.link:hover{color:#a53221}
details,.audioblock,.imageblock,.literalblock,.listingblock,.stemblock,.videoblock{margin-bottom:1.25em}
details>summary:first-of-type{cursor:pointer;display:list-item;outline:none;margin-bottom:.75em}
.admonitionblock td.content>.title,.audioblock>.title,.exampleblock>.title,.imageblock>.title,.listingblock>.title,.literalblock>.title,.stemblock>.title,.openblock>.title,.paragraph>.title,.quoteblock>.title,table.tableblock>.title,.verseblock>.title,.videoblock>.title,.dlist>.title,.olist>.title,.ulist>.title,.qlist>.title,.hdlist>.title{text-rendering:optimizeLegibility;text-align:left;font-family:"Noto Serif","DejaVu Serif",serif;font-size:1rem;font-style:italic}
table.tableblock.fit-content>caption.title{white-space:nowrap;width:0}
.paragraph.lead>p,#preamble>.sectionbody>[class="paragraph"]:first-of-type p{font-size:1.21875em;line-height:1.6;color:rgba(0,0,0,.85)}
table.tableblock #preamble>.sectionbody>[class="paragraph"]:first-of-type p{font-size:inherit}
.admonitionblock>table{border-collapse:separate;border:0;background:none;width:100%}
.admonitionblock>table td.icon{text-align:center;width:80px}
.admonitionblock>table td.icon img{max-width:none}
.admonitionblock>table td.icon .title{font-weight:bold;font-family:"Open Sans","DejaVu Sans",sans-serif;text-transform:uppercase}
.admonitionblock>table td.content{padding-left:1.125em;padding-right:1.25em;border-left:1px solid #dddddf;color:rgba(0,0,0,.6);word-wrap:anywhere}
.admonitionblock>table td.content>:last-child>:last-child{margin-bottom:0}
.exampleblock>.content{border-style:solid;border-width:1px;border-color:#e6e6e6;margin-bottom:1.25em;padding:1.25em;background:#fff;-webkit-border-radius:4px;border-radius:4px}
.exampleblock>.content>:first-child{margin-top:0}
.exampleblock>.content>:last-child{margin-bottom:0}
.sidebarblock{border-style:solid;border-width:1px;border-color:#dbdbd6;margin-bottom:1.25em;padding:1.25em;background:#f3f3f2;-webkit-border-radius:4px;border-radius:4px}
.sidebarblock>:first-child{margin-top:0}
.sidebarblock>:last-child{margin-bottom:0}
.sidebarblock>.content>.title{color:#7a2518;margin-top:0;text-align:center}
.exampleblock>.content>:last-child>:last-child,.exampleblock>.content .olist>ol>li:last-child>:last-child,.exampleblock>.content .ulist>ul>li:last-child>:last-child,.exampleblock>.content .qlist>ol>li:last-child>:last-child,.sidebarblock>.content>:last-child>:last-child,.sidebarblock>.content .olist>ol>li:last-child>:last-child,.sidebarblock>.content .ulist>ul>li:last-child>:last-child,.sidebarblock>.content .qlist>ol>li:last-child>:last-child{margin-bottom:0}
.literalblock pre,.listingblock>.content>pre{-webkit-border-radius:4px;border-radius:4px;overflow-x:auto;padding:1em;font-size:.8125em}
@media screen and (min-width:768px){.literalblock pre,.listingblock>.content>pre{font-size:.90625em}}
@media screen and (min-width:1280px){.literalblock pre,.listingblock>.content>pre{font-size:1em}}
.literalblock pre,.listingblock>.content>pre:not(.highlight),.listingblock>.content>pre[class="highlight"],.listingblock>.content>pre[class^="highlight "]{background:#f7f7f8}
.literalblock.output pre{color:#f7f7f8;background:rgba(0,0,0,.9)}
.listingblock>.content{position:relative}
.listingblock code[data-lang]::before{display:none;content:attr(data-lang);position:absolute;font-size:.75em;top:.425rem;right:.5rem;line-height:1;text-transform:uppercase;color:inherit;opacity:.5}
.listingblock:hover code[data-lang]::before{display:block}
.listingblock.terminal pre .command::before{content:attr(data-prompt);padding-right:.5em;color:inherit;opacity:.5}
.listingblock.terminal pre .command:not([data-prompt])::before{content:"$"}
.listingblock pre.highlightjs{padding:0}
.listingblock pre.highlightjs>code{padding:1em;-webkit-border-radius:4px;border-radius:4px}
.listingblock pre.prettyprint{border-width:0}
.prettyprint{background:#f7f7f8}
pre.prettyprint .linenums{line-height:1.45;margin-left:2em}
pre.prettyprint li{background:none;list-style-type:inherit;padding-left:0}
pre.prettyprint li code[data-lang]::before{opacity:1}
pre.prettyprint li:not(:first-child) code[data-lang]::before{display:none}
table.linenotable{border-collapse:separate;border:0;margin-bottom:0;background:none}
table.linenotable td[class]{color:inherit;vertical-align:top;padding:0;line-height:inherit;white-space:normal}
table.linenotable td.code{padding-left:.75em}
table.linenotable td.linenos{border-right:1px solid currentColor;opacity:.35;padding-right:.5em}
pre.pygments .lineno{border-right:1px solid currentColor;opacity:.35;display:inline-block;margin-right:.75em}
pre.pygments .lineno::before{content:"";margin-right:-.125em}
.quoteblock{margin:0 1em 1.25em 1.5em;display:table}
.quoteblock:not(.excerpt)>.title{margin-left:-1.5em;margin-bottom:.75em}
.quoteblock blockquote,.quoteblock p{color:rgba(0,0,0,.85);font-size:1.15rem;line-height:1.75;word-spacing:.1em;letter-spacing:0;font-style:italic;text-align:justify}
.quoteblock blockquote{margin:0;padding:0;border:0}
.quoteblock blockquote::before{content:"\201c";float:left;font-size:2.75em;font-weight:bold;line-height:.6em;margin-left:-.6em;color:#7a2518;text-shadow:0 1px 2px rgba(0,0,0,.1)}
.quoteblock blockquote>.paragraph:last-child p{margin-bottom:0}
.quoteblock .attribution{margin-top:.75em;margin-right:.5ex;text-align:right}
.verseblock{margin:0 1em 1.25em}
.verseblock pre{font-family:"Open Sans","DejaVu Sans",sans;font-size:1.15rem;color:rgba(0,0,0,.85);font-weight:300;text-rendering:optimizeLegibility}
.verseblock pre strong{font-weight:400}
.verseblock .attribution{margin-top:1.25rem;margin-left:.5ex}
.quoteblock .attribution,.verseblock .attribution{font-size:.9375em;line-height:1.45;font-style:italic}
.quoteblock .attribution br,.verseblock .attribution br{display:none}
.quoteblock .attribution cite,.verseblock .attribution cite{display:block;letter-spacing:-.025em;color:rgba(0,0,0,.6)}
.quoteblock.abstract blockquote::before,.quoteblock.excerpt blockquote::before,.quoteblock .quoteblock blockquote::before{display:none}
.quoteblock.abstract blockquote,.quoteblock.abstract p,.quoteblock.excerpt blockquote,.quoteblock.excerpt p,.quoteblock .quoteblock blockquote,.quoteblock .quoteblock p{line-height:1.6;word-spacing:0}
.quoteblock.abstract{margin:0 1em 1.25em;display:block}
.quoteblock.abstract>.title{margin:0 0 .375em;font-size:1.15em;text-align:center}
.quoteblock.excerpt>blockquote,.quoteblock .quoteblock{padding:0 0 .25em 1em;border-left:.25em solid #dddddf}
.quoteblock.excerpt,.quoteblock .quoteblock{margin-left:0}
.quoteblock.excerpt blockquote,.quoteblock.excerpt p,.quoteblock .quoteblock blockquote,.quoteblock .quoteblock p{color:inherit;font-size:1.0625rem}
.quoteblock.excerpt .attribution,.quoteblock .quoteblock .attribution{color:inherit;text-align:left;margin-right:0}
p.tableblock:last-child{margin-bottom:0}
td.tableblock>.content{margin-bottom:1.25em;word-wrap:anywhere}
td.tableblock>.content>:last-child{margin-bottom:-1.25em}
table.tableblock,th.tableblock,td.tableblock{border:0 solid #dedede}
table.grid-all>*>tr>*{border-width:1px}
table.grid-cols>*>tr>*{border-width:0 1px}
table.grid-rows>*>tr>*{border-width:1px 0}
table.frame-all{border-width:1px}
table.frame-ends{border-width:1px 0}
table.frame-sides{border-width:0 1px}
table.frame-none>colgroup+*>:first-child>*,table.frame-sides>colgroup+*>:first-child>*{border-top-width:0}
table.frame-none>:last-child>:last-child>*,table.frame-sides>:last-child>:last-child>*{border-bottom-width:0}
table.frame-none>*>tr>:first-child,table.frame-ends>*>tr>:first-child{border-left-width:0}
table.frame-none>*>tr>:last-child,table.frame-ends>*>tr>:last-child{border-right-width:0}
table.stripes-all tr,table.stripes-odd tr:nth-of-type(odd),table.stripes-even tr:nth-of-type(even),table.stripes-hover tr:hover{background:#f8f8f7}
th.halign-left,td.halign-left{text-align:left}
th.halign-right,td.halign-right{text-align:right}
th.halign-center,td.halign-center{text-align:center}
th.valign-top,td.valign-top{vertical-align:top}
th.valign-bottom,td.valign-bottom{vertical-align:bottom}
th.valign-middle,td.valign-middle{vertical-align:middle}
table thead th,table tfoot th{font-weight:bold}
tbody tr th{background:#f7f8f7}
tbody tr th,tbody tr th p,tfoot tr th,tfoot tr th p{color:rgba(0,0,0,.8);font-weight:bold}
p.tableblock>code:only-child{background:none;padding:0}
p.tableblock{font-size:1em}
ol{margin-left:1.75em}
ul li ol{margin-left:1.5em}
dl dd{margin-left:1.125em}
dl dd:last-child,dl dd:last-child>:last-child{margin-bottom:0}
ol>li p,ul>li p,ul dd,ol dd,.olist .olist,.ulist .ulist,.ulist .olist,.olist .ulist{margin-bottom:.625em}
ul.checklist,ul.none,ol.none,ul.no-bullet,ol.no-bullet,ol.unnumbered,ul.unstyled,ol.unstyled{list-style-type:none}
ul.no-bullet,ol.no-bullet,ol.unnumbered{margin-left:.625em}
ul.unstyled,ol.unstyled{margin-left:0}
ul.checklist{margin-left:.625em}
ul.checklist li>p:first-child>.fa-square-o:first-child,ul.checklist li>p:first-child>.fa-check-square-o:first-child{width:1.25em;font-size:.8em;position:relative;bottom:.125em}
ul.checklist li>p:first-child>input[type="checkbox"]:first-child{margin-right:.25em}
ul.inline{display:-ms-flexbox;display:-webkit-box;display:flex;-ms-flex-flow:row wrap;-webkit-flex-flow:row wrap;flex-flow:row wrap;list-style:none;margin:0 0 .625em -1.25em}
ul.inline>li{margin-left:1.25em}
.unstyled dl dt{font-weight:400;font-style:normal}
ol.arabic{list-style-type:decimal}
ol.decimal{list-style-type:decimal-leading-zero}
ol.loweralpha{list-style-type:lower-alpha}
ol.upperalpha{list-style-type:upper-alpha}
ol.lowerroman{list-style-type:lower-roman}
ol.upperroman{list-style-type:upper-roman}
ol.lowergreek{list-style-type:lower-greek}
.hdlist>table,.colist>table{border:0;background:none}
.hdlist>table>tbody>tr,.colist>table>tbody>tr{background:none}
td.hdlist1,td.hdlist2{vertical-align:top;padding:0 .625em}
td.hdlist1{font-weight:bold;padding-bottom:1.25em}
td.hdlist2{word-wrap:anywhere}
.literalblock+.colist,.listingblock+.colist{margin-top:-.5em}
.colist td:not([class]):first-child{padding:.4em .75em 0;line-height:1;vertical-align:top}
.colist td:not([class]):first-child img{max-width:none}
.colist td:not([class]):last-child{padding:.25em 0}
.thumb,.th{line-height:0;display:inline-block;border:solid 4px #fff;-webkit-box-shadow:0 0 0 1px #ddd;box-shadow:0 0 0 1px #ddd}
.imageblock.left{margin:.25em .625em 1.25em 0}
.imageblock.right{margin:.25em 0 1.25em .625em}
.imageblock>.title{margin-bottom:0}
.imageblock.thumb,.imageblock.th{border-width:6px}
.imageblock.thumb>.title,.imageblock.th>.title{padding:0 .125em}
.image.left,.image.right{margin-top:.25em;margin-bottom:.25em;display:inline-block;line-height:0}
.image.left{margin-right:.625em}
.image.right{margin-left:.625em}
a.image{text-decoration:none;display:inline-block}
a.image object{pointer-events:none}
sup.footnote,sup.footnoteref{font-size:.875em;position:static;vertical-align:super}
sup.footnote a,sup.footnoteref a{text-decoration:none}
sup.footnote a:active,sup.footnoteref a:active{text-decoration:underline}
#footnotes{padding-top:.75em;padding-bottom:.75em;margin-bottom:.625em}
#footnotes hr{width:20%;min-width:6.25em;margin:-.25em 0 .75em;border-width:1px 0 0}
#footnotes .footnote{padding:0 .375em 0 .225em;line-height:1.3334;font-size:.875em;margin-left:1.2em;margin-bottom:.2em}
#footnotes .footnote a:first-of-type{font-weight:bold;text-decoration:none;margin-left:-1.05em}
#footnotes .footnote:last-of-type{margin-bottom:0}
#content #footnotes{margin-top:-.625em;margin-bottom:0;padding:.75em 0}
.gist .file-data>table{border:0;background:#fff;width:100%;margin-bottom:0}
.gist .file-data>table td.line-data{width:99%}
div.unbreakable{page-break-inside:avoid}
.big{font-size:larger}
.small{font-size:smaller}
.underline{text-decoration:underline}
.overline{text-decoration:overline}
.line-through{text-decoration:line-through}
.aqua{color:#00bfbf}
.aqua-background{background:#00fafa}
.black{color:#000}
.black-background{background:#000}
.blue{color:#0000bf}
.blue-background{background:#0000fa}
.fuchsia{color:#bf00bf}
.fuchsia-background{background:#fa00fa}
.gray{color:#606060}
.gray-background{background:#7d7d7d}
.green{color:#006000}
.green-background{background:#007d00}
.lime{color:#00bf00}
.lime-background{background:#00fa00}
.maroon{color:#600000}
.maroon-background{background:#7d0000}
.navy{color:#000060}
.navy-background{background:#00007d}
.olive{color:#606000}
.olive-background{background:#7d7d00}
.purple{color:#600060}
.purple-background{background:#7d007d}
.red{color:#bf0000}
.red-background{background:#fa0000}
.silver{color:#909090}
.silver-background{background:#bcbcbc}
.teal{color:#006060}
.teal-background{background:#007d7d}
.white{color:#bfbfbf}
.white-background{background:#fafafa}
.yellow{color:#bfbf00}
.yellow-background{background:#fafa00}
span.icon>.fa{cursor:default}
a span.icon>.fa{cursor:inherit}
.admonitionblock td.icon [class^="fa icon-"]{font-size:2.5em;text-shadow:1px 1px 2px rgba(0,0,0,.5);cursor:default}
.admonitionblock td.icon .icon-note::before{content:"\f05a";color:#19407c}
.admonitionblock td.icon .icon-tip::before{content:"\f0eb";text-shadow:1px 1px 2px rgba(155,155,0,.8);color:#111}
.admonitionblock td.icon .icon-warning::before{content:"\f071";color:#bf6900}
.admonitionblock td.icon .icon-caution::before{content:"\f06d";color:#bf3400}
.admonitionblock td.icon .icon-important::before{content:"\f06a";color:#bf0000}
.conum[data-value]{display:inline-block;color:#fff!important;background:rgba(0,0,0,.8);-webkit-border-radius:50%;border-radius:50%;text-align:center;font-size:.75em;width:1.67em;height:1.67em;line-height:1.67em;font-family:"Open Sans","DejaVu Sans",sans-serif;font-style:normal;font-weight:bold}
.conum[data-value] *{color:#fff!important}
.conum[data-value]+b{display:none}
.conum[data-value]::after{content:attr(data-value)}
pre .conum[data-value]{position:relative;top:-.125em}
b.conum *{color:inherit!important}
.conum:not([data-value]):empty{display:none}
dt,th.tableblock,td.content,div.footnote{text-rendering:optimizeLegibility}
h1,h2,p,td.content,span.alt{letter-spacing:-.01em}
p strong,td.content strong,div.footnote strong{letter-spacing:-.005em}
p,blockquote,dt,td.content,span.alt{font-size:1.0625rem}
p{margin-bottom:1.25rem}
.sidebarblock p,.sidebarblock dt,.sidebarblock td.content,p.tableblock{font-size:1em}
.exampleblock>.content{background:#fffef7;border-color:#e0e0dc;-webkit-box-shadow:0 1px 4px #e0e0dc;box-shadow:0 1px 4px #e0e0dc}
.print-only{display:none!important}
@page{margin:1.25cm .75cm}
@media print{*{-webkit-box-shadow:none!important;box-shadow:none!important;text-shadow:none!important}
html{font-size:80%}
a{color:inherit!important;text-decoration:underline!important}
a.bare,a[href^="#"],a[href^="mailto:"]{text-decoration:none!important}
a[href^="http:"]:not(.bare)::after,a[href^="https:"]:not(.bare)::after{content:"(" attr(href) ")";display:inline-block;font-size:.875em;padding-left:.25em}
abbr[title]::after{content:" (" attr(title) ")"}
pre,blockquote,tr,img,object,svg{page-break-inside:avoid}
thead{display:table-header-group}
svg{max-width:100%}
p,blockquote,dt,td.content{font-size:1em;orphans:3;widows:3}
h2,h3,#toctitle,.sidebarblock>.content>.title{page-break-after:avoid}
#header,#content,#footnotes,#footer{max-width:none}
#toc,.sidebarblock,.exampleblock>.content{background:none!important}
#toc{border-bottom:1px solid #dddddf!important;padding-bottom:0!important}
body.book #header{text-align:center}
body.book #header>h1:first-child{border:0!important;margin:2.5em 0 1em}
body.book #header .details{border:0!important;display:block;padding:0!important}
body.book #header .details span:first-child{margin-left:0!important}
body.book #header .details br{display:block}
body.book #header .details br+span::before{content:none!important}
body.book #toc{border:0!important;text-align:left!important;padding:0!important;margin:0!important}
body.book #toc,body.book #preamble,body.book h1.sect0,body.book .sect1>h2{page-break-before:always}
.listingblock code[data-lang]::before{display:block}
#footer{padding:0 .9375em}
.hide-on-print{display:none!important}
.print-only{display:block!important}
.hide-for-print{display:none!important}
.show-for-print{display:inherit!important}}
@media print,amzn-kf8{#header>h1:first-child{margin-top:1.25rem}
.sect1{padding:0!important}
.sect1+.sect1{border:0}
#footer{background:none}
#footer-text{color:rgba(0,0,0,.6);font-size:.9em}}
@media amzn-kf8{#header,#content,#footnotes,#footer{padding:0}}
</style>
</head>
<body class="article toc2 toc-left">
<div id="header">
<h1>Reference Documentation for ODGI v0.6.0</h1>
<div class="details">
<span id="author" class="author">Simon Heumos, Andrea Guarracino, Erik Garrison</span><br>
</div>
<div id="toc" class="toc2">
<div id="toctitle">Table of Contents</div>
<ul class="sectlevel1">
<li><a href="#_odgi_1">1. odgi (1)</a></li>
<li><a href="#_odgi_build1">2. odgi build(1)</a></li>
<li><a href="#_odgi_stats1">3. odgi stats(1)</a></li>
<li><a href="#_odgi_sort1">4. odgi sort(1)</a></li>
<li><a href="#_odgi_view1">5. odgi view(1)</a></li>
<li><a href="#_odgi_kmers1">6. odgi kmers(1)</a></li>
<li><a href="#_odgi_unitig1">7. odgi unitig(1)</a></li>
<li><a href="#_odgi_viz1">8. odgi viz(1)</a></li>
<li><a href="#_odgi_paths1">9. odgi paths(1)</a></li>
<li><a href="#_odgi_prune1">10. odgi prune(1)</a></li>
<li><a href="#_odgi_unchop1">11. odgi unchop(1)</a></li>
<li><a href="#_odgi_normalize1">12. odgi normalize(1)</a></li>
<li><a href="#_odgi_subset1">13. odgi subset(1)</a></li>
<li><a href="#_odgi_bin1">14. odgi bin(1)</a></li>
<li><a href="#_odgi_matrix1">15. odgi matrix(1)</a></li>
<li><a href="#_odgi_chop1">16. odgi chop(1)</a></li>
<li><a href="#_odgi_layout1">17. odgi layout(1)</a></li>
<li><a href="#_odgi_flatten1">18. odgi flatten(1)</a></li>
<li><a href="#_odgi_break1">19. odgi break(1)</a></li>
<li><a href="#_odgi_pathindex1">20. odgi pathindex(1)</a></li>
<li><a href="#_odgi_panpos1">21. odgi panpos(1)</a></li>
<li><a href="#_odgi_server1">22. odgi server(1)</a></li>
<li><a href="#_odgi_test1">23. odgi test(1)</a></li>
<li><a href="#_odgi_version1">24. odgi version(1)</a></li>
</ul>
</div>
</div>
<div id="content">
<div class="sect1">
<h2 id="_odgi_1">1. odgi (1)</h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="_name">1.1. NAME</h3>
<div class="paragraph">
<p>odgi - dynamic succinct variation graph tool</p>
</div>
</div>
<div class="sect2">
<h3 id="_synopsis">1.2. SYNOPSIS</h3>
<div class="paragraph">
<p><strong>odgi</strong> <a href="#_odgi_build1">build</a> -g graph.gfa -o graph.og</p>
</div>
<div class="paragraph">
<p><strong>odgi</strong> <a href="#_odgi_stats1">stats</a> -i graph.og -S</p>
</div>
<div class="paragraph">
<p><strong>odgi</strong> <a href="odgi_cover.html#_odgi_cover1">cover</a> -i graph.og -o graph.paths.og</p>
</div>
<div class="paragraph">
<p><strong>odgi</strong> <a href="odgi_cover.html#_odgi_explode1">explode</a> -i graph.og -p prefix</p>
</div>
<div class="paragraph">
<p><strong>odgi</strong> <a href="odgi_cover.html#_odgi_squeeze1">squeeze</a> -f input_graphs.txt -o graphs.og</p>
</div>
<div class="paragraph">
<p><strong>odgi</strong> <a href="#_odgi_position1">position</a> -i target_graph.og -g</p>
</div>
<div class="paragraph">
<p><strong>odgi</strong> <a href="#_odgi_sort1">sort</a> -i graph.og -o graph.sorted.og -p bSnSnS</p>
</div>
<div class="paragraph">
<p><strong>odgi</strong> <a href="#_odgi_view1">view</a> -i graph.og -g</p>
</div>
<div class="paragraph">
<p><strong>odgi</strong> <a href="#_odgi_kmers1">kmers</a> -i graph.og -c -k 23 -e 34 -D 50</p>
</div>
<div class="paragraph">
<p><strong>odgi</strong> <a href="#_odgi_unitig1">unitig</a> -i graph.og -f -t 1324 -l 120</p>
</div>
<div class="paragraph">
<p><strong>odgi</strong> <a href="#_odgi_viz1">viz</a> -i graph.og -o graph.og.png -x 1920 -y 1080 -R -t 28</p>
</div>
<div class="paragraph">
<p><strong>odgi</strong> <a href="#_odgi_paths1">paths</a> -i graph.og -f</p>
</div>
<div class="paragraph">
<p><strong>odgi</strong> <a href="#_odgi_prune1">prune</a> -i graph.og -o graph.pruned.og -c 3 -C 345 -T</p>
</div>
<div class="paragraph">
<p><strong>odgi</strong> <a href="#_odgi_unchop1">unchop</a> -i graph.og -o graph.unchopped.og</p>
</div>
<div class="paragraph">
<p><strong>odgi</strong> <a href="#_odgi_normalize1">normalize</a> -i graph.og -o graph.normalized.og -I 100 -d</p>
</div>
<div class="paragraph">
<p><strong>odgi</strong> <a href="#_odgi_subset1">subset</a> -i graph.og -o graph.subsetted.og -n 1 -t 28</p>
</div>
<div class="paragraph">
<p><strong>odgi</strong> <a href="#_odgi_bin1">bin</a> -i graph.og -j -w 100 -s -g</p>
</div>
<div class="paragraph">
<p><strong>odgi</strong> <a href="#_odgi_matrix1">matrix</a> -i graph.og -e -d</p>
</div>
<div class="paragraph">
<p><strong>odgi</strong> <a href="#_odgi_chop1">chop</a> -i graph.og -o graph.choped.og -c 1000</p>
</div>
<div class="paragraph">
<p><strong>odgi</strong> <a href="odgi_groom.html#_odgi_groom1">groom</a> -i graph.og -o graph.groomed.og</p>
</div>
<div class="paragraph">
<p><strong>odgi</strong> <a href="#_odgi_layout1">layout</a> -i graph.og -o graph.svg -R 10 -m 100</p>
</div>
<div class="paragraph">
<p><strong>odgi</strong> <a href="#_odgi_break1">break</a> -i graph.og -o graph.broken.og -s 100 -d</p>
</div>
<div class="paragraph">
<p><strong>odgi</strong> <a href="#_odgi_pathindex1">pathindex</a> -i graph.og -o graph.xp</p>
</div>
<div class="paragraph">
<p><strong>odgi</strong> <a href="#_odgi_panpos1">panpos</a> -i graph.og -p Chr1 -n 4</p>
</div>
<div class="paragraph">
<p><strong>odgi</strong> <a href="#_odgi_server1">server</a> -i graph.og -p 4000 -ip 192.168.8.9</p>
</div>
<div class="paragraph">
<p><strong>odgi</strong> <a href="#_odgi_test1">test</a></p>
</div>
<div class="paragraph">
<p><strong>odgi</strong> <a href="#_odgi_version1">version</a></p>
</div>
</div>
<div class="sect2">
<h3 id="_description">1.3. DESCRIPTION</h3>
<div class="paragraph">
<p><strong>odgi</strong>, the <strong>Optimized Dynamic (genome) Graph Interface</strong>, links
a thrifty dynamic in-memory variation graph data model to a set of algorithms designed for scalable sorting, pruning,
transformation, and visualization of very large <a href="https://pangenome.github.io/">genome graphs</a>. <strong>odgi</strong> includes <a href="https://pangenome.github.io/odgi/odgipy.html">python bindings</a>
that can be used to
<a href="https://odgi.readthedocs.io/en/latest/rst/tutorial.html">directly
interface with its data model</a>. This <strong>odgi</strong> manual provides detailed information about its features and subcommands, including examples.</p>
</div>
</div>
<div class="sect2">
<h3 id="_commands">1.4. COMMANDS</h3>
<div class="paragraph">
<p>Each command has its own man page which can be viewed using e.g. <strong>man odgi_build.1</strong>. Below we have a brief summary of syntax and subcommand description.</p>
</div>
<div class="paragraph">
<p><strong>odgi build</strong> [<strong>-g, --gfa</strong>=<em>FILE</em>] [<strong>-o, --out</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;<br>
The odgi build(1) command constructs a succinct variation graph from a GFA. Currently, only GFA1 is supported. For details of the format please see <a href="https://github.com/GFA-spec/GFA-spec/blob/master/GFA1.md" class="bare">https://github.com/GFA-spec/GFA-spec/blob/master/GFA1.md</a>.</p>
</div>
<div class="paragraph">
<p><strong>odgi stats</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;<br>
The odgi stats(1) command produces statistics of a variation graph. Among other metrics, it can calculate the #nodes, #edges, #paths and the total nucleotide length of the graph. Various histogram summary options complement the tool. If [<strong>-B, --bed-multicov</strong>=<em>BED</em>] is set, the metrics will be produced for the intervals specified in the BED.</p>
</div>
<div class="paragraph">
<p><strong>odgi cover</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<strong>-o, --out</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;<br>
The odgi cover(1) command find a path cover of a variation graph, with a specified number of paths per component.</p>
</div>
<div class="paragraph">
<p><strong>odgi explode</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<strong>-p, --prefix</strong>=<em>STRING</em>] [<em>OPTION</em>]&#8230;&#8203;<br>
The odgi explode(1) command breaks a graph into connected components, writing each component in its own file.</p>
</div>
<div class="paragraph">
<p><strong>odgi squeeze</strong> [<strong>-f, --input-graphs</strong>=<em>FILE</em>] [<strong>-o, --out</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;
The odgi squeeze(1) command squeezes multiple graphs into the same file.</p>
</div>
<div class="paragraph">
<p><strong>odgi position</strong> [<strong>-i, --target</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;
The odgi position(1) command position parts of the graph as defined by query criteria.</p>
</div>
<div class="paragraph">
<p><strong>odgi sort</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<strong>-o, --out</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;<br>
The odgi sort(1) command sorts a succinct variation graph. Odgi sort offers a diverse palette of sorting algorithms to
determine the node order:</p>
</div>
<div class="ulist">
<ul>
<li>
<p>A topological sort: A graph can be sorted via <a href="https://en.wikipedia.org/wiki/Breadth-first_search">breadth-first search (BFS)</a> or <a href="https://en.wikipedia.org/wiki/Depth-first_search">depth-first search (DFS)</a>. Optionally,
a chunk size specifies how much of the graph to grab at once in each topological sorting phase. The sorting algorithm will continue the sort from the
next node in the prior graph order that has not been sorted, yet. The cycle breaking algorithm applies a DFS sort until
a cycle is found. We break and start a new DFS sort phase from where we stopped.</p>
</li>
<li>
<p>A random sort: The graph is randomly sorted. The node order is randomly shuffled from <a href="http://www.cplusplus.com/reference/random/mt19937/">Mersenne Twister pseudo-random</a> generated numbers.</p>
</li>
<li>
<p>A sparse matrix mondriaan sort: We can partition a hypergraph with integer weights and uniform hyperedge costs using the <a href="http://www.staff.science.uu.nl/~bisse101/Mondriaan/">Mondriaan</a> partitioner.</p>
</li>
<li>
<p>A 1D linear SGD sort: Odgi implements a 1D linear, variation graph adjusted, multi-threaded version of the <a href="https://arxiv.org/abs/1710.04626">Graph Drawing
by Stochastic Gradient Descent</a> algorithm. The force-directed graph drawing algorithm minimizes the graph&#8217;s energy function
or stress level. It applies stochastic gradient descent (SGD) to move a single pair of nodes at a time.</p>
</li>
<li>
<p>An eades algorithmic sort: Use <a href="http://www.it.usyd.edu.au/~pead6616/old_spring_paper.pdf">Peter Eades' heuristic for graph drawing</a>.</p>
</li>
</ul>
</div>
<div class="paragraph">
<p>Sorting the paths in a graph my refine the sorting process. For the users' convenience, it is possible to specify a whole
pipeline of sorts within one parameter.</p>
</div>
<div class="paragraph">
<p><strong>odgi view</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;<br>
The odgi view(1) command can convert a graph in odgi format to GFAv1. It can reveal a graph&#8217;s internal structures for e.g. debugging processes.</p>
</div>
<div class="paragraph">
<p><strong>odgi kmers</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<strong>-c, --stdout</strong>] [<em>OPTION</em>]&#8230;&#8203;<br>
Given a kmer length, the odgi kmers(1) command can emit all kmers. The output can be refined by setting the maximum number
of furcations at edges or by not considering nodes above a given node degree limit.</p>
</div>
<div class="paragraph">
<p><strong>odgi unitig</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;<br>
The odgi unitig(1) command can print all unitigs of a given odgi graph to standard output in FASTA format. Unitigs can also be emitted
in a fixed sequence quality FASTQ format. Various parameters can refine the unitigs to print.</p>
</div>
<div class="paragraph">
<p><strong>odgi viz</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<strong>-o, --out</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;<br>
The odgi viz(1) command can produce a linear, static visualization of an odgi variation graph. It aggregates the pangenome into bins
and directly renders a raster image. The binning level depends on the target width of the PNG to emit. Can be used to produce visualizations for gigabase scale pangenomes. For more information
about the binning process, please refer to <a href="#_odgi_bin1">odgi bin</a>. If reverse coloring was selected, only
the bins with a reverse rate of at least 0.5 are colored. Currently, there is no parameter to color according to the
sequence coverage in bins available.</p>
</div>
<div class="paragraph">
<p><strong>odgi paths</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;<br>
The odgi paths(1) command allows the investigation of paths of a given variation graph. It can calculate overlap statistics
of groupings of paths.</p>
</div>
<div class="paragraph">
<p><strong>odgi prune</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<strong>-o, --out</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;<br>
The odgi prune(1) command can remove complex parts of a graph. One can drop paths, nodes by a certain kind of edge coverage,
edges and graph tips. Specifying a kmer length and a maximum number of furcations, the graph can be broken at edges not
fitting into these conditions.</p>
</div>
<div class="paragraph">
<p><strong>odgi unchop</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<strong>-o, --out</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;<br>
The odgi unchop(1) command merges each unitig into a single node.</p>
</div>
<div class="paragraph">
<p><strong>odgi normalize</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<strong>-o, --out</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;<br>
The odgi normalize(1) command <a href="#_odgi_unchop1">unchops</a> a given variation graph and simplifies redundant furcations.</p>
</div>
<div class="paragraph">
<p><strong>odgi subset</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<strong>-o, --out</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;<br>
Extracting a node subset of a variation graph is the task of the odgi subset(1) command. Users can specify a node, a list of nodes
or a the context of which to generate a subset from.</p>
</div>
<div class="paragraph">
<p><strong>odgi matrix</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;<br>
The odgi matrix(1) command generates a sparse matrix format out of the graph topology of a given variation graph.</p>
</div>
<div class="paragraph">
<p><strong>odgi bin</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;<br>
The odgi bin(1) command bins a given variation graph. The pangenome sequence, the one-time traversal of all nodes from smallest to
largest node identifier, can be summed up into bins of a specified size. For each bin, the path metainformation is summarized.
This enables a summarized view of gigabase scale graphs. Each step of a path is a bin and connected to its next bin via a link.
A link has a start bin identifier and an end bin identifier.<br>
The concept of odgi bin is also applied in odgi <a href="#_odgi_viz1">viz</a>.
A demonstration of how the odgi bin JSON output can be used for an interactive visualization is realized in the <a href="https://graph-genome.github.io/">Pantograph</a>
project. Per default, odgi bin writes the bins to stdout in a tab-delimited format: <strong>path.name</strong>, <strong>path.prefix</strong>, <strong>path.suffix</strong>,
<strong>bin</strong> (bin identifier), <strong>mean.cov</strong> (mean coverage of the path in this bin), <strong>mean.inv</strong> (mean inversion rate of this path in this bin),
<strong>mean.pos</strong> (mean nucleotide position of this path in this bin), <strong>first.nucl</strong> (first nucleotide position of this path in this bin),
<strong>last.nucl</strong> (last nucleotide position of this path in this bin). These nucleotide ranges might span positions that are not present in the bin. Example:
A range of 1-100 means that the first nucleotide has position 1 and the last has position 100, but nucleotide 45 could be located in
another bin. For an exact positional output, please specify [<strong>-j, --json</strong>].</p>
</div>
<div class="paragraph">
<p><strong>odgi chop</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<strong>-o, --out</strong>=<em>FILE</em>] [<strong>-c, --chop-to</strong>=<em>N</em>] [<em>OPTION</em>]&#8230;&#8203;<br>
The odgi chop(1) command chops long nodes into short ones while preserving the graph topology.</p>
</div>
<div class="paragraph">
<p><strong>odgi layout</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<strong>-o, --out</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;<br>
The odgi layout(1) command draws 2D layouts of the graph using stochastic gradient descent (SGD). The input graph must be sorted
and id-compacted. The algorithm itself is described in <a href="https://arxiv.org/abs/1710.04626">Graph Drawing by Stochastic Gradient Descent</a>.
The force-directed graph drawing algorithm minimizes the graph&#8217;s energy function or stress level.
It applies SGD to move a single pair of nodes at a time. The rendered graph is written in SVG format.</p>
</div>
<div class="paragraph">
<p><strong>odgi flatten</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;<br>
The odgi flatten(1) command projects the graph sequence and paths into FASTA and BED.</p>
</div>
<div class="paragraph">
<p><strong>odgi break</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<strong>-o, --out</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;<br>
The odgi break(1) command finds cycles in a graph via <a href="https://en.wikipedia.org/wiki/Breadth-first_search">breadth-first search (BFS)</a> and breaks them, also dropping
the graph&#8217;s paths.</p>
</div>
<div class="paragraph">
<p><strong>odgi pathindex</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<strong>-o, --out</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;<br>
The odgi pathindex(1) command generates a path index of a graph. It uses succinct data structures to encode the index.
The path index represents a subset of the features of a fully realized <a href="https://github.com/vgteam/xg">xg index</a>. Having a path index, we can use
odgi <a href="#_odgi_panpos1">panpos</a> to go from <strong>path:position</strong> &#8594; <strong>pangenome:position</strong> which is important when
navigating large graphs in an interactive manner like in the <a href="https://graph-genome.github.io/">Pantograph</a> project.</p>
</div>
<div class="paragraph">
<p><strong>odgi panpos</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<strong>-p, --path</strong>=<em>STRING</em>] [<strong>-n, --nuc-pos</strong>=<em>N</em>] [<em>OPTION</em>]&#8230;&#8203;<br>
The odgi panpos(1) command give a pangenome position for a given path and nucleotide position. It requires a path index,
which can be created with odgi <a href="#_odgi_pathindex1">pathindex</a>. Going from <strong>path:position</strong> &#8594; <strong>pangenome:position</strong> is important when
navigating large graphs in an interactive manner like in the <a href="https://graph-genome.github.io/">Pantograph</a> project. All
input and output positions are 1-based.</p>
</div>
<div class="paragraph">
<p><strong>odgi server</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<strong>-p, --port</strong>=<em>N</em>] [<em>OPTION</em>]&#8230;&#8203;<br>
The odgi server(1) command starts an HTTP server with a given path index as input. The idea is that we can go from
<strong>path:position</strong> &#8594; <strong>pangenome:position</strong> via GET requests to the HTTP server. The server headers do not block cross origin requests.
Example GET request: <em><a href="http://localost:3000/path_name/nucleotide_position" class="bare">http://localost:3000/path_name/nucleotide_position</a></em>.<br>
The required path index can be created with odgi <a href="#_odgi_pathindex1">pathindex</a>. Going from <strong>path:position</strong> &#8594; <strong>pangenome:position</strong> is important when
navigating large graphs in an interactive manner like in the <a href="https://graph-genome.github.io/">Pantograph</a> project. All
input and output positions are 1-based. If no IP address is specified, the server will run on localhost.</p>
</div>
<div class="paragraph">
<p><strong>odgi test</strong> [&lt;TEST NAME|PATTERN|TAGS&gt; &#8230;&#8203;] [<em>OPTION</em>]&#8230;&#8203;<br>
The odgi test(1) command starts all unit tests that are implemented in odgi. For targeted testing, a subset of tests can
be selected. odgi test(1) depends on <a href="https://github.com/catchorg/Catch2">Catch2</a>. In the default setting, all results are printed to stdout.</p>
</div>
<div class="paragraph">
<p><strong>odgi version</strong> [<em>OPTION</em>]&#8230;&#8203;<br>
The odgi version(1) command prints the current git version with tags and codename to stdout (like <em>v-44-g89d022b "back to old ABI"</em>). Optionally, only the release, version or codename can be printed.</p>
</div>
</div>
<div class="sect2">
<h3 id="_bugs">1.5. BUGS</h3>
<div class="paragraph">
<p>Refer to the <strong>odgi</strong> issue tracker at <a href="https://github.com/vgteam/odgi/issues" class="bare">https://github.com/vgteam/odgi/issues</a>.</p>
</div>
</div>
<div class="sect2">
<h3 id="_authors">1.6. AUTHORS</h3>
<div class="paragraph">
<p>Erik Garrison from the University of California Santa Cruz wrote the whole <strong>odgi</strong> tool. Simon Heumos from the Quantitative Biology Center Tübingen wrote <strong>odgi pathindex</strong>, <strong>odgi panpos</strong>, <strong>odgi server</strong>, and this documentation. Andrea Guarracino from the University of Rome Tor Vergata wrote <strong>odgi viz</strong>, <strong>odgi cover</strong>, <strong>odgi explode</strong>, <strong>odgi squeeze</strong>, and this documentation*.</p>
</div>
</div>
<div class="sect2">
<h3 id="_resources">1.7. RESOURCES</h3>
<div class="paragraph">
<p><strong>Project web site:</strong> <a href="https://github.com/vgteam/odgi" class="bare">https://github.com/vgteam/odgi</a></p>
</div>
<div class="paragraph">
<p><strong>Git source repository on GitHub:</strong> <a href="https://github.com/vgteam/odgi" class="bare">https://github.com/vgteam/odgi</a></p>
</div>
<div class="paragraph">
<p><strong>GitHub organization:</strong> <a href="https://github.com/vgteam" class="bare">https://github.com/vgteam</a></p>
</div>
<div class="paragraph">
<p><strong>Discussion list / forum:</strong> <a href="https://github.com/vgteam/odgi/issues" class="bare">https://github.com/vgteam/odgi/issues</a></p>
</div>
</div>
<div class="sect2">
<h3 id="_copying">1.8. COPYING</h3>
<div class="paragraph">
<p>The MIT License (MIT)</p>
</div>
<div class="paragraph">
<p>Copyright (c) 2019 Erik Garrison</p>
</div>
<div class="paragraph">
<p>Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
the Software, and to permit persons to whom the Software is furnished to do so,
subject to the following conditions:</p>
</div>
<div class="paragraph">
<p>The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.</p>
</div>
<div class="paragraph">
<p>THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.</p>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_odgi_build1">2. odgi build(1)</h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="_name_2">2.1. NAME</h3>
<div class="paragraph">
<p>odgi_build - construct a dynamic succinct variation graph</p>
</div>
</div>
<div class="sect2">
<h3 id="_synopsis_2">2.2. SYNOPSIS</h3>
<div class="paragraph">
<p><strong>odgi build</strong> [<strong>-g, --gfa</strong>=<em>FILE</em>] [<strong>-o, --out</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;</p>
</div>
</div>
<div class="sect2">
<h3 id="_description_2">2.3. DESCRIPTION</h3>
<div class="paragraph">
<p>The odgi build(1) command constructs a succinct variation graph from a GFA. Currently, only GFA1 is supported. For details of the format please see <a href="https://github.com/GFA-spec/GFA-spec/blob/master/GFA1.md" class="bare">https://github.com/GFA-spec/GFA-spec/blob/master/GFA1.md</a>.</p>
</div>
</div>
<div class="sect2">
<h3 id="_options">2.4. OPTIONS</h3>
<div class="sect3">
<h4 id="_graph_files_io">2.4.1. Graph Files IO</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-g, --gfa</strong>=<em>FILE</em></dt>
<dd>
<p>GFA1 file containing the nodes, edges and paths to build a dynamic succinct variation graph from.</p>
</dd>
<dt class="hdlist1"><strong>-o, --out</strong>=<em>FILE</em></dt>
<dd>
<p>Write the dynamic succinct variation graph to this file. A file ending with <em>.og</em> is recommended.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_graph_sorting">2.4.2. Graph Sorting</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-s, --sort</strong></dt>
<dd>
<p>Apply a general topological sort to the graph and order the node ids accordingly. A bidirected adaptation of Kahn&#8217;s topological sort (1962) is used, which can handle components with no heads or tails. Here, both heads and tails are taken into account.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_processing_information">2.4.3. Processing Information</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-p, --progress</strong></dt>
<dd>
<p>Print progress updates to stdout.</p>
</dd>
<dt class="hdlist1"><strong>-d, --debug</strong></dt>
<dd>
<p>Verbosely print graph information to stderr. This includes the maximum node_id, the minimum node_id, the handle to node_id mapping, the deleted nodes and the path metadata.</p>
</dd>
<dt class="hdlist1"><strong>--trace</strong></dt>
<dd>
<p>Include backtrace information when reporting errors.</p>
</dd>
<dt class="hdlist1"><strong>-v, --verbose</strong></dt>
<dd>
<p>Verbosely print processing information to stderr, including debug-level log messages.</p>
</dd>
<dt class="hdlist1"><strong>-w, --warnings</strong></dt>
<dd>
<p>Turn on script warnings (applies to executed code).</p>
</dd>
<dt class="hdlist1"><strong>-t, --threads</strong>=<em>N</em></dt>
<dd>
<p>Number of threads to use for the parallel operations.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_program_information">2.4.4. Program Information</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-h, --help</strong></dt>
<dd>
<p>Print a help message for <strong>odgi build</strong>.</p>
</dd>
</dl>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_exit_status">2.5. EXIT STATUS</h3>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>0</strong></dt>
<dd>
<p>Success.</p>
</dd>
<dt class="hdlist1"><strong>1</strong></dt>
<dd>
<p>Failure (syntax or usage error; parameter error; file processing failure; unexpected error).</p>
</dd>
</dl>
</div>
</div>
<div class="sect2">
<h3 id="_bugs_2">2.6. BUGS</h3>
<div class="paragraph">
<p>Refer to the <strong>odgi</strong> issue tracker at <a href="https://github.com/vgteam/odgi/issues" class="bare">https://github.com/vgteam/odgi/issues</a>.</p>
</div>
</div>
<div class="sect2">
<h3 id="_authors_2">2.7. AUTHORS</h3>
<div class="paragraph">
<p><strong>odgi build</strong> was written by Erik Garrison.</p>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_odgi_stats1">3. odgi stats(1)</h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="_name_3">3.1. NAME</h3>
<div class="paragraph">
<p>odgi_stats - metrics describing variation graphs</p>
</div>
</div>
<div class="sect2">
<h3 id="_synopsis_3">3.2. SYNOPSIS</h3>
<div class="paragraph">
<p><strong>odgi stats</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;</p>
</div>
</div>
<div class="sect2">
<h3 id="_description_3">3.3. DESCRIPTION</h3>
<div class="paragraph">
<p>The odgi stats(1) command produces statistics of a variation graph. Among other metrics, it can calculate the #nodes, #edges, #paths and the total nucleotide length of the graph. Various histogram summary options complement the tool. If [<strong>-B, --bed-multicov</strong>=<em>BED</em>] is set, the metrics will be produced for the intervals specified in the BED.</p>
</div>
</div>
<div class="sect2">
<h3 id="_options_2">3.4. OPTIONS</h3>
<div class="sect3">
<h4 id="_graph_files_io_2">3.4.1. Graph Files IO</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-i, --idx</strong>=<em>FILE</em></dt>
<dd>
<p>File containing the succinct variation graph to create statistics from. The file name usually ends with <em>.og</em>.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_summary_options">3.4.2. Summary Options</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-S, --summarize</strong></dt>
<dd>
<p>Summarize the graph properties and dimensions. Print to stdout the #nucleotides, #nodes, #edges and #paths of the graph.</p>
</dd>
<dt class="hdlist1"><strong>-W, --weak-connected-components</strong></dt>
<dd>
<p>Shows the properties of the weakly connected components.</p>
</dd>
<dt class="hdlist1"><strong>-b, --base-content</strong></dt>
<dd>
<p>Describe the base content of the graph. Print to stdout the #A, #C, #G and #T of the graph.</p>
</dd>
<dt class="hdlist1"><strong>-C, --coverage</strong></dt>
<dd>
<p>Provide a histogram of path coverage over bases in the graph. Print three tab-delimited columns to stdout: <strong>type</strong>, <strong>cov</strong>, <strong>N</strong>. <strong>type</strong> is one of <em>full</em> or <em>uniq</em> and determines if the histogram corresponds to the full graph or only to a unique paths graph. <strong>cov</strong> implies the #paths. <strong>N</strong> implies the #nucleotides.</p>
</dd>
<dt class="hdlist1"><strong>-V, --set-coverage</strong></dt>
<dd>
<p>Provide a histogram of coverage over unique set of paths. Print two tab-delimited columns to stdout: <strong>cov</strong>, <strong>sets</strong>. <strong>cov</strong> implies #nucleotides. <strong>sets</strong> lists the unique set of paths in a comma separated list. Sets with a <strong>cov</strong> of one and no paths in <strong>sets</strong> are listed, too.</p>
</dd>
<dt class="hdlist1"><strong>-M, --multi-coverage</strong></dt>
<dd>
<p>Provide a histogram of coverage over unique multiset, the combination with possible repetition of paths. Print two tab-delimited columns to stdout: <strong>cov</strong>, <strong>sets</strong>. <strong>cov</strong> implies #nucleotides. <strong>sets</strong> lists the unique multisets of paths in a comma separated list. Multisets with a <strong>cov</strong> of one and no paths in <strong>sets</strong> are listed, too.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_bed_interval">3.4.3. BED Interval</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-B, --bed-multicov</strong>=<em>BED</em></dt>
<dd>
<p>For each BED entry, provide a table of path coverage over unique multisets of paths in the graph. Each unique multiset of paths overlapping a given BED interval is described in terms of its length relative to the total interval, the number of path traversals and unique paths involved in these traversals.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_sorting_goodness_evaluation">3.4.4. Sorting goodness evaluation</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-l, --mean-links-length</strong></dt>
<dd>
<p>Calculate the mean links length. This metric is path-guided and computable in 1D and 2D.</p>
</dd>
<dt class="hdlist1"><strong>-g, --no-gap-links</strong></dt>
<dd>
<p>Don&#8217;t penalize gap links in the mean links length. A gap link is a link which connects two nodes that are consecutive in the linear pangenomic order. This option is specifiable only to compute the mean links length in 1D.</p>
</dd>
<dt class="hdlist1"><strong>-s, --sum-path-nodes-distances</strong></dt>
<dd>
<p>Calculate the sum of path nodes distances. This metric is path-guided and computable in 1D and 2D. For each path, it iterates from node to node, summing their distances, and normalizing by the path length. In 1D, if a link goes back in the linearized viewpoint of the graph, this is penalized (adding 3 times its length in the sum).</p>
</dd>
<dt class="hdlist1"><strong>-d, --penalize-different-orientation</strong></dt>
<dd>
<p>If a link connects two nodes which have different orientations, this is penalized (adding 2 times its length in the sum).</p>
</dd>
<dt class="hdlist1"><strong>-c, --coords-in</strong></dt>
<dd>
<p>File containing the layout coordinates of the succinct variation graph specified as input. The file name usually ends with <em>.lay</em>. When the layout coordinates are provided, the mean links length and the sum path nodes distances statistics are evaluated in 2D, else in 1D.</p>
</dd>
<dt class="hdlist1"><strong>-P, --path-statistics</strong></dt>
<dd>
<p>Display the statistics (mean links length or sum path nodes distances) for each path.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_threading">3.4.5. Threading</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-t, --threads</strong>=<em>N</em></dt>
<dd>
<p>Number of threads to use.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_program_information_2">3.4.6. Program Information</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-h, --help</strong></dt>
<dd>
<p>Print a help message for <strong>odgi stats</strong>.</p>
</dd>
</dl>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_exit_status_2">3.5. EXIT STATUS</h3>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>0</strong></dt>
<dd>
<p>Success.</p>
</dd>
<dt class="hdlist1"><strong>1</strong></dt>
<dd>
<p>Failure (syntax or usage error; parameter error; file processing failure; unexpected error).</p>
</dd>
</dl>
</div>
</div>
<div class="sect2">
<h3 id="_bugs_3">3.6. BUGS</h3>
<div class="paragraph">
<p>Refer to the <strong>odgi</strong> issue tracker at <a href="https://github.com/vgteam/odgi/issues" class="bare">https://github.com/vgteam/odgi/issues</a>.</p>
</div>
</div>
<div class="sect2">
<h3 id="_authors_3">3.7. AUTHORS</h3>
<div class="paragraph">
<p><strong>odgi stats</strong> was written by Erik Garrison.</p>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_odgi_sort1">4. odgi sort(1)</h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="_name_4">4.1. NAME</h3>
<div class="paragraph">
<p>odgi_sort - sort a variation graph</p>
</div>
</div>
<div class="sect2">
<h3 id="_synopsis_4">4.2. SYNOPSIS</h3>
<div class="paragraph">
<p><strong>odgi sort</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<strong>-o, --out</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;</p>
</div>
</div>
<div class="sect2">
<h3 id="_description_4">4.3. DESCRIPTION</h3>
<div class="paragraph">
<p>The odgi sort(1) command sorts a succinct variation graph. Odgi sort offers a diverse palette of sorting algorithms to
determine the node order:</p>
</div>
<div class="ulist">
<ul>
<li>
<p>A topological sort: A graph can be sorted via <a href="https://en.wikipedia.org/wiki/Breadth-first_search">breadth-first search (BFS)</a> or <a href="https://en.wikipedia.org/wiki/Depth-first_search">depth-first search (DFS)</a>. Optionally,
a chunk size specifies how much of the graph to grab at once in each topological sorting phase. The sorting algorithm will continue the sort from the
next node in the prior graph order that has not been sorted, yet. The cycle breaking algorithm applies a DFS sort until
a cycle is found. We break and start a new DFS sort phase from where we stopped.</p>
</li>
<li>
<p>A random sort: The graph is randomly sorted. The node order is randomly shuffled from <a href="http://www.cplusplus.com/reference/random/mt19937/">Mersenne Twister pseudo-random</a> generated numbers.</p>
</li>
<li>
<p>A 1D linear SGD sort: Odgi implements a 1D linear, variation graph adjusted, multi-threaded version of the <a href="https://arxiv.org/abs/1710.04626">Graph Drawing
by Stochastic Gradient Descent</a> algorithm. The force-directed graph drawing algorithm minimizes the graph&#8217;s energy function
or stress level. It applies stochastic gradient descent (SGD) to move a single pair of nodes at a time.</p>
</li>
<li>
<p>A path guided, 1D linear SGD sort: Odgi implements a 1D linear, variation graph adjusted, multi-threaded version of the <a href="https://arxiv.org/abs/1710.04626">Graph Drawing
 by Stochastic Gradient Descent</a> algorithm. The force-directed graph drawing algorithm minimizes the graph&#8217;s energy function
 or stress level. It applies stochastic gradient descent (SGD) to move a single pair of nodes at a time. The path index is used to pick the terms to move stochastically. If ran with 1 thread only,
the resulting order of the graph is deterministic. The seed is adjustable.</p>
</li>
</ul>
</div>
<div class="paragraph">
<p>Sorting the paths in a graph my refine the sorting process. For the users' convenience, it is possible to specify a whole
pipeline of sorts within one parameter.</p>
</div>
</div>
<div class="sect2">
<h3 id="_options_3">4.4. OPTIONS</h3>
<div class="sect3">
<h4 id="_graph_files_io_3">4.4.1. Graph Files IO</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-i, --idx</strong>=<em>FILE</em></dt>
<dd>
<p>File containing the succinct variation graph to sort. The file name usually ends with <em>.og</em>.</p>
</dd>
<dt class="hdlist1"><strong>-o, --out</strong>=<em>FILE</em></dt>
<dd>
<p>Write the sorted dynamic succinct variation graph to this file. A file ending with <em>.og</em> is recommended.</p>
</dd>
<dt class="hdlist1"><strong>-s, --sort-order</strong>=<em>FILE</em></dt>
<dd>
<p>File containing the sort order. Each line contains one node identifier.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_topological_sorts">4.4.2. Topological Sorts</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-b, --breadth-first</strong></dt>
<dd>
<p>Use a (chunked) breadth first topological sort.</p>
</dd>
<dt class="hdlist1"><strong>-B, --breadth-first-chunk</strong>=<em>N</em></dt>
<dd>
<p>Chunk size for breadth first topological sort. Specify how many nucleotides to grap at once in each BFS phase.</p>
</dd>
<dt class="hdlist1"><strong>-z, --depth-first</strong></dt>
<dd>
<p>Use a (chunked) depth first topological sort.</p>
</dd>
<dt class="hdlist1"><strong>-Z, --depth-first-chunk</strong>=<em>N</em></dt>
<dd>
<p>Chunk size for the depth first topological sort. Specify how many nucleotides to grap at once in each DFS phace.</p>
</dd>
<dt class="hdlist1"><strong>-w, --two-way</strong></dt>
<dd>
<p>Use a two-way topological algorithm for sorting. It is a maximum of head-first and tail-first topological sort.</p>
</dd>
<dt class="hdlist1"><strong>-n, --no-seeds</strong></dt>
<dd>
<p>Don&#8217;t use heads or tails to seed topological sort.</p>
</dd>
<dt class="hdlist1"><strong>-c, --cycle-breaking</strong></dt>
<dd>
<p>Use a cycle breaking sort.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_random_sort">4.4.3. Random Sort</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-r, --random</strong></dt>
<dd>
<p>Randomly sort the graph.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_path_guided_1d_linear_sgd_sort">4.4.4. Path Guided 1D Linear SGD Sort</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-Y, --path-sgd</strong></dt>
<dd>
<p>Apply path guided 1D linear SGD algorithm to organize the graph.</p>
</dd>
<dt class="hdlist1"><strong>-X, --path-index</strong>=<em>FILE</em></dt>
<dd>
<p>Load the path index from this <em>FILE</em>.</p>
</dd>
<dt class="hdlist1"><strong>-f, --path-sgd-use-paths</strong>=FILE</dt>
<dd>
<p>Specify a line separated list of paths to sample from for the on the fly term generation process in the path guided linear 1D SGD. The default value are <em>all paths</em>.</p>
</dd>
<dt class="hdlist1"><strong>-G, --path-sgd-min-term-updates-paths</strong>=<em>N</em></dt>
<dd>
<p>The minimum number of terms to be updated before a new path guided linear 1D SGD iteration with adjusted learning rate eta starts, expressed as a multiple of total path steps. The default value is <em>1.0</em>. Can be overwritten by <em>-U, -path-sgd-min-term-updates-nodes=N</em>.</p>
</dd>
<dt class="hdlist1"><strong>-U, --path-sgd-min-term-updates-nodes</strong>=<em>N</em></dt>
<dd>
<p>The minimum number of terms to be updated before a new path guided linear 1D SGD iteration with adjusted learning rate eta starts, expressed as a multiple of the number of nodes. Per default, the argument is not set. The default of <em>-G, path-sgd-min-term-updates-paths=N</em> is used).</p>
</dd>
<dt class="hdlist1"><strong>-j, --path-sgd-delta</strong>=<em>N</em></dt>
<dd>
<p>The threshold of maximum displacement approximately in bp at which to stop path guided linear 1D SGD. Default values is <em>0.0</em>.</p>
</dd>
<dt class="hdlist1"><strong>-g, --path-sgd-eps</strong>=<em>N</em></dt>
<dd>
<p>The final learning rate for path guided linear 1D SGD model. The default value is <em>0.01</em>.</p>
</dd>
<dt class="hdlist1"><strong>-v, --path-sgd-eta-max</strong>=<em>N</em></dt>
<dd>
<p>The first and maximum learning rate for path guided linear 1D SGD model. The default value is <em>squared steps of longest path in graph</em>.</p>
</dd>
<dt class="hdlist1"><strong>-a, --path-sgd-zipf-theta</strong>=<em>N</em></dt>
<dd>
<p>The theta value for the Zipfian distribution which is used as the sampling method for the second node of one term in the path guided linear 1D SGD model. The default value is <em>0.99</em>.</p>
</dd>
<dt class="hdlist1"><strong>-x, --path-sgd-iter-max</strong>=<em>N</em></dt>
<dd>
<p>The maximum number of iterations for path guided linear 1D SGD model. The default value is <em>30</em>.</p>
</dd>
<dt class="hdlist1"><strong>-F, --iteration-max-learning-rate</strong>=<em>N</em></dt>
<dd>
<p>The iteration where the learning rate is max for path guided linear 1D SGD model. The default value is <em>0</em>.</p>
</dd>
<dt class="hdlist1"><strong>-k, --path-sgd-zipf-space</strong>=<em>N</em></dt>
<dd>
<p>The maximum space size of the Zipfian distribution which is used as the sampling method for the second node of one term in the path guided linear 1D SGD model. The default value is the <em>longest path length</em>.</p>
</dd>
<dt class="hdlist1"><strong>-I, --path-sgd-zipf-space-max</strong>=<em>N</em></dt>
<dd>
<p>The maximum space size of the Zipfian distribution beyond which quantization occurs. Default value is <em>100</em>.</p>
</dd>
<dt class="hdlist1"><strong>-l, --path-sgd-zipf-space-quantization-step</strong>=<em>N</em></dt>
<dd>
<p>Quantization step size when the maximum space size of the Zipfian distribution is exceeded. Default value is <em>100</em>.</p>
</dd>
<dt class="hdlist1"><strong>-y, --path-sgd-zipf-max-num-distributions</strong>=<em>N</em></dt>
<dd>
<p>Approximate maximum number of Zipfian distributions to calculate. The default value is <em>100</em>.</p>
</dd>
<dt class="hdlist1"><strong>-q, --path-sgd-seed</strong>=<em>N</em></dt>
<dd>
<p>Set the seed for the deterministic 1-threaded path guided linear 1D SGD model. The default value is <em>pangenomic!</em>.</p>
</dd>
<dt class="hdlist1"><strong>-u, --path-sgd-snapshot</strong>=<em>STRING</em></dt>
<dd>
<p>Set the prefix to which each snapshot graph of a path guided 1D SGD iteration should be written to. This is turned off per default.
This argument only works when <em>-Y, --path-sgd</em> was specified. Not applicable in a pipeline of sorts.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_path_sorting_options">4.4.5. Path Sorting Options</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-L, --paths-min</strong></dt>
<dd>
<p>Sort paths by their lowest contained node identifier.</p>
</dd>
<dt class="hdlist1"><strong>-M, --paths-max</strong></dt>
<dd>
<p>Sort paths by their highest contained node identifier.</p>
</dd>
<dt class="hdlist1"><strong>-A, --paths-avg</strong></dt>
<dd>
<p>Sort paths by their average contained node identifier.</p>
</dd>
<dt class="hdlist1"><strong>-R, --paths-avg-rev</strong></dt>
<dd>
<p>Sort paths in reverse by their average contained node identifier.</p>
</dd>
<dt class="hdlist1"><strong>-D, --path-delim</strong>=<em>path-delim</em></dt>
<dd>
<p>Sort paths in bins by their prefix up to this delimiter.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_pipeline_sorting">4.4.6. Pipeline Sorting</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-p, --pipeline</strong>=<em>STRING</em></dt>
<dd>
<p>Apply a series of sorts, based on single character command line arguments given to this command. The default sort is
<em>s</em>. The reverse sort would be specified via <em>f</em>.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_additional_parameters">4.4.7. Additional Parameters</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-d, --dagify-sort</strong></dt>
<dd>
<p>Sort on the basis of a DAGified graph.</p>
</dd>
<dt class="hdlist1"><strong>-O, --Optimize</strong></dt>
<dd>
<p>Use the MutableHandleGraph::optimize method to compact the node identifier space.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_threading_2">4.4.8. Threading</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-t, --threads</strong>=<em>N</em></dt>
<dd>
<p>Number of threads to use for the parallel operations.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_processing_information_2">4.4.9. Processing Information</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-P, --progress</strong></dt>
<dd>
<p>Print sort progress to stdout.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_program_information_3">4.4.10. Program Information</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-h, --help</strong></dt>
<dd>
<p>Print a help message for <strong>odgi sort</strong>.</p>
</dd>
</dl>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_exit_status_3">4.5. EXIT STATUS</h3>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>0</strong></dt>
<dd>
<p>Success.</p>
</dd>
<dt class="hdlist1"><strong>1</strong></dt>
<dd>
<p>Failure (syntax or usage error; parameter error; file processing failure; unexpected error).</p>
</dd>
</dl>
</div>
</div>
<div class="sect2">
<h3 id="_bugs_4">4.6. BUGS</h3>
<div class="paragraph">
<p>Refer to the <strong>odgi</strong> issue tracker at <a href="https://github.com/vgteam/odgi/issues" class="bare">https://github.com/vgteam/odgi/issues</a>.</p>
</div>
</div>
<div class="sect2">
<h3 id="_authors_4">4.7. AUTHORS</h3>
<div class="paragraph">
<p><strong>odgi sort</strong> was written by Erik Garrison.</p>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_odgi_view1">5. odgi view(1)</h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="_name_5">5.1. NAME</h3>
<div class="paragraph">
<p>odgi_view - projection of graphs into other formats</p>
</div>
</div>
<div class="sect2">
<h3 id="_synopsis_5">5.2. SYNOPSIS</h3>
<div class="paragraph">
<p><strong>odgi view</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;</p>
</div>
</div>
<div class="sect2">
<h3 id="_description_5">5.3. DESCRIPTION</h3>
<div class="paragraph">
<p>The odgi view(1) command can convert a graph in odgi format to GFAv1. It can reveal a graph&#8217;s internal structures for e.g. debugging processes.</p>
</div>
</div>
<div class="sect2">
<h3 id="_options_4">5.4. OPTIONS</h3>
<div class="sect3">
<h4 id="_graph_files_io_4">5.4.1. Graph Files IO</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-i, --idx</strong>=<em>FILE</em></dt>
<dd>
<p>File containing the succinct variation graph to convert from. The file name usually ends with <em>.og</em>.</p>
</dd>
<dt class="hdlist1"><strong>-g, --to-gfa</strong></dt>
<dd>
<p>Write the graph in GFAv1 format to standard output.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_summary_options_2">5.4.2. Summary Options</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-d, --display</strong></dt>
<dd>
<p>Show the internal structures of a graph. Print to stdout the maximum node identifier, the minimum node identifier,
the nodes vector, the delete nodes bit vector and the path metadata, each in a separate line.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_program_information_4">5.4.3. Program Information</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-h, --help</strong></dt>
<dd>
<p>Print a help message for <strong>odgi view</strong>.</p>
</dd>
</dl>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_exit_status_4">5.5. EXIT STATUS</h3>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>0</strong></dt>
<dd>
<p>Success.</p>
</dd>
<dt class="hdlist1"><strong>1</strong></dt>
<dd>
<p>Failure (syntax or usage error; parameter error; file processing failure; unexpected error).</p>
</dd>
</dl>
</div>
</div>
<div class="sect2">
<h3 id="_bugs_5">5.6. BUGS</h3>
<div class="paragraph">
<p>Refer to the <strong>odgi</strong> issue tracker at <a href="https://github.com/vgteam/odgi/issues" class="bare">https://github.com/vgteam/odgi/issues</a>.</p>
</div>
</div>
<div class="sect2">
<h3 id="_authors_5">5.7. AUTHORS</h3>
<div class="paragraph">
<p><strong>odgi view</strong> was written by Erik Garrison.</p>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_odgi_kmers1">6. odgi kmers(1)</h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="_name_6">6.1. NAME</h3>
<div class="paragraph">
<p>odgi_kmers - show and characterize the kmer space of the graph</p>
</div>
</div>
<div class="sect2">
<h3 id="_synopsis_6">6.2. SYNOPSIS</h3>
<div class="paragraph">
<p><strong>odgi kmers</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<strong>-c, --stdout</strong>] [<em>OPTION</em>]&#8230;&#8203;</p>
</div>
</div>
<div class="sect2">
<h3 id="_description_6">6.3. DESCRIPTION</h3>
<div class="paragraph">
<p>Given a kmer length, the odgi kmers(1) command can emit all kmers. The output can be refined by setting the maximum number
of furcations at edges or by not considering nodes above a given node degree limit.</p>
</div>
</div>
<div class="sect2">
<h3 id="_options_5">6.4. OPTIONS</h3>
<div class="sect3">
<h4 id="_graph_files_io_5">6.4.1. Graph Files IO</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-i, --idx</strong>=<em>FILE</em></dt>
<dd>
<p>File containing the succinct variation graph to convert from. The file name usually ends with <em>.og</em>.</p>
</dd>
<dt class="hdlist1"><strong>-c, --stdout</strong>=</dt>
<dd>
<p>Write the kmers to standard output. Kmers are line-separated.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_kmer_options">6.4.2. Kmer Options</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-k, --kmer-length</strong>=<em>N</em></dt>
<dd>
<p>The kmer length to generate kmers from.</p>
</dd>
<dt class="hdlist1"><strong>-e, --max-furcations</strong>=<em>N</em></dt>
<dd>
<p>Break at edges that would induce this many furcations when generating a kmer.</p>
</dd>
<dt class="hdlist1"><strong>-D, --max-degree</strong>=<em>N</em></dt>
<dd>
<p>Don&#8217;t take nodes into account that have a degree greater than <em>N</em>.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_threading_3">6.4.3. Threading</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-t, --threads</strong>=<em>N</em></dt>
<dd>
<p>Number of threads to use.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_program_information_5">6.4.4. Program Information</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-h, --help</strong></dt>
<dd>
<p>Print a help message for <strong>odgi kmers</strong>.</p>
</dd>
</dl>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_exit_status_5">6.5. EXIT STATUS</h3>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>0</strong></dt>
<dd>
<p>Success.</p>
</dd>
<dt class="hdlist1"><strong>1</strong></dt>
<dd>
<p>Failure (syntax or usage error; parameter error; file processing failure; unexpected error).</p>
</dd>
</dl>
</div>
</div>
<div class="sect2">
<h3 id="_bugs_6">6.6. BUGS</h3>
<div class="paragraph">
<p>Refer to the <strong>odgi</strong> issue tracker at <a href="https://github.com/vgteam/odgi/issues" class="bare">https://github.com/vgteam/odgi/issues</a>.</p>
</div>
</div>
<div class="sect2">
<h3 id="_authors_6">6.7. AUTHORS</h3>
<div class="paragraph">
<p><strong>odgi kmers</strong> was written by Erik Garrison.</p>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_odgi_unitig1">7. odgi unitig(1)</h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="_name_7">7.1. NAME</h3>
<div class="paragraph">
<p>odgi_unitig - output unitigs of the graph</p>
</div>
</div>
<div class="sect2">
<h3 id="_synopsis_7">7.2. SYNOPSIS</h3>
<div class="paragraph">
<p><strong>odgi unitig</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;</p>
</div>
</div>
<div class="sect2">
<h3 id="_description_7">7.3. DESCRIPTION</h3>
<div class="paragraph">
<p>The odgi unitig(1) command can print all <a href="https://github.com/mcveanlab/mccortex/wiki/unitig">unitigs</a> of a given odgi graph to standard output in FASTA format. Unitigs can also be emitted
in a fixed sequence quality FASTQ format. Various parameters can refine the unitigs to print.</p>
</div>
</div>
<div class="sect2">
<h3 id="_options_6">7.4. OPTIONS</h3>
<div class="sect3">
<h4 id="_graph_files_io_6">7.4.1. Graph Files IO</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-i, --idx</strong>=<em>FILE</em></dt>
<dd>
<p>File containing the succinct variation graph to convert from. The file name usually ends with <em>.og</em>.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_fastq_options">7.4.2. FASTQ Options</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-f, --fake-fastq</strong></dt>
<dd>
<p>Write the unitigs in FASTQ format to stdout with a fixed quality value of <em>I</em>.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_unitig_options">7.4.3. Unitig Options</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-t, --sample-to</strong>=<em>N</em></dt>
<dd>
<p>Continue unitigs with a random walk in the graph so that they have at least the given <em>N</em> length.</p>
</dd>
<dt class="hdlist1"><strong>-p, --sample-plus</strong>=<em>N</em></dt>
<dd>
<p>Continue unitigs with a random walk in the graph by <em>N</em> past their natural end.</p>
</dd>
<dt class="hdlist1"><strong>-l, --min-begin-node-length</strong>=<em>N</em></dt>
<dd>
<p>Only begin unitigs collection from nodes which have at least length <em>N</em>.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_program_information_6">7.4.4. Program Information</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-h, --help</strong></dt>
<dd>
<p>Print a help message for <strong>odgi unitig</strong>.</p>
</dd>
</dl>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_exit_status_6">7.5. EXIT STATUS</h3>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>0</strong></dt>
<dd>
<p>Success.</p>
</dd>
<dt class="hdlist1"><strong>1</strong></dt>
<dd>
<p>Failure (syntax or usage error; parameter error; file processing failure; unexpected error).</p>
</dd>
</dl>
</div>
</div>
<div class="sect2">
<h3 id="_bugs_7">7.6. BUGS</h3>
<div class="paragraph">
<p>Refer to the <strong>odgi</strong> issue tracker at <a href="https://github.com/vgteam/odgi/issues" class="bare">https://github.com/vgteam/odgi/issues</a>.</p>
</div>
</div>
<div class="sect2">
<h3 id="_authors_7">7.7. AUTHORS</h3>
<div class="paragraph">
<p><strong>odgi unitig</strong> was written by Erik Garrison.</p>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_odgi_viz1">8. odgi viz(1)</h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="_name_8">8.1. NAME</h3>
<div class="paragraph">
<p>odgi_viz - variation graph visualizations</p>
</div>
</div>
<div class="sect2">
<h3 id="_synopsis_8">8.2. SYNOPSIS</h3>
<div class="paragraph">
<p><strong>odgi viz</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<strong>-o, --out</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;</p>
</div>
</div>
<div class="sect2">
<h3 id="_description_8">8.3. DESCRIPTION</h3>
<div class="paragraph">
<p>The odgi viz(1) command can produce a linear, static visualization of an odgi variation graph. It can aggregate the pangenome into bins
and directly renders a raster image. The binning level can be specified in input or it is calculated from the target width of the PNG to emit.
Can be used to produce visualizations for gigabase scale pangenomes. For more information about the binning process,
please refer to <a href="#_odgi_bin1">odgi bin</a>.</p>
</div>
</div>
<div class="sect2">
<h3 id="_options_7">8.4. OPTIONS</h3>
<div class="sect3">
<h4 id="_graph_files_io_7">8.4.1. Graph Files IO</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-i, --idx</strong>=<em>FILE</em></dt>
<dd>
<p>File containing the succinct variation graph to convert from. The file name usually ends with <em>.og</em>.</p>
</dd>
<dt class="hdlist1"><strong>-o, --out</strong>=<em>FILE</em></dt>
<dd>
<p>Write the visualization in PNG format to this file.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_visualization_options">8.4.2. Visualization Options</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-x, --width</strong>=<em>N</em></dt>
<dd>
<p>Set the width in pixels of the output image.</p>
</dd>
<dt class="hdlist1"><strong>-y, --height</strong>=<em>N</em></dt>
<dd>
<p>Set the height in pixels of the output image.</p>
</dd>
<dt class="hdlist1"><strong>-P, --path-height</strong>=<em>N</em></dt>
<dd>
<p>The height in pixels for a path.</p>
</dd>
<dt class="hdlist1"><strong>-X, --path-x-padding</strong>=<em>N</em></dt>
<dd>
<p>The padding in pixels on the x-axis for a path.</p>
</dd>
<dt class="hdlist1"><strong>-R, --pack-paths</strong></dt>
<dd>
<p>Pack all paths rather than displaying a single path per row.</p>
</dd>
<dt class="hdlist1"><strong>-L, --link-path-pieces</strong>=<em>FLOAT</em></dt>
<dd>
<p>Show thin links of this relative width to connect path pieces.</p>
</dd>
<dt class="hdlist1"><strong>-A, --alignment-prefix</strong>=<em>STRING</em></dt>
<dd>
<p>Apply alignment related visual motifs to paths which have this name prefix. It affects the [<strong>-S, --show-strand</strong>] and
[<strong>-d, --change-darkness</strong>] options.</p>
</dd>
<dt class="hdlist1"><strong>-S, --show-strand</strong></dt>
<dd>
<p>Use red and blue coloring to display forward and reverse alignments. This parameter can be set in combination with
[<strong>-A, --alignment-prefix</strong>=<em>STRING</em>].</p>
</dd>
<dt class="hdlist1"><strong>-z, --color-by-mean-inversion-rate</strong></dt>
<dd>
<p>Change the color respect to the node strandness (black for forward, red for reverse); in binned mode (<strong>-b, --binned-mode</strong>),
change the color respect to the mean inversion rate of the path for each bin, from black (no inversions) to red (bin
mean inversion rate equals to 1).</p>
</dd>
<dt class="hdlist1"><strong>-s, --color-by-prefix</strong></dt>
<dd>
<p>Colors paths by their names looking at the prefix before the given character C.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_intervals_selection">8.4.3. Intervals selection</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-r, --pangenomic-range</strong></dt>
<dd>
<p>Nucleotide range to visualize: <code>STRING=[PATH:]start-end</code>. <code>*-end</code> for <code>[0,end]</code>; <code>start-*</code> for <code>[start,pangenome_length]</code>.
If no PATH is specified, the nucleotide positions refer to the pangenome&#8217;s sequence (i.e., the sequence obtained arranging all the graph&#8217;s node
from left to right).</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_paths_selection">8.4.4. Paths selection</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-p, --paths-to-display</strong></dt>
<dd>
<p>List of paths to display in the specified order; the file must contain one path name per line and a subset of all
paths can be specified.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_path_names_visualization_options">8.4.5. Path names visualization Options</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-H, --hide-path-names</strong></dt>
<dd>
<p>Hide the path names on the left of the generated image.</p>
</dd>
<dt class="hdlist1"><strong>-C, --color-path-names-background</strong></dt>
<dd>
<p>Color path names background with the same color as paths</p>
</dd>
<dt class="hdlist1"><strong>-c, --max-num-of-characters</strong></dt>
<dd>
<p>Maximum number of characters to display for each path name (max 128 characters). The default value is
<em>the length of the longest path name</em> (up to 32 characters).</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_binned_mode_options">8.4.6. Binned Mode Options</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-b, --binned-mode</strong></dt>
<dd>
<p>The variation graph is binned before its visualization. Each pixel in the output image will correspond to a bin.
For more information about the binning process, please refer to <a href="#_odgi_bin1">odgi bin</a>.</p>
</dd>
<dt class="hdlist1"><strong>-w, --bin-width</strong>=<em>N</em></dt>
<dd>
<p>The bin width specifies the size of each bin in the binned mode. If it is not specified, the bin width is calculated
from the width in pixels of the output image.</p>
</dd>
<dt class="hdlist1"><strong>-g, --no-gap-links</strong></dt>
<dd>
<p>We divide links into 2 classes:</p>
<div class="olist arabic">
<ol class="arabic">
<li>
<p>the links which help to follow complex variations. They need to be drawn, else one could not follow the sequence of a path.</p>
</li>
<li>
<p>the links helping to follow simple variations. These links are called <strong>gap-links</strong>. Such links solely connecting a
path from left to right may not be relevant to understand a path&#8217;s traversal through the bins. Therefore, when this option
is set, the gap-links are not drawn in binned mode.</p>
</li>
</ol>
</div>
</dd>
<dt class="hdlist1"><strong>-m, --color-by-mean-coverage</strong></dt>
<dd>
<p>Change the color respect to the mean coverage of the path for each bin, from black (no coverage) to blue (max bin mean
coverage in the entire graph).</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_gradient_mode_also_known_as_position_mode_options">8.4.7. Gradient Mode (also known as Position Mode) Options</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-d, --change-darkness</strong></dt>
<dd>
<p>Change the color darkness based on nucleotide position in the path. When it is used in binned mode, the mean inversion
rate of the bin node is considered to set the color gradient starting position: when this rate is greater than 0.5, the
bin is considered inverted, and the color gradient starts from the right-end of the bin. This parameter can be set in
combination with [<strong>-A, --alignment-prefix</strong>=<em>STRING</em>].</p>
</dd>
<dt class="hdlist1"><strong>-l, --longest-path</strong></dt>
<dd>
<p>Use the longest path length to change the color darkness.</p>
</dd>
<dt class="hdlist1"><strong>-u, --white-to-black</strong></dt>
<dd>
<p>Change the color darkness from white (for the first nucleotide position) to black (for the last nucleotide position).</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_program_information_7">8.4.8. Program Information</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-h, --help</strong></dt>
<dd>
<p>Print a help message for <strong>odgi viz</strong>.</p>
</dd>
</dl>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_exit_status_7">8.5. EXIT STATUS</h3>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>0</strong></dt>
<dd>
<p>Success.</p>
</dd>
<dt class="hdlist1"><strong>1</strong></dt>
<dd>
<p>Failure (syntax or usage error; parameter error; file processing failure; unexpected error).</p>
</dd>
</dl>
</div>
</div>
<div class="sect2">
<h3 id="_bugs_8">8.6. BUGS</h3>
<div class="literalblock">
<div class="content">
<pre>Refer to the *odgi* issue tracker at https://github.com/vgteam/odgi/issues.</pre>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_authors_8">8.7. AUTHORS</h3>
<div class="paragraph">
<p><strong>odgi viz</strong> was written by Erik Garrison.</p>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_odgi_paths1">9. odgi paths(1)</h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="_name_9">9.1. NAME</h3>
<div class="paragraph">
<p>odgi_paths - embedded path interrogation</p>
</div>
</div>
<div class="sect2">
<h3 id="_synopsis_9">9.2. SYNOPSIS</h3>
<div class="paragraph">
<p><strong>odgi paths</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;</p>
</div>
</div>
<div class="sect2">
<h3 id="_description_9">9.3. DESCRIPTION</h3>
<div class="paragraph">
<p>The odgi paths(1) command allows the investigation of paths of a given variation graph. It can calculate overlap statistics
of groupings of paths.</p>
</div>
</div>
<div class="sect2">
<h3 id="_options_8">9.4. OPTIONS</h3>
<div class="sect3">
<h4 id="_graph_files_io_8">9.4.1. Graph Files IO</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-i, --idx</strong>=<em>FILE</em></dt>
<dd>
<p>File containing the succinct variation graph to investigate the paths from. The file name usually ends with <em>.og</em>.</p>
</dd>
<dt class="hdlist1"><strong>-O, --overlaps</strong>=<em>FILE</em></dt>
<dd>
<p>Read in the path grouping file to generate the overlap statistics from. The file must be tab-delimited. The first column
lists a grouping and the second the path itself. Each line has one path entry. For each group the pairwise overlap statistics
for each pairing will be calculated and printed to stdout.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_investigation_options">9.4.2. Investigation Options</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-L, --list-paths</strong></dt>
<dd>
<p>Print the paths in the graph to stdout. Each path is printed in its own line.</p>
</dd>
<dt class="hdlist1"><strong>-H, --haplotypes</strong></dt>
<dd>
<p>Print to stdout the paths in an approximate binary haplotype matrix based on the graph&#8217;s sort order. The output is tab-delimited:
<strong>path.name</strong>, <strong>path.length</strong>, <strong>node.count</strong>, <strong>node.1</strong>, <strong>node.2</strong>, <strong>node.n</strong>. Each path entry is printed in its own line.</p>
</dd>
<dt class="hdlist1"><strong>-D, --delim</strong>=<em>CHAR</em></dt>
<dd>
<p>The part of each path name before this delimiter is a group identifier. This parameter should only be set in combination
with [<strong>-H, --haplotypes</strong>]. Prints an additional, first column <strong>group.name</strong> to stdout.</p>
</dd>
<dt class="hdlist1"><strong>-d, --distance</strong></dt>
<dd>
<p>Provides a sparse distance matrix for paths. If [<strong>-D, --delim</strong>] is set, it will be path groups distances.</p>
</dd>
<dt class="hdlist1"><strong>-f, --fasta</strong></dt>
<dd>
<p>Print paths in FASTA format to stdout.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_threading_4">9.4.3. Threading</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-t, --threads</strong>=<em>N</em></dt>
<dd>
<p>Number of threads to use.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_program_information_8">9.4.4. Program Information</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-h, --help</strong></dt>
<dd>
<p>Print a help message for <strong>odgi paths</strong>.</p>
</dd>
</dl>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_exit_status_8">9.5. EXIT STATUS</h3>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>0</strong></dt>
<dd>
<p>Success.</p>
</dd>
<dt class="hdlist1"><strong>1</strong></dt>
<dd>
<p>Failure (syntax or usage error; parameter error; file processing failure; unexpected error).</p>
</dd>
</dl>
</div>
</div>
<div class="sect2">
<h3 id="_bugs_9">9.6. BUGS</h3>
<div class="paragraph">
<p>Refer to the <strong>odgi</strong> issue tracker at <a href="https://github.com/vgteam/odgi/issues" class="bare">https://github.com/vgteam/odgi/issues</a>.</p>
</div>
</div>
<div class="sect2">
<h3 id="_authors_9">9.7. AUTHORS</h3>
<div class="paragraph">
<p><strong>odgi paths</strong> was written by Erik Garrison.</p>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_odgi_prune1">10. odgi prune(1)</h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="_name_10">10.1. NAME</h3>
<div class="paragraph">
<p>odgi_prune - remove complex parts of the graph</p>
</div>
</div>
<div class="sect2">
<h3 id="_synopsis_10">10.2. SYNOPSIS</h3>
<div class="paragraph">
<p><strong>odgi prune</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<strong>-o, --out</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;</p>
</div>
</div>
<div class="sect2">
<h3 id="_description_10">10.3. DESCRIPTION</h3>
<div class="paragraph">
<p>The odgi prune(1) command can remove complex parts of a graph. One can drop paths, nodes by a certain kind of edge coverage,
edges and graph tips. Specifying a kmer length and a maximum number of furcations, the graph can be broken at edges not
fitting into these conditions.</p>
</div>
</div>
<div class="sect2">
<h3 id="_options_9">10.4. OPTIONS</h3>
<div class="sect3">
<h4 id="_graph_files_io_9">10.4.1. Graph Files IO</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-i, --idx</strong>=<em>FILE</em></dt>
<dd>
<p>File containing the succinct variation graph to load in. The file name usually ends with <em>.og</em>.</p>
</dd>
<dt class="hdlist1"><strong>-o, --out</strong>=<em>FILE</em></dt>
<dd>
<p>Write the pruned graph to <em>FILE</em>. The file name should end with <em>.og</em>.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_kmer_options_2">10.4.2. Kmer Options</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-k, --kmer-length</strong>=<em>N</em></dt>
<dd>
<p>The length of the kmers to consider.</p>
</dd>
<dt class="hdlist1"><strong>-e, --max-furcations</strong>=<em>N</em></dt>
<dd>
<p>Break at edges that would induce <em>N</em> many furcations in a kmer.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_node_options">10.4.3. Node Options</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-d, --max-degree</strong>=<em>N</em></dt>
<dd>
<p>Remove nodes that have a higher node degree than <em>N</em>.</p>
</dd>
<dt class="hdlist1"><strong>-c, --min-coverage</strong>=<em>N</em></dt>
<dd>
<p>Remove nodese covered by fewer than <em>N</em> number of path steps.</p>
</dd>
<dt class="hdlist1"><strong>-C, --max-coverage</strong>=<em>N</em></dt>
<dd>
<p>Remove nodes covered by more than <em>N</em> number of path steps.</p>
</dd>
<dt class="hdlist1"><strong>-T, --cut-tips</strong>=<em>N</em></dt>
<dd>
<p>Remove nodes which are graph tips.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_edge_options">10.4.4. Edge Options</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-E, --edge-coverage</strong></dt>
<dd>
<p>Remove edges outside of the minimum and maximum coverage rather than nodes. Only set this argument in combination with
[<strong>-c, --min-coverage</strong>=<em>N</em>] and [<strong>-C, --max-coverage</strong>=<em>N</em>].</p>
</dd>
<dt class="hdlist1"><strong>-b, --best-edges</strong>=<em>N</em></dt>
<dd>
<p>Only keep the <em>N</em> most covered inbound and output edges of each node.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_path_options">10.4.5. Path Options</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-D, --drop-paths</strong></dt>
<dd>
<p>Remove the paths from the graph.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_threading_5">10.4.6. Threading</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-t, --threads</strong>=<em>N</em></dt>
<dd>
<p>Number of threads to use.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_program_information_9">10.4.7. Program Information</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-h, --help</strong></dt>
<dd>
<p>Print a help message for <strong>odgi prune</strong>.</p>
</dd>
</dl>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_exit_status_9">10.5. EXIT STATUS</h3>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>0</strong></dt>
<dd>
<p>Success.</p>
</dd>
<dt class="hdlist1"><strong>1</strong></dt>
<dd>
<p>Failure (syntax or usage error; parameter error; file processing failure; unexpected error).</p>
</dd>
</dl>
</div>
</div>
<div class="sect2">
<h3 id="_bugs_10">10.6. BUGS</h3>
<div class="paragraph">
<p>Refer to the <strong>odgi</strong> issue tracker at <a href="https://github.com/vgteam/odgi/issues" class="bare">https://github.com/vgteam/odgi/issues</a>.</p>
</div>
</div>
<div class="sect2">
<h3 id="_authors_10">10.7. AUTHORS</h3>
<div class="paragraph">
<p><strong>odgi prune</strong> was written by Erik Garrison.</p>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_odgi_unchop1">11. odgi unchop(1)</h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="_name_11">11.1. NAME</h3>
<div class="paragraph">
<p>odgi_unchop - merge unitigs into single nodes</p>
</div>
</div>
<div class="sect2">
<h3 id="_synopsis_11">11.2. SYNOPSIS</h3>
<div class="paragraph">
<p><strong>odgi unchop</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<strong>-o, --out</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;</p>
</div>
</div>
<div class="sect2">
<h3 id="_description_11">11.3. DESCRIPTION</h3>
<div class="paragraph">
<p>The odgi unchop(1) command merges each unitig into a single node preserving the node order.</p>
</div>
</div>
<div class="sect2">
<h3 id="_options_10">11.4. OPTIONS</h3>
<div class="sect3">
<h4 id="_graph_files_io_10">11.4.1. Graph Files IO</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-i, --idx</strong>=<em>FILE</em></dt>
<dd>
<p>File containing the succinct variation graph to unchop. The file name usually ends with <em>.og</em>.</p>
</dd>
<dt class="hdlist1"><strong>-o, --out</strong>=<em>FILE</em></dt>
<dd>
<p>Write the unchopped dynamic succinct variation graph to this file. A file ending with <em>.og</em> is recommended.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_processing_information_3">11.4.2. Processing Information</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-d, --debug</strong></dt>
<dd>
<p>Print information about the process to stderr.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_threading_6">11.4.3. Threading</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-t, --threads</strong>=<em>N</em></dt>
<dd>
<p>Number of threads to use for the parallel operations.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_program_information_10">11.4.4. Program Information</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-h, --help</strong></dt>
<dd>
<p>Print a help message for <strong>odgi unchop</strong>.</p>
</dd>
</dl>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_exit_status_10">11.5. EXIT STATUS</h3>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>0</strong></dt>
<dd>
<p>Success.</p>
</dd>
<dt class="hdlist1"><strong>1</strong></dt>
<dd>
<p>Failure (syntax or usage error; parameter error; file processing failure; unexpected error).</p>
</dd>
</dl>
</div>
</div>
<div class="sect2">
<h3 id="_bugs_11">11.6. BUGS</h3>
<div class="paragraph">
<p>Refer to the <strong>odgi</strong> issue tracker at <a href="https://github.com/vgteam/odgi/issues" class="bare">https://github.com/vgteam/odgi/issues</a>.</p>
</div>
</div>
<div class="sect2">
<h3 id="_authors_11">11.7. AUTHORS</h3>
<div class="paragraph">
<p><strong>odgi unchop</strong> was written by Erik Garrison.</p>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_odgi_normalize1">12. odgi normalize(1)</h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="_name_12">12.1. NAME</h3>
<div class="paragraph">
<p>odgi_normalize - compact unitigs and simplify redundant furcations</p>
</div>
</div>
<div class="sect2">
<h3 id="_synopsis_12">12.2. SYNOPSIS</h3>
<div class="paragraph">
<p><strong>odgi normalize</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<strong>-o, --out</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;</p>
</div>
</div>
<div class="sect2">
<h3 id="_description_12">12.3. DESCRIPTION</h3>
<div class="paragraph">
<p>The odgi normalize(1) command <a href="#_odgi_unchop1">unchops</a> a given variation graph and simplifies redundant furcations.</p>
</div>
</div>
<div class="sect2">
<h3 id="_options_11">12.4. OPTIONS</h3>
<div class="sect3">
<h4 id="_graph_files_io_11">12.4.1. Graph Files IO</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-i, --idx</strong>=<em>FILE</em></dt>
<dd>
<p>File containing the succinct variation graph to normalize. The file name usually ends with <em>.og</em>.</p>
</dd>
<dt class="hdlist1"><strong>-o, --out</strong>=<em>FILE</em></dt>
<dd>
<p>Write the normalized dynamic succinct variation graph to this file. A file ending with <em>.og</em> is recommended.</p>
</dd>
<dt class="hdlist1"><strong>-I, --max-iterations</strong>=<em>N</em></dt>
<dd>
<p>Iterate the normalization up to <em>N</em> many times. The default is <em>10</em>.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_program_debugging">12.4.2. Program Debugging</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-d, --debug</strong></dt>
<dd>
<p>Print information about the normalization process to stdout.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_program_information_11">12.4.3. Program Information</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-h, --help</strong></dt>
<dd>
<p>Print a help message for <strong>odgi normalize</strong>.</p>
</dd>
</dl>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_exit_status_11">12.5. EXIT STATUS</h3>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>0</strong></dt>
<dd>
<p>Success.</p>
</dd>
<dt class="hdlist1"><strong>1</strong></dt>
<dd>
<p>Failure (syntax or usage error; parameter error; file processing failure; unexpected error).</p>
</dd>
</dl>
</div>
</div>
<div class="sect2">
<h3 id="_bugs_12">12.6. BUGS</h3>
<div class="paragraph">
<p>Refer to the <strong>odgi</strong> issue tracker at <a href="https://github.com/vgteam/odgi/issues" class="bare">https://github.com/vgteam/odgi/issues</a>.</p>
</div>
</div>
<div class="sect2">
<h3 id="_authors_12">12.7. AUTHORS</h3>
<div class="paragraph">
<p><strong>odgi normalize</strong> was written by Erik Garrison.</p>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_odgi_subset1">13. odgi subset(1)</h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="_name_13">13.1. NAME</h3>
<div class="paragraph">
<p>odgi_subset - extract subsets of the graph as defined by query criteria</p>
</div>
</div>
<div class="sect2">
<h3 id="_synopsis_13">13.2. SYNOPSIS</h3>
<div class="paragraph">
<p><strong>odgi subset</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<strong>-o, --out</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;</p>
</div>
</div>
<div class="sect2">
<h3 id="_description_13">13.3. DESCRIPTION</h3>
<div class="paragraph">
<p>Extracting a node subset of a variation graph is the task of the odgi subset(1) command. Users can specify a node, a list of nodes
or a context of which to generate a subset from.</p>
</div>
</div>
<div class="sect2">
<h3 id="_options_12">13.4. OPTIONS</h3>
<div class="sect3">
<h4 id="_graph_files_io_12">13.4.1. Graph Files IO</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-i, --idx</strong>=<em>FILE</em></dt>
<dd>
<p>File containing the succinct variation graph to subset. The file name usually ends with <em>.og</em>.</p>
</dd>
<dt class="hdlist1"><strong>-o, --out</strong>=<em>FILE</em></dt>
<dd>
<p>Write the subset to this file.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_traversal_options">13.4.2. Traversal Options</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-l, --node-list</strong>=<em>FILE</em></dt>
<dd>
<p>A file with one node identifier per line. All nodes specified here will be extract as a subset.</p>
</dd>
<dt class="hdlist1"><strong>-n, --node</strong>=<em>N</em></dt>
<dd>
<p>Specify the node identifier from which our traversal should begin.</p>
</dd>
<dt class="hdlist1"><strong>-c, --context</strong>=<em>N</em></dt>
<dd>
<p>The number of steps away from which to begin our traversal.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_threading_7">13.4.3. Threading</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-t, --threads</strong>=<em>N</em></dt>
<dd>
<p>Number of threads to use.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_program_information_12">13.4.4. Program Information</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-h, --help</strong></dt>
<dd>
<p>Print a help message for <strong>odgi subset</strong>.</p>
</dd>
</dl>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_exit_status_12">13.5. EXIT STATUS</h3>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>0</strong></dt>
<dd>
<p>Success.</p>
</dd>
<dt class="hdlist1"><strong>1</strong></dt>
<dd>
<p>Failure (syntax or usage error; parameter error; file processing failure; unexpected error).</p>
</dd>
</dl>
</div>
</div>
<div class="sect2">
<h3 id="_bugs_13">13.6. BUGS</h3>
<div class="paragraph">
<p>Refer to the <strong>odgi</strong> issue tracker at <a href="https://github.com/vgteam/odgi/issues" class="bare">https://github.com/vgteam/odgi/issues</a>.</p>
</div>
</div>
<div class="sect2">
<h3 id="_authors_13">13.7. AUTHORS</h3>
<div class="paragraph">
<p><strong>odgi subset</strong> was written by Erik Garrison.</p>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_odgi_bin1">14. odgi bin(1)</h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="_name_14">14.1. NAME</h3>
<div class="paragraph">
<p>odgi_bin - binning of pangenome sequence and path information in the graph</p>
</div>
</div>
<div class="sect2">
<h3 id="_synopsis_14">14.2. SYNOPSIS</h3>
<div class="paragraph">
<p><strong>odgi bin</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;</p>
</div>
</div>
<div class="sect2">
<h3 id="_description_14">14.3. DESCRIPTION</h3>
<div class="paragraph">
<p>The odgi bin(1) command bins a given variation graph. The pangenome sequence, the one-time traversal of all nodes from smallest to
largest node identifier, can be summed up into bins of a specified size. For each bin, the path metainformation is summarized.
This enables a summarized view of gigabase scale graphs. Each step of a path is a bin and connected to its next bin via a link.
A link has a start bin identifier and an end bin identifier.<br>
The concept of odgi bin is also applied in odgi <a href="#_odgi_viz1">viz</a>.
A demonstration of how the odgi bin JSON output can be used for an interactive visualization is realized in the <a href="https://graph-genome.github.io/">Pantograph</a>
project. Per default, odgi bin writes the bins to stdout in a tab-delimited format: <strong>path.name</strong>, <strong>path.prefix</strong>, <strong>path.suffix</strong>,
<strong>bin</strong> (bin identifier), <strong>mean.cov</strong> (mean coverage of the path in this bin), <strong>mean.inv</strong> (mean inversion rate of this path in this bin),
<strong>mean.pos</strong> (mean nucleotide position of this path in this bin), <strong>first.nucl</strong> (first nucleotide position of this path in this bin),
<strong>last.nucl</strong> (last nucleotide position of this path in this bin). These nucleotide ranges might span positions that are not present in the bin. Example:
A range of 1-100 means that the first nucleotide has position 1 and the last has position 100, but nucleotide 45 could be located in
another bin. For an exact positional output, please specify [<strong>-j, --json</strong>].<br>
Running odgi bin in <a href="https://github.com/tpook92/HaploBlocker">HaploBlocker</a> mode, only arguments [<strong>-b, --haplo-blocker</strong>],
[<strong>-p[N], --haplo-blocker-min-paths[N]</strong>], and [<strong>-c[N], --haplo-blocker-min-coverage[N]</strong>] are required. A TSV is printed to stdout:
Each row corresponds to a node. Each column corresponds to a path. Each value is the coverage of a specific node of a specific path.</p>
</div>
</div>
<div class="sect2">
<h3 id="_options_13">14.4. OPTIONS</h3>
<div class="sect3">
<h4 id="_graph_files_io_13">14.4.1. Graph Files IO</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-i, --idx</strong>=<em>FILE</em></dt>
<dd>
<p>File containing the succinct variation graph to investigate the bin from. The file name usually ends with <em>.og</em>.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_fasta_options">14.4.2. FASTA Options</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-f, --fasta</strong>=<em>FILE</em></dt>
<dd>
<p>Write the pangenome sequence to <em>FILE</em> in FASTA format.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_bin_options">14.4.3. Bin Options</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-n, --number-bins</strong>=<em>N</em></dt>
<dd>
<p>The number of bins the pangenome sequence should be chopped up to.</p>
</dd>
<dt class="hdlist1"><strong>-w, --bin-width</strong>=<em>N</em></dt>
<dd>
<p>The bin width specifies the size of each bin.</p>
</dd>
<dt class="hdlist1"><strong>-D, --path-delim</strong>=<em>STRING</em></dt>
<dd>
<p>Annotate rows by prefix and suffix of this delimiter.</p>
</dd>
<dt class="hdlist1"><strong>-a, --aggregate-delim</strong></dt>
<dd>
<p>Aggregate on path prefix delimiter. Argument depends on [<strong>-D, --path-delim</strong>=<em>STRING</em>].</p>
</dd>
<dt class="hdlist1"><strong>-j, --json</strong></dt>
<dd>
<p>Print bins and links to stdout in pseudo JSON format. Each line is a valid JSON object, but the whole file is not a valid JSON!
First, each bin including its pangenome sequence is printed to stdout per line. Second, for each path in the graph, its
traversed bins including metainformation: <strong>bin</strong> (bin identifier), <strong>mean.cov</strong> (mean coverage of the path in this bin), <strong>mean.inv</strong> (mean inversion rate of this path in this bin),
<strong>mean.pos</strong> (mean nucleotide position of this path in this bin), and an array of ranges determining the nucleotide position
of the path in this bin. Switching first and last nucleotide in a range represents a complement reverse
orientation of that particular sequence.</p>
</dd>
<dt class="hdlist1"><strong>-s, --no-seqs</strong></dt>
<dd>
<p>If [<strong>-j, --json</strong>] is set, no nucleotide sequences will be printed to stdout in order to save disk space.</p>
</dd>
<dt class="hdlist1"><strong>-g, --no-gap-links</strong></dt>
<dd>
<p>We divide links into 2 classes:</p>
<div class="olist arabic">
<ol class="arabic">
<li>
<p>the links which help to follow complex variations. They need to be drawn, else one could not follow the sequence of a path.</p>
</li>
<li>
<p>the links helping to follow simple variations. These links are called <strong>gap-links</strong>. Such links solely connecting a
path from left to right may not be relevant to understand a path&#8217;s traversal through the bins. Therefore, when this option
is set, the gap-links are left out saving disk space</p>
</li>
</ol>
</div>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_haploblocker_options">14.4.4. HaploBlocker Options</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-b, --haplo-blocker</strong></dt>
<dd>
<p>Write a TSV to stdout formatted in a way ready for HaploBlocker: Each row corresponds to a node. Each column corresponds to a path.
Each value is the coverage of a specific node of a specific path.</p>
</dd>
<dt class="hdlist1"><strong>-p[N], --haplo-blocker-min-paths[N]</strong></dt>
<dd>
<p>Specify the minimum number of paths that need to be present in the bin to actually report that bin. The default value is 1.</p>
</dd>
<dt class="hdlist1"><strong>-c[N], --haplo-blocker-min-coverage[N]</strong></dt>
<dd>
<p>Specify the minimum coverage a path needs to have in a bin to actually report that bin. The default value is 1.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_program_information_13">14.4.5. Program Information</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-h, --help</strong></dt>
<dd>
<p>Print a help message for <strong>odgi bin</strong>.</p>
</dd>
<dt class="hdlist1"><strong>-P, --progress</strong></dt>
<dd>
<p>Write the current progress to stderr.</p>
</dd>
</dl>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_exit_status_13">14.5. EXIT STATUS</h3>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>0</strong></dt>
<dd>
<p>Success.</p>
</dd>
<dt class="hdlist1"><strong>1</strong></dt>
<dd>
<p>Failure (syntax or usage error; parameter error; file processing failure; unexpected error).</p>
</dd>
</dl>
</div>
</div>
<div class="sect2">
<h3 id="_bugs_14">14.6. BUGS</h3>
<div class="paragraph">
<p>Refer to the <strong>odgi</strong> issue tracker at <a href="https://github.com/vgteam/odgi/issues" class="bare">https://github.com/vgteam/odgi/issues</a>.</p>
</div>
</div>
<div class="sect2">
<h3 id="_authors_14">14.7. AUTHORS</h3>
<div class="paragraph">
<p><strong>odgi bin</strong> was written by Erik Garrison, Simon Heumos</p>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_odgi_matrix1">15. odgi matrix(1)</h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="_name_15">15.1. NAME</h3>
<div class="paragraph">
<p>odgi_matrix - write the graph topology in sparse matrix formats</p>
</div>
</div>
<div class="sect2">
<h3 id="_synopsis_15">15.2. SYNOPSIS</h3>
<div class="paragraph">
<p><strong>odgi matrix</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;</p>
</div>
</div>
<div class="sect2">
<h3 id="_description_15">15.3. DESCRIPTION</h3>
<div class="paragraph">
<p>The odgi matrix(1) command generates a sparse matrix format out of the graph topology of a given variation graph.</p>
</div>
</div>
<div class="sect2">
<h3 id="_options_14">15.4. OPTIONS</h3>
<div class="sect3">
<h4 id="_graph_files_io_14">15.4.1. Graph Files IO</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-i, --idx</strong>=<em>FILE</em></dt>
<dd>
<p>File containing the succinct variation graph to create the sparse matrix from. The file name usually ends with <em>.og</em>.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_matrix_options">15.4.2. Matrix Options</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-e, --edge-depth-weight</strong></dt>
<dd>
<p>Weigh edges by their path depth.</p>
</dd>
<dt class="hdlist1"><strong>-d, --delta-weight</strong></dt>
<dd>
<p>Weigh edges by their inverse id delta.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_program_information_14">15.4.3. Program Information</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-h, --help</strong></dt>
<dd>
<p>Print a help message for <strong>odgi matrix</strong>.</p>
</dd>
</dl>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_exit_status_14">15.5. EXIT STATUS</h3>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>0</strong></dt>
<dd>
<p>Success.</p>
</dd>
<dt class="hdlist1"><strong>1</strong></dt>
<dd>
<p>Failure (syntax or usage error; parameter error; file processing failure; unexpected error).</p>
</dd>
</dl>
</div>
</div>
<div class="sect2">
<h3 id="_bugs_15">15.6. BUGS</h3>
<div class="paragraph">
<p>Refer to the <strong>odgi</strong> issue tracker at <a href="https://github.com/vgteam/odgi/issues" class="bare">https://github.com/vgteam/odgi/issues</a>.</p>
</div>
</div>
<div class="sect2">
<h3 id="_authors_15">15.7. AUTHORS</h3>
<div class="paragraph">
<p><strong>odgi matrix</strong> was written by Erik Garrison.</p>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_odgi_chop1">16. odgi chop(1)</h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="_name_16">16.1. NAME</h3>
<div class="paragraph">
<p>odgi_chop - divide nodes into smaller pieces</p>
</div>
</div>
<div class="sect2">
<h3 id="_synopsis_16">16.2. SYNOPSIS</h3>
<div class="paragraph">
<p><strong>odgi chop</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<strong>-o, --out</strong>=<em>FILE</em>] [<strong>-c, --chop-to</strong>=<em>N</em>] [<em>OPTION</em>]&#8230;&#8203;</p>
</div>
</div>
<div class="sect2">
<h3 id="_description_16">16.3. DESCRIPTION</h3>
<div class="paragraph">
<p>The odgi chop(1) command chops long nodes into short ones while preserving the graph topology and node order.</p>
</div>
</div>
<div class="sect2">
<h3 id="_options_15">16.4. OPTIONS</h3>
<div class="sect3">
<h4 id="_graph_files_io_15">16.4.1. Graph Files IO</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-i, --idx</strong>=<em>FILE</em></dt>
<dd>
<p>File containing the succinct variation graph to chop. The file name usually ends with <em>.og</em>.</p>
</dd>
<dt class="hdlist1"><strong>-o, --out</strong>=<em>FILE</em></dt>
<dd>
<p>Write the choped succinct variation graph to <em>FILE</em>. The file name usually ends with <em>.og</em>.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_chop_options">16.4.2. Chop Options</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-c, --chop-to</strong>=<em>N</em></dt>
<dd>
<p>Divide nodes that longer than <em>N</em> into nodes no longer than <em>N</em> while maintaining graph topology.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_threading_8">16.4.3. Threading</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-t, --threads</strong>=<em>N</em></dt>
<dd>
<p>Number of threads to use for the parallel operations.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_processing_information_4">16.4.4. Processing Information</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-d, --debug</strong></dt>
<dd>
<p>Print information about the process to stderr.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_program_information_15">16.4.5. Program Information</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-h, --help</strong></dt>
<dd>
<p>Print a help message for <strong>odgi chop</strong>.</p>
</dd>
</dl>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_exit_status_15">16.5. EXIT STATUS</h3>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>0</strong></dt>
<dd>
<p>Success.</p>
</dd>
<dt class="hdlist1"><strong>1</strong></dt>
<dd>
<p>Failure (syntax or usage error; parameter error; file processing failure; unexpected error).</p>
</dd>
</dl>
</div>
</div>
<div class="sect2">
<h3 id="_bugs_16">16.6. BUGS</h3>
<div class="paragraph">
<p>Refer to the <strong>odgi</strong> issue tracker at <a href="https://github.com/vgteam/odgi/issues" class="bare">https://github.com/vgteam/odgi/issues</a>.</p>
</div>
</div>
<div class="sect2">
<h3 id="_authors_16">16.7. AUTHORS</h3>
<div class="paragraph">
<p><strong>odgi chop</strong> was written by Erik Garrison.</p>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_odgi_layout1">17. odgi layout(1)</h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="_name_17">17.1. NAME</h3>
<div class="paragraph">
<p>odgi_layout - use SGD to make 2D layouts of the graph</p>
</div>
</div>
<div class="sect2">
<h3 id="_synopsis_17">17.2. SYNOPSIS</h3>
<div class="paragraph">
<p><strong>odgi layout</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<strong>-o, --out</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;</p>
</div>
</div>
<div class="sect2">
<h3 id="_description_17">17.3. DESCRIPTION</h3>
<div class="paragraph">
<p>The odgi layout(1) command draws 2D layouts of the graph using stochastic gradient descent (SGD). The input graph must be sorted
and id-compacted. The algorithm itself is described in <a href="https://arxiv.org/abs/1710.04626">Graph Drawing by Stochastic Gradient Descent</a>.
The force-directed graph drawing algorithm minimizes the graph&#8217;s energy function or stress level.
It applies SGD to move a single pair of nodes at a time. The rendered graph is written in SVG format.</p>
</div>
</div>
<div class="sect2">
<h3 id="_options_16">17.4. OPTIONS</h3>
<div class="sect3">
<h4 id="_graph_files_io_16">17.4.1. Graph Files IO</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-i, --idx</strong>=<em>FILE</em></dt>
<dd>
<p>File containing the succinct variation graph to layout. The file name usually ends with <em>.og</em>.</p>
</dd>
<dt class="hdlist1"><strong>-o, --out</strong>=<em>FILE</em></dt>
<dd>
<p>Write the rendered layout in SVG format to <em>FILE</em>.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_sgd_options">17.4.2. SGD Options</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-m, --iter-max</strong>=<em>N</em></dt>
<dd>
<p>The maximum number of iterations to run the layout. Default is <em>30</em>.</p>
</dd>
<dt class="hdlist1"><strong>-p, --n-pivots</strong>=<em>N</em></dt>
<dd>
<p>The number of pivots for sparse layout. Default is <em>0</em> leading to a non-sparse layout.</p>
</dd>
<dt class="hdlist1"><strong>-e, --eps</strong>=<em>N</em></dt>
<dd>
<p>The learning rate for SGD layout. Default is <em>0.01</em>.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_svg_options">17.4.3. SVG Options</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-x, --x-padding</strong>=<em>N</em></dt>
<dd>
<p>The padding between the connected component layouts. Default is <em>10.0</em>.</p>
</dd>
<dt class="hdlist1"><strong>-R, --render-scale</strong>=<em>N</em></dt>
<dd>
<p>SVG scaling Default is <em>5.0</em>.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_processing_information_5">17.4.4. Processing Information</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-d, --debug</strong></dt>
<dd>
<p>Print information about the components to stdout.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_program_information_16">17.4.5. Program Information</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-h, --help</strong></dt>
<dd>
<p>Print a help message for <strong>odgi layout</strong>.</p>
</dd>
</dl>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_exit_status_16">17.5. EXIT STATUS</h3>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>0</strong></dt>
<dd>
<p>Success.</p>
</dd>
<dt class="hdlist1"><strong>1</strong></dt>
<dd>
<p>Failure (syntax or usage error; parameter error; file processing failure; unexpected error).</p>
</dd>
</dl>
</div>
</div>
<div class="sect2">
<h3 id="_bugs_17">17.6. BUGS</h3>
<div class="paragraph">
<p>Refer to the <strong>odgi</strong> issue tracker at <a href="https://github.com/vgteam/odgi/issues" class="bare">https://github.com/vgteam/odgi/issues</a>.</p>
</div>
</div>
<div class="sect2">
<h3 id="_authors_17">17.7. AUTHORS</h3>
<div class="paragraph">
<p><strong>odgi layout</strong> was written by Erik Garrison.</p>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_odgi_flatten1">18. odgi flatten(1)</h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="_name_18">18.1. NAME</h3>
<div class="paragraph">
<p>odgi_flatten - generate linearization of the graph</p>
</div>
</div>
<div class="sect2">
<h3 id="_synopsis_18">18.2. SYNOPSIS</h3>
<div class="paragraph">
<p><strong>odgi flatten</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;</p>
</div>
</div>
<div class="sect2">
<h3 id="_description_18">18.3. DESCRIPTION</h3>
<div class="paragraph">
<p>The odgi flatten(1) command projects the graph sequence and paths into FASTA and BED.</p>
</div>
</div>
<div class="sect2">
<h3 id="_options_17">18.4. OPTIONS</h3>
<div class="sect3">
<h4 id="_graph_files_io_17">18.4.1. Graph Files IO</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-i, --idx</strong>=<em>FILE</em></dt>
<dd>
<p>File containing the succinct variation graph to flatten. The file name usually ends with <em>.og</em>.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_output_options">18.4.2. Output Options</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-f, --fasta</strong>=<em>FILE</em></dt>
<dd>
<p>Write the concatenated node sequences in FASTA format to <em>FILE</em>.</p>
</dd>
<dt class="hdlist1"><strong>-n, --name-seq</strong>=<em>STRING</em></dt>
<dd>
<p>The name to use for the concatenated graph sequence. Default is the name of the input file which was specified via [<strong>-i, --idx</strong>=<em>FILE</em>].</p>
</dd>
<dt class="hdlist1"><strong>-b, --bed</strong>=<em>FILE</em></dt>
<dd>
<p>Write the mapping between graph paths and the linearized FASTA sequence in BED format to <em>FILE</em>.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_program_information_17">18.4.3. Program Information</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-h, --help</strong></dt>
<dd>
<p>Print a help message for <strong>odgi flatten</strong>.</p>
</dd>
</dl>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_exit_status_17">18.5. EXIT STATUS</h3>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>0</strong></dt>
<dd>
<p>Success.</p>
</dd>
<dt class="hdlist1"><strong>1</strong></dt>
<dd>
<p>Failure (syntax or usage error; parameter error; file processing failure; unexpected error).</p>
</dd>
</dl>
</div>
</div>
<div class="sect2">
<h3 id="_bugs_18">18.6. BUGS</h3>
<div class="paragraph">
<p>Refer to the <strong>odgi</strong> issue tracker at <a href="https://github.com/vgteam/odgi/issues" class="bare">https://github.com/vgteam/odgi/issues</a>.</p>
</div>
</div>
<div class="sect2">
<h3 id="_authors_18">18.7. AUTHORS</h3>
<div class="paragraph">
<p><strong>odgi flatten</strong> was written by Erik Garrison.</p>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_odgi_break1">19. odgi break(1)</h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="_name_19">19.1. NAME</h3>
<div class="paragraph">
<p>odgi_break - break cycles in the graph and drop its paths</p>
</div>
</div>
<div class="sect2">
<h3 id="_synopsis_19">19.2. SYNOPSIS</h3>
<div class="paragraph">
<p><strong>odgi break</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<strong>-o, --out</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;</p>
</div>
</div>
<div class="sect2">
<h3 id="_description_19">19.3. DESCRIPTION</h3>
<div class="paragraph">
<p>The odgi break(1) command finds cycles in a graph via <a href="https://en.wikipedia.org/wiki/Breadth-first_search">breadth-first search (BFS)</a> and breaks them, also dropping
the graph&#8217;s paths.</p>
</div>
</div>
<div class="sect2">
<h3 id="_options_18">19.4. OPTIONS</h3>
<div class="sect3">
<h4 id="_graph_files_io_18">19.4.1. Graph Files IO</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-i, --idx</strong>=<em>FILE</em></dt>
<dd>
<p>File containing the succinct variation graph to break. The file name usually ends with <em>.og</em>.</p>
</dd>
<dt class="hdlist1"><strong>-o, --out</strong>=<em>FILE</em></dt>
<dd>
<p>Write the broken graph to <em>FILE</em>.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_cycle_options">19.4.2. Cycle Options</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-c, --cycle-max-bp</strong>=<em>N</em></dt>
<dd>
<p>The maximum cycle length at which to break.</p>
</dd>
<dt class="hdlist1"><strong>-s, --max-search-bp</strong>=<em>N</em></dt>
<dd>
<p>The maximum search space of each BFS given in number of base pairs.</p>
</dd>
<dt class="hdlist1"><strong>-u, --repeat-up-to</strong>=<em>N</em></dt>
<dd>
<p>Iterate cycle breaking up to <em>N</em> times or stop if no new edges are removed.</p>
</dd>
<dt class="hdlist1"><strong>-d, --show</strong></dt>
<dd>
<p>Print the edges we would remove to stdout.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_program_information_18">19.4.3. Program Information</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-h, --help</strong></dt>
<dd>
<p>Print a help message for <strong>odgi break</strong>.</p>
</dd>
</dl>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_exit_status_18">19.5. EXIT STATUS</h3>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>0</strong></dt>
<dd>
<p>Success.</p>
</dd>
<dt class="hdlist1"><strong>1</strong></dt>
<dd>
<p>Failure (syntax or usage error; parameter error; file processing failure; unexpected error).</p>
</dd>
</dl>
</div>
</div>
<div class="sect2">
<h3 id="_bugs_19">19.6. BUGS</h3>
<div class="paragraph">
<p>Refer to the <strong>odgi</strong> issue tracker at <a href="https://github.com/vgteam/odgi/issues" class="bare">https://github.com/vgteam/odgi/issues</a>.</p>
</div>
</div>
<div class="sect2">
<h3 id="_authors_19">19.7. AUTHORS</h3>
<div class="paragraph">
<p><strong>odgi break</strong> was written by Erik Garrison.</p>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_odgi_pathindex1">20. odgi pathindex(1)</h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="_name_20">20.1. NAME</h3>
<div class="paragraph">
<p>odgi_pathindex - create a path index for a given path</p>
</div>
</div>
<div class="sect2">
<h3 id="_synopsis_20">20.2. SYNOPSIS</h3>
<div class="paragraph">
<p><strong>odgi pathindex</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<strong>-o, --out</strong>=<em>FILE</em>] [<em>OPTION</em>]&#8230;&#8203;</p>
</div>
</div>
<div class="sect2">
<h3 id="_description_20">20.3. DESCRIPTION</h3>
<div class="paragraph">
<p>The odgi pathindex(1) command generates a path index of a graph. It uses succinct data structures to encode the index.
The path index represents a subset of the features of a fully realized <a href="https://github.com/vgteam/xg">xg index</a>. Having a path index, we can use
odgi <a href="#_odgi_panpos1">panpos</a> to go from <strong>path:position</strong> &#8594; <strong>pangenome:position</strong> which is important when
navigating large graphs in an interactive manner like in the <a href="https://graph-genome.github.io/">Pantograph</a> project.</p>
</div>
</div>
<div class="sect2">
<h3 id="_options_19">20.4. OPTIONS</h3>
<div class="sect3">
<h4 id="_graph_files_io_19">20.4.1. Graph Files IO</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-i, --idx</strong>=<em>FILE</em></dt>
<dd>
<p>File containing the succinct variation graph to generate a path index from. The file name usually ends with <em>.og</em>.</p>
</dd>
<dt class="hdlist1"><strong>-o, --out</strong>=<em>FILE</em></dt>
<dd>
<p>Write the path index to <em>FILE</em>.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_program_information_19">20.4.2. Program Information</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-h, --help</strong></dt>
<dd>
<p>Print a help message for <strong>odgi pathindex</strong>.</p>
</dd>
</dl>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_exit_status_19">20.5. EXIT STATUS</h3>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>0</strong></dt>
<dd>
<p>Success.</p>
</dd>
<dt class="hdlist1"><strong>1</strong></dt>
<dd>
<p>Failure (syntax or usage error; parameter error; file processing failure; unexpected error).</p>
</dd>
</dl>
</div>
</div>
<div class="sect2">
<h3 id="_bugs_20">20.6. BUGS</h3>
<div class="paragraph">
<p>Refer to the <strong>odgi</strong> issue tracker at <a href="https://github.com/vgteam/odgi/issues" class="bare">https://github.com/vgteam/odgi/issues</a>.</p>
</div>
</div>
<div class="sect2">
<h3 id="_authors_20">20.7. AUTHORS</h3>
<div class="paragraph">
<p><strong>odgi pathindex</strong> was written by Simon Heumos.</p>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_odgi_panpos1">21. odgi panpos(1)</h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="_name_21">21.1. NAME</h3>
<div class="paragraph">
<p>odgi_panpos - get the pangenome position of a given path and nucleotide position (1-based)</p>
</div>
</div>
<div class="sect2">
<h3 id="_synopsis_21">21.2. SYNOPSIS</h3>
<div class="paragraph">
<p><strong>odgi panpos</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<strong>-p, --path</strong>=<em>STRING</em>] [<strong>-n, --nuc-pos</strong>=<em>N</em>] [<em>OPTION</em>]&#8230;&#8203;</p>
</div>
</div>
<div class="sect2">
<h3 id="_description_21">21.3. DESCRIPTION</h3>
<div class="paragraph">
<p>The odgi panpos(1) command give a pangenome position for a given path and nucleotide position. It requires a path index,
which can be created with odgi <a href="#_odgi_pathindex1">pathindex</a>. Going from <strong>path:position</strong> &#8594; <strong>pangenome:position</strong> is important when
navigating large graphs in an interactive manner like in the <a href="https://graph-genome.github.io/">Pantograph</a> project. All
input and output positions are 1-based.</p>
</div>
</div>
<div class="sect2">
<h3 id="_options_20">21.4. OPTIONS</h3>
<div class="sect3">
<h4 id="_graph_files_io_20">21.4.1. Graph Files IO</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-i, --idx</strong>=<em>FILE</em></dt>
<dd>
<p>File containing the succinct variation graph index to find the pangenome position in. The file name usually ends with <em>.xp</em>.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_position_options">21.4.2. Position Options</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-p, --path</strong>=<em>STRING</em></dt>
<dd>
<p>The path name of the query.</p>
</dd>
<dt class="hdlist1"><strong>-n, --nuc-pos</strong>=<em>STRING</em></dt>
<dd>
<p>The nucleotide sequence of the query.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_program_information_20">21.4.3. Program Information</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-h, --help</strong></dt>
<dd>
<p>Print a help message for <strong>odgi panpos</strong>.</p>
</dd>
</dl>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_exit_status_20">21.5. EXIT STATUS</h3>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>0</strong></dt>
<dd>
<p>Success.</p>
</dd>
<dt class="hdlist1"><strong>1</strong></dt>
<dd>
<p>Failure (syntax or usage error; parameter error; file processing failure; unexpected error).</p>
</dd>
</dl>
</div>
</div>
<div class="sect2">
<h3 id="_bugs_21">21.6. BUGS</h3>
<div class="paragraph">
<p>Refer to the <strong>odgi</strong> issue tracker at <a href="https://github.com/vgteam/odgi/issues" class="bare">https://github.com/vgteam/odgi/issues</a>.</p>
</div>
</div>
<div class="sect2">
<h3 id="_authors_21">21.7. AUTHORS</h3>
<div class="paragraph">
<p><strong>odgi panpos</strong> was written by Simon Heumos.</p>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_odgi_server1">22. odgi server(1)</h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="_name_22">22.1. NAME</h3>
<div class="paragraph">
<p>odgi_server - start a HTTP server with a given index file to query a pangenome position</p>
</div>
</div>
<div class="sect2">
<h3 id="_synopsis_22">22.2. SYNOPSIS</h3>
<div class="paragraph">
<p><strong>odgi server</strong> [<strong>-i, --idx</strong>=<em>FILE</em>] [<strong>-p, --port</strong>=<em>N</em>] [<em>OPTION</em>]&#8230;&#8203;</p>
</div>
</div>
<div class="sect2">
<h3 id="_description_22">22.3. DESCRIPTION</h3>
<div class="paragraph">
<p>The odgi server(1) command starts an HTTP server with a given path index as input. The idea is that we can go from
<strong>path:position</strong> &#8594; <strong>pangenome:position</strong> via GET requests to the HTTP server. The server headers do not block cross origin requests.
Example GET request: <strong><a href="http://localost:3000/path_name/nucleotide_position" class="bare">http://localost:3000/path_name/nucleotide_position</a></strong>.<br>
The required path index can be created with odgi <a href="#_odgi_pathindex1">pathindex</a>. Going from <strong>path:position</strong> &#8594; <strong>pangenome:position</strong> is important when
navigating large graphs in an interactive manner like in the <a href="https://graph-genome.github.io/">Pantograph</a> project. All
input and output positions are 1-based. If no IP address is specified, the server will run on localhost.</p>
</div>
</div>
<div class="sect2">
<h3 id="_options_21">22.4. OPTIONS</h3>
<div class="sect3">
<h4 id="_graph_files_io_21">22.4.1. Graph Files IO</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-i, --idx</strong>=<em>FILE</em></dt>
<dd>
<p>File containing the succinct variation graph index to host in a HTTP server. The file name usually ends with <em>.xp</em>.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_http_options">22.4.2. HTTP Options</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-p, --port</strong>=<em>N</em></dt>
<dd>
<p>Run the server under this port.</p>
</dd>
<dt class="hdlist1"><strong>-a, --ip</strong>=<em>IP</em></dt>
<dd>
<p>Run the server under this IP address. If not specified, <em>IP</em> will be <em>localhost</em>.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_program_information_21">22.4.3. Program Information</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-h, --help</strong></dt>
<dd>
<p>Print a help message for <strong>odgi server</strong>.</p>
</dd>
</dl>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_exit_status_21">22.5. EXIT STATUS</h3>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>0</strong></dt>
<dd>
<p>Success.</p>
</dd>
<dt class="hdlist1"><strong>1</strong></dt>
<dd>
<p>Failure (syntax or usage error; parameter error; file processing failure; unexpected error).</p>
</dd>
</dl>
</div>
</div>
<div class="sect2">
<h3 id="_bugs_22">22.6. BUGS</h3>
<div class="paragraph">
<p>Refer to the <strong>odgi</strong> issue tracker at <a href="https://github.com/vgteam/odgi/issues" class="bare">https://github.com/vgteam/odgi/issues</a>.</p>
</div>
</div>
<div class="sect2">
<h3 id="_authors_22">22.7. AUTHORS</h3>
<div class="paragraph">
<p><strong>odgi server</strong> was written by Simon Heumos.</p>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_odgi_test1">23. odgi test(1)</h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="_name_23">23.1. NAME</h3>
<div class="paragraph">
<p>odgi_test - run odgi unit tests</p>
</div>
</div>
<div class="sect2">
<h3 id="_synopsis_23">23.2. SYNOPSIS</h3>
<div class="paragraph">
<p><strong>odgi test</strong> [&lt;TEST NAME|PATTERN|TAGS&gt; &#8230;&#8203;] [<em>OPTION</em>]&#8230;&#8203;</p>
</div>
</div>
<div class="sect2">
<h3 id="_description_23">23.3. DESCRIPTION</h3>
<div class="paragraph">
<p>The odgi test(1) command starts all unit tests that are implemented in odgi. For targeted testing, a subset of tests can
be selected. odgi test(1) depends on <a href="https://github.com/catchorg/Catch2">Catch2</a>. In the default setting, all results are printed to stdout.</p>
</div>
</div>
<div class="sect2">
<h3 id="_options_22">23.4. OPTIONS</h3>
<div class="sect3">
<h4 id="_testing_options">23.4.1. Testing Options</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-l, --list-tests</strong></dt>
<dd>
<p>List all test cases. If a pattern was specified, all matching test cases are listed.</p>
</dd>
<dt class="hdlist1"><strong>-t, --list-tags</strong></dt>
<dd>
<p>List all tags. If a pattern was specified, all matching tags are listed.</p>
</dd>
<dt class="hdlist1"><strong>-s, --success</strong></dt>
<dd>
<p>Include successful tests in output.</p>
</dd>
<dt class="hdlist1"><strong>-b, --break</strong></dt>
<dd>
<p>Break into debugger mode upon failed test.</p>
</dd>
<dt class="hdlist1"><strong>-e, --nothrow</strong></dt>
<dd>
<p>Skip exception tests.</p>
</dd>
<dt class="hdlist1"><strong>-i, --invisibles</strong></dt>
<dd>
<p>Show invisibles like tabs or newlines.</p>
</dd>
<dt class="hdlist1"><strong>-o, --out</strong>=<em>FILE</em></dt>
<dd>
<p>Write all output to <em>FILE</em>.</p>
</dd>
<dt class="hdlist1"><strong>-r, --reporter</strong>=<em>STRING</em></dt>
<dd>
<p>Reporter to use. Default is console.</p>
</dd>
<dt class="hdlist1"><strong>-n, --name</strong>=<em>STRING</em></dt>
<dd>
<p>Suite name.</p>
</dd>
<dt class="hdlist1"><strong>-a, --abort</strong></dt>
<dd>
<p>Abort at first failure.</p>
</dd>
<dt class="hdlist1"><strong>-x, --abortx</strong>=<em>N</em></dt>
<dd>
<p>Abort after <em>N</em> failures.</p>
</dd>
<dt class="hdlist1"><strong>-w, --warn</strong>=<em>STRING</em></dt>
<dd>
<p>Enable warnings.</p>
</dd>
<dt class="hdlist1"><strong>-d, --durations</strong>=<em>yes|no</em></dt>
<dd>
<p>Show test durations. Default is <em>no</em>.</p>
</dd>
<dt class="hdlist1"><strong>-f, --input-file</strong>=<em>FILE</em></dt>
<dd>
<p>Load test names from a file.</p>
</dd>
<dt class="hdlist1"><strong>-#, --filenames-as-tags</strong></dt>
<dd>
<p>Adds a tag for the file name.</p>
</dd>
<dt class="hdlist1"><strong>-c, --section</strong>=<em>STRING</em></dt>
<dd>
<p>Specify the section to run the tests on.</p>
</dd>
<dt class="hdlist1"><strong>-v, --verosity</strong>=<em>quiet|normal|high</em></dt>
<dd>
<p>Set output verbosity. Default is <em>normal</em>.</p>
</dd>
<dt class="hdlist1"><strong>--list-test-names-only</strong></dt>
<dd>
<p>List all test cases names only. If a pattern was specified, all matching test cases are listed.</p>
</dd>
<dt class="hdlist1"><strong>--list-reporters</strong></dt>
<dd>
<p>List all reporters.</p>
</dd>
<dt class="hdlist1"><strong>--order</strong>=<em>decl|lex|rand</em></dt>
<dd>
<p>Test case order. Default ist <em>decl</em>.</p>
</dd>
<dt class="hdlist1"><strong>--rng-seed</strong>=<em>time|number</em></dt>
<dd>
<p>Set a specific seed for random numbers.</p>
</dd>
<dt class="hdlist1"><strong>--use-color</strong>=<em>yes|no</em></dt>
<dd>
<p>Should the output be colorized? Default is <em>yes</em>.</p>
</dd>
<dt class="hdlist1"><strong>--libidentify</strong></dt>
<dd>
<p>Report name and version according to libidentify.</p>
</dd>
<dt class="hdlist1"><strong>--wait-for-keypress</strong>=<em>start|exit|both</em></dt>
<dd>
<p>Waits for a keypress before <em>start|exit|both</em>.</p>
</dd>
<dt class="hdlist1"><strong>--benchmark-resolution-multiple</strong></dt>
<dd>
<p>Multiple of clock resolution to run benchmarks.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_program_information_22">23.4.2. Program Information</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-?, -h, --help</strong></dt>
<dd>
<p>Print a help message for <strong>odgi test</strong>.</p>
</dd>
</dl>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_exit_status_22">23.5. EXIT STATUS</h3>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>0</strong></dt>
<dd>
<p>Success.</p>
</dd>
<dt class="hdlist1"><strong>1</strong></dt>
<dd>
<p>Failure (syntax or usage error; parameter error; file processing failure; unexpected error).</p>
</dd>
</dl>
</div>
</div>
<div class="sect2">
<h3 id="_bugs_23">23.6. BUGS</h3>
<div class="paragraph">
<p>Refer to the <strong>odgi</strong> issue tracker at <a href="https://github.com/vgteam/odgi/issues" class="bare">https://github.com/vgteam/odgi/issues</a>.</p>
</div>
</div>
<div class="sect2">
<h3 id="_authors_23">23.7. AUTHORS</h3>
<div class="paragraph">
<p><strong>odgi test</strong> was written by Erik Garrison and Simon Heumos.</p>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_odgi_version1">24. odgi version(1)</h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="_name_24">24.1. NAME</h3>
<div class="paragraph">
<p>odgi_version - display the version of odgi</p>
</div>
</div>
<div class="sect2">
<h3 id="_synopsis_24">24.2. SYNOPSIS</h3>
<div class="paragraph">
<p><strong>odgi version</strong> [<em>OPTION</em>]&#8230;&#8203;</p>
</div>
</div>
<div class="sect2">
<h3 id="_description_24">24.3. DESCRIPTION</h3>
<div class="paragraph">
<p>The odgi version(1) command prints the current git version with tags and codename to stdout (like <em>v-44-g89d022b "back to old ABI"</em>). Optionally, only the release, version or codename can be printed.</p>
</div>
</div>
<div class="sect2">
<h3 id="_options_23">24.4. OPTIONS</h3>
<div class="sect3">
<h4 id="_version_options">24.4.1. Version Options</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-v, --version</strong>=</dt>
<dd>
<p>Print only the version (like <em>v-44-g89d022b</em>).</p>
</dd>
<dt class="hdlist1"><strong>-c, --codename</strong></dt>
<dd>
<p>Print only the codename (like <em>back to old ABI</em>).</p>
</dd>
<dt class="hdlist1"><strong>-r, --release</strong></dt>
<dd>
<p>Print only the release (like <em>v</em>).</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_program_information_23">24.4.2. Program Information</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-h, --help</strong></dt>
<dd>
<p>Print a help message for <strong>odgi version</strong>.</p>
</dd>
</dl>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_exit_status_23">24.5. EXIT STATUS</h3>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>0</strong></dt>
<dd>
<p>Success.</p>
</dd>
<dt class="hdlist1"><strong>1</strong></dt>
<dd>
<p>Failure (syntax or usage error; parameter error; file processing failure; unexpected error).</p>
</dd>
</dl>
</div>
</div>
<div class="sect2">
<h3 id="_bugs_24">24.6. BUGS</h3>
<div class="paragraph">
<p>Refer to the <strong>odgi</strong> issue tracker at <a href="https://github.com/vgteam/odgi/issues" class="bare">https://github.com/vgteam/odgi/issues</a>.</p>
</div>
</div>
<div class="sect2">
<h3 id="_authors_24">24.7. AUTHORS</h3>
<div class="paragraph">
<p><strong>odgi version</strong> was written by Simon Heumos.</p>
</div>
</div>
</div>
</div>
</div>
<div id="footer">
<div id="footer-text">
Last updated 2021-03-12 15:33:48 +0100
</div>
</div>
</body>
</html>