<!DOCTYPE html>
<html lang="en">
<title>Practical Graphical Pangenomics</title>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<link rel="stylesheet" href="style/w3.css">
<link rel="stylesheet" href="style/w3-theme-black.css">
<link rel="stylesheet" href="style/Roboto.css">
<link rel="stylesheet" href="style/font-awesome.min.css">
<style>
  html,
  body,
  h1,
  h2,
  h3,
  h4,
  h5,
  h6 {
    font-family: "Roboto", sans-serif;
  }

  h1, h3 {
    text-align: center;
  }

  .w3-sidebar {
    z-index: 3;
    width: 250px;
    top: 43px;
    bottom: 0;
    height: inherit;
  }
</style>

<body>

  <!-- Navbar -->
  <!--
    <div class="w3-top">
        <div class="w3-bar w3-theme w3-top w3-left-align w3-large">
            <a class="w3-bar-item w3-button w3-right w3-hide-large w3-hover-white w3-large w3-theme-l1"
                href="javascript:void(0)" onclick="w3_open()"><i class="fa fa-bars"></i></a>
            <a href="#" class="w3-bar-item w3-button w3-theme-l1">Logo</a>
            <a href="#" class="w3-bar-item w3-button w3-hide-small w3-hover-white">About</a>
            <a href="#" class="w3-bar-item w3-button w3-hide-small w3-hover-white">Values</a>
            <a href="#" class="w3-bar-item w3-button w3-hide-small w3-hover-white">News</a>
            <a href="#" class="w3-bar-item w3-button w3-hide-small w3-hover-white">Contact</a>
            <a href="#" class="w3-bar-item w3-button w3-hide-small w3-hide-medium w3-hover-white">Clients</a>
            <a href="#" class="w3-bar-item w3-button w3-hide-small w3-hide-medium w3-hover-white">Partners</a>
        </div>
    </div>
    -->

  <!-- Sidebar -->
  <!--
    <nav class="w3-sidebar w3-bar-block w3-collapse w3-large w3-theme-l5 w3-animate-left" id="mySidebar">
        <a href="javascript:void(0)" onclick="w3_close()"
            class="w3-right w3-xlarge w3-padding-large w3-hover-black w3-hide-large" title="Close Menu">
            <i class="fa fa-remove"></i>
        </a>
        <h4 class="w3-bar-item"><b>Menu</b></h4>
        <a class="w3-bar-item w3-button w3-hover-black" href="#">Link</a>
        <a class="w3-bar-item w3-button w3-hover-black" href="#">Link</a>
        <a class="w3-bar-item w3-button w3-hover-black" href="#">Link</a>
        <a class="w3-bar-item w3-button w3-hover-black" href="#">Link</a>
    </nav>
    -->

  <!-- Overlay effect when opening sidebar on small screens -->
  <!--
    <div class="w3-overlay w3-hide-large" onclick="w3_close()" style="cursor:pointer" title="close side menu"
        id="myOverlay"></div>
    -->

  <!-- Main content: shift it to the right by 250 pixels when the sidebar is visible -->


    <div class="w3-main">
      <!-- style="margin-left:250px">  -->
      <h1>Practical Graphical Pangenomics</h1>
      <h3>tools and workflows based on genome variation graphs</h3>
      <div class="w3-third w3-container">
        <a href="https://doi.org/10.1146/annurev-genom-120219-080406"><h2>Pangenomic methods</h2></a>
        <p>
          Standard approaches to genome inference and analysis relate sequences to a single linear reference genome.
          This is efficient but has a fundamental problem:
          Differences from this reference are hard to observe and describe in a coherent way.
          Variation and sequence are separated.
        </p>
        <a href="images/genomic-vs-pangenomic-analysis.png"><img src="images/genomic-vs-pangenomic-analysis.png" width="45%"></a>
        <a href="images/genomic-vs-pangenomic-models.png"><img src="images/genomic-vs-pangenomic-models.png" width="54%"></a>
        <p>
          Pangenomic methods allow us to relate all genomes or sequences in our analysis directly to each other.
          Sequence and variation are combined into a coherent data structure.
          This practice is still new, and research into ways to design, implement, and apply this model is ongoing.
          However, there is a growing consensus around best practices.
          Many methods work on an augmented sequence graph model and use a handful of common data formats for input and
          output.
        </p>
        <p>
          The <i>variation graph</i> data model describes the all-to-all alignment of many sequences (genomes or genes for
          instance) as walks through a graph whose nodes are labeled with DNA sequences:
        </p>
        <a href="images/HLA-DRB1-3123-all-viz-caption.png"><img src="images/HLA-DRB1-3123-odgi-viz.png" width="100%"></a>
        <p>
          Here, we document tools and workflows that operate on this graphical pangenomic data model.
          Our goal is to provide greater clarity for students and scientists working with this new paradigm for genomic
          research.
        </p>
      </div>

      <div class="w3-third w3-container">
        <a href="https://github.com/vgteam/vg"><h2>vg</h2></a>
        The variation graph toolkit <strong><a href="https://github.com/vgteam/vg">vg</a></strong> provides computational methods for creating and manipulating of genome variation graphs. It's pangenome representation of a set of genomes overcomes reference bias and improves read mapping.
        This is highlighted in the <a href="https://www.nature.com/articles/nbt.4227">Nature Biotechnology publication</a>.
        Users can receive support on <a href="https://www.biostars.org/t/vg/"><strong>vg</strong>'s Biostars page</a>.
      </div>

      <div class="w3-third w3-container">
        <a href="https://github.com/pangenome/pggb"><h2>PanGenome Graph Builder (pggb)</h2></a>
        This pangenome graph construction pipeline renders a collection of sequences into a pangenome graph (in the variation graph model). Its goal is to build a graph that is locally directed and acyclic while preserving large-scale variation. Maintaining local linearity is important for the interpretation, visualization, and reuse of pangenome variation graphs.
      </div>

      <div class="w3-third w3-container">
        <a href="https://github.com/pangenome/pgvf-spec"><h2>Pangenome Graph Variation Format (PGVF)</h2></a>
        PGVF is a hard fork of the GFAv1 format that allows the description of graph-to-graph alignments. It represents a collection of aligned graphs as a network of walks through an underlying merged sequence graph. While pangenome graphs let us represent differences between genomes, we have no mechanism to represent differences between pangenome graphs, or to combine multiple pangenome graphs into one structure without losing information. This motivates the development of a new biological data format.
      </div>

      <div class="w3-third w3-container">
        <a href="https://github.com/vgteam/xg"><h2>xg</h2></a>
        The succinct graph index <strong><a href="https://github.com/vgteam/xg">xg</a></strong> presents a static index of nodes, edges and paths of a variation graph.
        <strong>xg</strong> can be used to annotate graph nodes with their reference path relative positions.
        It was a key component of early development in <strong>vg</strong>, and was use to scale short read mapping to large genomes.
        It implements the <a href="https://github.com/vgteam/libhandlegraph">libhandlegraph</a> API.
      </div>

      <div class="w3-third w3-container">
        <a href="odgi/index.html"><h2>odgi</h2></a>
        <a href="odgi/index.html"><strong>odgi</strong></a>, the Optimized Dynamic (genome) Graph Interface, links a thrifty dynamic in-memory variation graph data model to a set of algorithms designed for scalable sorting, pruning, transformation, and visualization of very large genome graphs.
        <strong>odgi</strong> includes <a href="odgi/odgipy.html">python bindings</a> that can be used to
        <a href="https://odgi.readthedocs.io/en/latest/rst/tutorial.html">directly interface with its data model</a>.
        The <a href="odgi/odgi_docs.html">odgi manual</a> provides detailed information about its features and subcommands, including examples.
      </div>

      <div class="w3-third w3-container">
        <a href="https://github.com/jltsiren/gbwt"><h2>GWBT</h2></a>
        <strong><a href="https://github.com/jltsiren/gbwt">GWBT</a> -</strong> <strong>Graph BWT</strong>
        is a substring index for paths in a <a href="#vg">variation graph</a>.
        It is based on the positional Burrows-Wheeler transform (PBWT) and independently implements its graph extension (gPBWT).
        The <strong>GBWT</strong> supports extreme compression of genome sequences, requiring only 1 bit per 1 kilobasepair of sequence to store a  1000 Genomes Project.
        For documentation see the <a href="https://github.com/jltsiren/gbwt/wiki"><strong>GBWT</strong> wiki</a>.
      </div>

      <div class="w3-third w3-container">
        <a href="https://github.com/pangenome/spodgi"><h2>spodgi</h2></a>
        <strong><a href="https://github.com/pangenome/spodgi">SpOdgi</a></strong> transforms any 
        <a href="#odgi">odgi</a> genome variation graph file into a SPARQL capable database. 
        The RDF semantics are described in the <a href="https://github.com/vgteam/vg/tree/master/ontology">vg ontology directory</a>.
        This transformation allows us to connect variation graphs to other RDF resources, supporting their query using logic programming.
        Many operations or queries that are implemented in custom code in other pangenome tools can be expressed in compact SPARQL queries executed against <strong>SpOdgi</strong>.
      </div>

      <div class="w3-third w3-container">
        <a href="handlegraph/index.html"><h2>HandleGraph</h2></a>
        Lessons learned when designing algorithms for variation graphs guided the development of the <a href="handlegraph/index.html"><strong>HandleGraph</strong> model</a> and API.
        It formalizes ways of addressing, traversing, and manipulating the fundamental units of variation graphs.
        Implemented in <a href="https://github.com/vgteam/libhandlegraph"><strong>libhandlegraph</strong></a>, this C++ API hierarchy defines a set of capabilities from the simplest: <a href="https://github.com/vgteam/libhandlegraph/blob/master/src/include/handlegraph/handle_graph.hpp">a static bidirected DNA sequence graph</a>, to the most complex: <a href="https://github.com/vgteam/libhandlegraph/blob/master/src/include/handlegraph/mutable_path_deletable_handle_graph.hpp">dynamic graphs with modifiable paths</a>, and covers a number of important capabilities including serialization and positional indexing.
      </div>

      <div class="w3-third w3-container">
        <a href="https://github.com/vgteam/libbdsg"><h2>libbdsg</h2></a>
        <a href="https://github.com/vgteam/libbdsg"><strong>libbdsg</strong></a> brings together a collection of dynamic <a href="https://github.com/vgteam/libhandlegraph"><strong>HandleGraph</strong></a> implementations.
        <strong><a href="https://github.com/vgteam/libbdsg/blob/master/src/packed_graph.cpp">PackedGraph</a></strong> is designed to have a very low memory footprint.
        <strong><a href="https://github.com/vgteam/libbdsg/blob/master/src/hash_graph.cpp">HashGraph</a></strong> is implemented using a collection high-performance hash tables with the goal of providing the highest-possible runtime performance at the cost of increased memory usage.
        <!-- The library also wraps a copy of the <strong><a href="https://github.com/vgteam/libbdsg/blob/master/src/odgi.cpp">odgi</a></strong> graph model. -->
        For more details see the <a href="https://github.com/vgteam/handlegraph-paper">handle graph API comparison paper</a>. The <a href="https://bdsg.readthedocs.io/en/master/index.html">bdsg Read the Docs!</a> provides detailed information about starting a project with <strong>libbdsg</strong>, its python interface, tutorials, and an overview of available methods.
      </div>

      <div class="w3-third w3-container">
        <a href="https://github.com/ekg/seqwish"><h2>seqwish</h2></a>
        The alignment to variation graph inducer <a href="https://github.com/ekg/seqwish"><strong>seqwish</strong></a> renders a set of sequences and alignments into the equivalent variation graph.
        It accomplishes this using a number of tricks to reduce its memory footprint while maintaining a high degree of parallelism.
        The result is entirely dependent on the input alignments, which it represents losslessly.
        <strong>seqwish</strong> is generic: it can induce variation graphs from a collection of human genomes, or a set of noisy nanopore reads.
      </div>

      <div class="w3-third w3-container">
        <a href="https://github.com/maickrau/GraphAligner"><h2>GraphAligner</h2></a>
        The sequence to graph aligner <strong>GraphAligner</strong> implements a novel, <a href="https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btz162/5372677">high performance alignment algorithm</a> capable of aligning to graphs of arbitrary topological complexity with minimal overhead relative to a linear aligner.
        Its seeding strategy, which is based on exact matches (minimizers) in whole nodes, limits it to longer reads.
        It produces GAM and GAF alignment formats compatible with other pangenome graph based tools.
      </div>

      <div class="w3-third w3-container">
        <a href="https://github.com/GFA-spec/GFA-spec"><h2>Pangenomic data formats</h2></a>
        Graphical pangenomes are usually exchanged using a subset of <strong><a href="https://github.com/GFA-spec/GFA-spec/blob/master/GFA1.md">GFAv1</a> - Graphical Fragment Assembly</strong> format.
        Graph nodes are stored in sequence records (S), edges represented in link (L) records, and embedded sequences in path records (P).
        Mappings to <strong>GFA</strong> can be encoded in <strong><a href="https://github.com/vgteam/vg/wiki/File-Formats#gam-graph-alignment--map-vgs-bam">GAM</a></strong> (Graph Alignment/Map format, <a href="https://github.com/vgteam/vg">vg</a>'s BAM equivalent) or the text-based <strong><a href="https://github.com/lh3/gfatools/blob/master/doc/rGFA.md#the-graph-alignment-format-gaf">GAF</a></strong> (Graph Alignment Format).
      </div>

      <!--
      <div class="w3-third w3-container">
        <a href="https://github.com/lh3/minigraph"><h2>minigraph</h2></a>
        A <i>proof-of-concept</i> sequence-to-graph mapper and graph generator is <strong><a href="https://github.com/lh3/minigraph">minigraph</a></strong>.
        The pangenome graph is build in incremental <a href="https://github.com/lh3/minimap2">minimap</a> steps. Approximate mappings are found through chaining of local hits.
        Current <a href="https://github.com/lh3/minigraph#limitations">limitations</a> do not allow the algorithm to map short segments or to perform on a base-level resultion.
        This compromise leads to the construction of an accurate pangenome graph, but may leave out complex events.
      </div>
      -->

      <div class="w3-third w3-container">
        <a href="https://github.com/vgteam/sequenceTubeMap"><h2>Sequence Tube Map</h2></a>
        <strong><a href="https://github.com/vgteam/sequenceTubeMap">Sequence Tube Map</a></strong> is a javascript module visualizing variation graphs in a <i>tube-map-like</i> layout.
        It renders variation graphs using a <a href="images/sequence-tube-map-example.png">"tube map" model</a> in which paths representing genomes flow through the sequence nodes of the graph.
        Currently, it can only handle graphs created with <a href="https://github.com/vgteam/vg">vg</a>.
      </div>

      <div class="w3-third w3-container">
        <a href="https://graph-genome.github.io/"><h2>Pantograph</h2></a>
        The <strong><a href="https://graph-genome.github.io/">Pantograph</a></strong> project 
        aims to build an interactive pangenome visualization tool for COVID-19 data that includes annotation and metadata. 
        On the long run, it should be capable to visualize a pangenome of 1,000s of individuals and gigabase genomes,
        scaling from nucleotide to whole chromosome level.
      </div>

      <div class="w3-third w3-container">
        <a href="https://rrwick.github.io/Bandage"><h2>Bandage</h2></a>
        Originally developed for assembly graph visualization, <strong><a href="https://rrwick.github.io/Bandage">Bandage</a></strong> is an indispensable tool for visual inspection of variation graphs as well as assembly graphs.
      </div>

      <div class="w3-third w3-container">
        <a href="https://github.com/MoMI-G/MoMI-G/"><h2>MoMI-G</h2></a>
        <strong><a href="https://github.com/MoMI-G/MoMI-G/">MoMI-G</a> - MOdular Multi-scale Integrated Genome graph browser</strong> is a <i>mult-view</i>
        graph browser combining the base-level differences of <a href="https://github.com/vgteam/sequenceTubeMap">Sequence Tube Map</a>
        with a <a href="http://circos.ca/">CIRCOS</a> plot of chromosomal-scale connections and an interval card deck to efficiently browse structural variants.
        It displays evidences such as short and long read alignments, read depth, and annotations.
      </div>

      <!--
          <div class="w3-third w3-container">
            <p class="w3-border w3-padding-large w3-padding-32 w3-center">AD</p>
            <p class="w3-border w3-padding-large w3-padding-64 w3-center">AD</p>
          </div>
          -->


          <!-- END MAIN -->
    </div>

</body>

</html>
