

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Working with the HandleGraph abstraction &mdash; odgi 00d68c2 documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> odgi
          

          
          </a>

          
            
            
              <div class="version">
                v0.5.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">Welcome</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="quick_start.html">Quick Start</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="commands.html">Commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="binding.html">Python Binding</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">odgi</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Working with the HandleGraph abstraction</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../_sources/rst/handlegraph.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="working-with-the-handlegraph-abstraction">
<h1>Working with the HandleGraph abstraction<a class="headerlink" href="#working-with-the-handlegraph-abstraction" title="Permalink to this headline">¶</a></h1>
<p><em>Variation graphs</em> combine sequences (usually DNA) with information
about variation between them. They can be used to represent the mutual
alignment of multiple genomes, or more simply, a single reference and
variation found in a population. In these graphs, nodes are labeled with
sequences. Edges represent allowed linkages between nodes. Paths, or
walks traversing a series of nodes, represent sequences of interest.</p>
<div class="section" id="basic-handlegraph-types">
<h2>Basic HandleGraph types<a class="headerlink" href="#basic-handlegraph-types" title="Permalink to this headline">¶</a></h2>
<p>To interact with these graphs in precise, programmatic ways, we need to
be able to identify their basic elements. This is the core motivation
behind the <a class="reference external" href="https://github.com/vgteam/libhandlegraph">HandleGraph</a>
abstraction used in odgi and other dynamic graph implementations in
<a class="reference external" href="https://github.com/vgteam/libbdsg">libbdsg</a>. Because DNA sequence
graphs have two strands, we need a more precise way of addressing
elements in the graph than nodes, which implicitly represent both
strands. The <em>handle</em>, which is the core concept in the Handle Graph
abstraction, allows us to refer to one strand of a single node, which is
the smallest addressable unit in a variation graph. Nodes have numeric
identifiers (or _ID_s), and associated sequences. Edges link two
handles. Paths are a series of steps which link a path identifier and a
handle. Breaking this down (with reference to the C++ types), we have a
set of references to graph elements:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">handle_t</span></code>: an oriented traversal of a node (an opaque 64-bit
identifier)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">id_t</span></code>: a node identifier (a 64-bit number)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">edge_t</span></code>: a pair of <code class="docutils literal notranslate"><span class="pre">handle_t</span></code>s, where the edge is directed
from the first member of the pair to the second</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">path_handle_t</span></code>: a reference to a path (an opaque 64-bit
identifier)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">step_handle_t</span></code>: a reference to a single step of a path on one node
traversal (an opaque 128-bit identifier)</p></li>
</ul>
<p>The handle types are “opaque” in that they don’t directly correspond to
an external identifier. For instance, the <code class="docutils literal notranslate"><span class="pre">path_handle_t</span></code> is just a
64-bit type, not a path name that would have been used in the
<a class="reference external" href="https://github.com/GFA-spec/GFA-spec">GFA</a> serialization that we
loaded the graph from. The abstraction provides no guarantees about the
order or meaning of the identifiers, but they are guaranteed to be
stable as long as we don’t modify the graph.</p>
</div>
<div class="section" id="design-considerations-that-lead-to-the-handlegraph-abstraction">
<h2>Design considerations that lead to the HandleGraph abstraction<a class="headerlink" href="#design-considerations-that-lead-to-the-handlegraph-abstraction" title="Permalink to this headline">¶</a></h2>
<p>You might wonder why we use these opaque references to these elements,
rather than more literal identifiers. For instance, we could refer to
paths by their names, and steps by path names and indexes in the path.
We could just refer to nodes by their IDs.</p>
<p>The basic problem is that the representation of large, dynamic sequence
graphs requires careful data structure design that often makes accessing
elements by their literal names expensive. Except for very small or
simple graphs, we can’t expect to be able to save a copy of each path
name for every node it traverses. Instead, we encode some identifier
that refers to the path. By using a generic identifier, we are able to
encode the reference in a succinct, packed, or even implict manner that
can be memory efficient.</p>
<p>Graph implementations that provide the HandleGraph API can thus have
internal identifiers for graph elements that provide good performance
for functions that interact with, query, or traverse the graph. Defining
an abstract handle type allows each implementation to choose the right
internal identifier model to use. In the HandleGraph API, functions are
provided to convert between these internal identifiers and external
ones. For instance, given a <code class="docutils literal notranslate"><span class="pre">handle_t</span></code>, you can ask the graph to tell
you what node <code class="docutils literal notranslate"><span class="pre">id_t</span></code> and orientation (forward or reverse complement)
it represents. We can ask the graph to give us the <code class="docutils literal notranslate"><span class="pre">path_handle_t</span></code>
corresponding to a given path name, or find which <code class="docutils literal notranslate"><span class="pre">path_handle_t</span></code> and
<code class="docutils literal notranslate"><span class="pre">handle_t</span></code> a given <code class="docutils literal notranslate"><span class="pre">step_handle_t</span></code> references.</p>
</div>
<div class="section" id="why-should-we-use-a-handlegraph-implementation">
<h2>Why should we use a HandleGraph implementation?<a class="headerlink" href="#why-should-we-use-a-handlegraph-implementation" title="Permalink to this headline">¶</a></h2>
<p>Naive implementations of graph genomic data structures can consume tens
of bytes per input graph. So a GFA file of 30GB, a typical size for a
whole human genome graph including the 1000 Genomes Projects variation,
might be impossible to load on typical systems available (as of 2020).
(The initial VG implementation required 300GB to load the 30GB human
genome variation graph.) HandleGraph implementations developed by the
vgteam resolve this issue through careful use of succinct data
structures. For most of these implementations, including odgi, the
in-memory size of the graph data structures is around the same as the
uncompressed input GFA. At the same time, these models provide efficient
random access to (and in most cases, modification of) all of the graph
elements.</p>
<p>Our goal with the HandleGraph API is to provide a consistent and
reasoned interface to these graphs that allows for the development of
algorithms independent of the particular internals of a given graph data
structure. This frees users (bioinformatics and genomics researchers and
developers) to focus on algorithms that operate at a higher level,
without worrying about the details of efficient graph data structure
implementations. Our hope is that, by working with this consistent
interface, we can develop libraries of standard algorithms that achieve
common goals in graph genomic operations, thus reducing the effort
required to work with genome graphs.</p>
</div>
<div class="section" id="the-handlegraph-api-hierarchy">
<h2>The HandleGraph API hierarchy<a class="headerlink" href="#the-handlegraph-api-hierarchy" title="Permalink to this headline">¶</a></h2>
<p>The libhandlegraph APIs describe a hierarchy of increasingly complex
capabilities, from <a class="reference external" href="https://github.com/vgteam/libhandlegraph/blob/master/src/include/handlegraph/handle_graph.hpp">simple static variation graphs without
paths</a>,
to <a class="reference external" href="https://github.com/vgteam/libhandlegraph/blob/master/src/include/handlegraph/mutable_handle_graph.hpp">mutable
graphs</a>
with
<a class="reference external" href="https://github.com/vgteam/libhandlegraph/blob/master/src/include/handlegraph/mutable_path_handle_graph.hpp">paths</a>
and <a class="reference external" href="https://github.com/vgteam/libhandlegraph/blob/master/src/include/handlegraph/path_position_handle_graph.hpp">positional
indexes</a>.</p>
<p>This hierarchy allows us to build a consistent interface to
implementations that may be optimial in a particular application but
which lack generic functionality. For example, a static graph which
cannot be modified is usually more efficient in terms of memory and
access time than a dynamic graph, but it will not be able to match parts
of the API which require modification of the graph.</p>
<p>In this document, we’ll avoid the complexity of this C++ API hierarchy,
and focus on the most generic type of graph, the
<code class="docutils literal notranslate"><span class="pre">MutablePathDeletableHandleGraph</span></code> model implemented in
<code class="docutils literal notranslate"><span class="pre">`odgi</span></code> &lt;<a class="reference external" href="https://github.com/pangenome/odgi/blob/master/src/odgi.hpp">https://github.com/pangenome/odgi/blob/master/src/odgi.hpp</a>&gt;`__,
the
<code class="docutils literal notranslate"><span class="pre">`PackedGraph</span></code> &lt;<a class="reference external" href="https://github.com/vgteam/libbdsg/blob/master/include/bdsg/packed_graph.hpp">https://github.com/vgteam/libbdsg/blob/master/include/bdsg/packed_graph.hpp</a>&gt;`__
and
<code class="docutils literal notranslate"><span class="pre">`HashGraph</span></code> &lt;<a class="reference external" href="https://github.com/vgteam/libbdsg/blob/master/include/bdsg/hash_graph.hpp">https://github.com/vgteam/libbdsg/blob/master/include/bdsg/hash_graph.hpp</a>&gt;`__
from libbdsg. This model allows both query and modification (deletion,
addition, division, unification) of all graph elements. As such, it is
suitable for generic operations on genome graphs, both their
construction and interrogation.</p>
</div>
<div class="section" id="odgis-python-api">
<h2>odgi’s python API<a class="headerlink" href="#odgis-python-api" title="Permalink to this headline">¶</a></h2>
<p>Most of the elements in the HandleGraph API are <a class="reference external" href="https://github.com/pangenome/odgi/blob/master/src/pythonmodule.cpp">wrapped in a python
module</a>.
Python makes for good pseudocode, and so we can use it here to provide
some examples of how to work with the HandleGraph abstraction.</p>
<div class="section" id="loading-a-graph">
<h3>Loading a graph<a class="headerlink" href="#loading-a-graph" title="Permalink to this headline">¶</a></h3>
<p>Given a <a class="reference external" href="https://github.com/pangenome/odgi/blob/master/docs/assets/lil.gfa">graph in GFA
format</a>,
we can build the odgi serialization of it using <code class="docutils literal notranslate"><span class="pre">odgi</span> <span class="pre">build</span></code>.</p>
<div class="figure align-default" id="id1">
<img alt="example graph" src="https://github.com/pangenome/odgi/blob/master/docs/assets/lil.png" />
<p class="caption"><span class="caption-text">example graph</span><a class="headerlink" href="#id1" title="Permalink to this image">¶</a></p>
</div>
<div class="highlight-txt notranslate"><div class="highlight"><pre><span></span>H   VN:Z:1.0
S   1   CAAATAAG
S   2   A
S   3   G
S   4   T
S   5   C
S   6   TTG
S   7   A
S   8   G
S   9   AAATTTTCTGGAGTTCTAT
S   10  A
S   11  T
S   12  ATAT
S   13  A
S   14  T
S   15  CCAACTCTCTG
P   x   1+,3+,5+,6+,8+,9+,11+,12+,14+,15+   8M,1M,1M,3M,1M,19M,1M,4M,1M,11M
P   y   1+,2+,4+,6+,7+,9+,11+,12+,14+,15+   8M,1M,1M,3M,1M,19M,1M,4M,1M,11M
P   z   1+,3+,5+,6+,7+,9+,10+,12+,13+,15+   8M,1M,1M,3M,1M,19M,1M,4M,1M,11M
L   1   +   2   +   0M
L   1   +   3   +   0M
L   2   +   4   +   0M
L   2   +   5   +   0M
L   3   +   4   +   0M
L   3   +   5   +   0M
L   4   +   6   +   0M
L   5   +   6   +   0M
L   6   +   7   +   0M
L   6   +   8   +   0M
L   7   +   9   +   0M
L   8   +   9   +   0M
L   9   +   10  +   0M
L   9   +   11  +   0M
L   10  +   12  +   0M
L   11  +   12  +   0M
L   12  +   13  +   0M
L   12  +   14  +   0M
L   13  +   15  +   0M
L   14  +   15  +   0M
</pre></div>
</div>
<p>Transforming the graph into odgi’s succinct self index:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>odgi build -g lil.gfa -o lil.odgi
</pre></div>
</div>
<p>We can now load this into python, using the odgi python module:</p>
<p>First, we need to make sure our <code class="docutils literal notranslate"><span class="pre">PYTHONPATH</span></code> environment variable
points to the directory where our python module file lives. Assuming we
built <code class="docutils literal notranslate"><span class="pre">odgi</span></code> in our home directory, we could do this:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nv">PYTHONPATH</span><span class="o">=</span>~/odgi/lib python3
</pre></div>
</div>
<p>Now we can load the graph and check how big it is:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">odgi</span>

<span class="n">g</span> <span class="o">=</span> <span class="n">odgi</span><span class="o">.</span><span class="n">graph</span><span class="p">()</span> <span class="c1"># instatiate a graph</span>
<span class="n">g</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s1">&#39;lil.odgi&#39;</span><span class="p">)</span>
<span class="n">g</span><span class="o">.</span><span class="n">get_node_count</span><span class="p">()</span> <span class="c1"># 15</span>
</pre></div>
</div>
</div>
<div class="section" id="interrogating-parts-of-the-graph">
<h3>Interrogating parts of the graph<a class="headerlink" href="#interrogating-parts-of-the-graph" title="Permalink to this headline">¶</a></h3>
<p>We can examine an individual node:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">h</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">get_handle</span><span class="p">(</span><span class="mi">9</span><span class="p">)</span>
<span class="n">g</span><span class="o">.</span><span class="n">get_id</span><span class="p">(</span><span class="n">h</span><span class="p">)</span> <span class="c1"># returns 9</span>
<span class="n">g</span><span class="o">.</span><span class="n">get_is_reverse</span><span class="p">(</span><span class="n">h</span><span class="p">)</span>  <span class="c1"># False, by default, we get the forward handle</span>
<span class="n">r</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">get_handle</span><span class="p">(</span><span class="mi">9</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>  <span class="c1"># get the reverse handle</span>
<span class="n">g</span><span class="o">.</span><span class="n">get_is_reverse</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>  <span class="c1"># True, this handle is reverse</span>
<span class="nb">print</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">get_sequence</span><span class="p">(</span><span class="n">h</span><span class="p">))</span>  <span class="c1"># AAATTTTCTGGAGTTCTAT --- same as the node in the graph</span>
<span class="nb">print</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">get_sequence</span><span class="p">(</span><span class="n">r</span><span class="p">))</span>  <span class="c1"># ATAGAACTCCAGAAAATTT --- the reverse complement</span>
</pre></div>
</div>
<p>And we can check which paths overlap nodes:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">h</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">get_handle</span><span class="p">(</span><span class="mi">11</span><span class="p">)</span>
<span class="n">g</span><span class="o">.</span><span class="n">for_each_step_on_handle</span><span class="p">(</span><span class="n">h</span><span class="p">,</span>
    <span class="k">lambda</span> <span class="n">s</span><span class="p">:</span> <span class="nb">print</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">get_path_name</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">get_path_handle_of_step</span><span class="p">(</span><span class="n">s</span><span class="p">))))</span>
<span class="c1"># x</span>
<span class="c1"># y</span>
<span class="n">g</span><span class="o">.</span><span class="n">for_each_step_on_handle</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">get_handle</span><span class="p">(</span><span class="mi">9</span><span class="p">),</span>
    <span class="k">lambda</span> <span class="n">s</span><span class="p">:</span> <span class="nb">print</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">get_path_name</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">get_path_handle_of_step</span><span class="p">(</span><span class="n">s</span><span class="p">))))</span>
<span class="c1"># x</span>
<span class="c1"># y</span>
<span class="c1"># z</span>
</pre></div>
</div>
</div>
<div class="section" id="iterating-over-graph-elements">
<h3>Iterating over graph elements<a class="headerlink" href="#iterating-over-graph-elements" title="Permalink to this headline">¶</a></h3>
<p>Matching the C++ API, most of the methods of iterating over elements in
the graph use callback functions (note our use of lambda’s above). In
the current version of the python API, this causes some difficulty, as
python lacks true functional
<a class="reference external" href="https://en.wikipedia.org/wiki/Closure_(computer_programming)">closures</a>
and prevents assignments within callbacks. Future versions of this API
we provide generator functions thot support memory-efficient and
pythonic iteration over graph elements.</p>
<p>For instance, to iterate over our nodes, we can call
<code class="docutils literal notranslate"><span class="pre">for_each_handle</span></code>, which will invoke a callback for each forward
handle in our graph.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">g</span><span class="o">.</span><span class="n">for_each_handle</span><span class="p">(</span><span class="k">lambda</span> <span class="n">h</span><span class="p">:</span> <span class="nb">print</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">get_id</span><span class="p">(</span><span class="n">h</span><span class="p">),</span> <span class="n">g</span><span class="o">.</span><span class="n">get_sequence</span><span class="p">(</span><span class="n">h</span><span class="p">)))</span>
<span class="c1"># writes out each node id and its sequence</span>
</pre></div>
</div>
<p>We can enumerate the paths and get their names:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">g</span><span class="o">.</span><span class="n">for_each_path_handle</span><span class="p">(</span><span class="k">lambda</span> <span class="n">p</span><span class="p">:</span> <span class="nb">print</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">get_path_name</span><span class="p">(</span><span class="n">p</span><span class="p">)))</span>
<span class="c1"># x</span>
<span class="c1"># y</span>
<span class="c1"># z</span>
</pre></div>
</div>
<p>And we can iterate over the steps in a given path, finding which node
and orientation each step has:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># a function to call for each step in the path</span>
<span class="k">def</span> <span class="nf">process_step</span><span class="p">(</span><span class="n">s</span><span class="p">):</span>
    <span class="n">h</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">get_handle_of_step</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="c1"># gets the handle (both node and orientation) of the step</span>
    <span class="n">is_rev</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">get_is_reverse</span><span class="p">(</span><span class="n">h</span><span class="p">)</span>
    <span class="nb">id</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">get_id</span><span class="p">(</span><span class="n">h</span><span class="p">)</span>
    <span class="k">return</span> <span class="nb">str</span><span class="p">(</span><span class="nb">id</span><span class="p">)</span> <span class="o">+</span> <span class="p">(</span><span class="s2">&quot;+&quot;</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">is_rev</span> <span class="k">else</span> <span class="s2">&quot;-&quot;</span><span class="p">)</span>

<span class="n">p</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">get_path_handle</span><span class="p">(</span><span class="s1">&#39;z&#39;</span><span class="p">)</span>
<span class="n">q</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">g</span><span class="o">.</span><span class="n">for_each_step_in_path</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">s</span><span class="p">:</span> <span class="n">q</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">process_step</span><span class="p">(</span><span class="n">s</span><span class="p">)))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">get_path_name</span><span class="p">(</span><span class="n">p</span><span class="p">),</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">q</span><span class="p">))</span>
<span class="c1"># z 1+,3+,5+,6+,7+,9+,10+,12+,13+,15+</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="modifying-the-graph">
<h2>Modifying the graph<a class="headerlink" href="#modifying-the-graph" title="Permalink to this headline">¶</a></h2>
<p>It’s possible to add and delete nodes from the graph using the python
API:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">g</span><span class="o">.</span><span class="n">get_node_count</span><span class="p">()</span> <span class="c1"># 15</span>
<span class="n">h</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">create_handle</span><span class="p">(</span><span class="s2">&quot;GATTACA&quot;</span><span class="p">)</span>
<span class="n">g</span><span class="o">.</span><span class="n">get_node_count</span><span class="p">()</span> <span class="c1"># 16</span>
</pre></div>
</div>
<p>We can also add edges:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">g</span><span class="o">.</span><span class="n">create_edge</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">get_handle</span><span class="p">(</span><span class="mi">15</span><span class="p">),</span> <span class="n">h</span><span class="p">)</span> <span class="c1"># connects node 15 to 16</span>
</pre></div>
</div>
<p>And add path steps:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">h</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">get_handle</span><span class="p">(</span><span class="mi">16</span><span class="p">)</span>
<span class="n">p</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">get_path_handle</span><span class="p">(</span><span class="s1">&#39;z&#39;</span><span class="p">)</span>
<span class="n">g</span><span class="o">.</span><span class="n">append_step</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="n">h</span><span class="p">)</span>
</pre></div>
</div>
<p>Path <code class="docutils literal notranslate"><span class="pre">z</span></code> now ends at 16+, and running the path enumeration code above
would yield:</p>
<div class="highlight-txt notranslate"><div class="highlight"><pre><span></span>z 1+,3+,5+,6+,7+,9+,10+,12+,13+,15+,16+
</pre></div>
</div>
<p>We can divide a node without breaking the paths that overlap it:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">g</span><span class="o">.</span><span class="n">divide_handle</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">get_handle</span><span class="p">(</span><span class="mi">9</span><span class="p">),</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span><span class="mi">7</span><span class="p">])</span>
</pre></div>
</div>
<p>Assuming you’ve executed all the code up to this point, the paths now
walk through new nodes 17, 18, and 19 in place of node 9.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">g</span><span class="o">.</span><span class="n">to_gfa</span><span class="p">()</span>
<span class="c1"># ...</span>
<span class="c1"># P x 1+,3+,5+,6+,8+,17+,18+,19+,11+,12+,14+,15+      *,*,*,*,*,*,*,*,*,*,*</span>
<span class="c1"># P y 1+,2+,4+,6+,7+,17+,18+,19+,11+,12+,14+,15+      *,*,*,*,*,*,*,*,*,*,*</span>
<span class="c1"># P z 1+,3+,5+,6+,7+,17+,18+,19+,10+,12+,13+,15+,16+  *,*,*,*,*,*,*,*,*,*,*,*</span>
</pre></div>
</div>
</div>
<div class="section" id="summary">
<h2>Summary<a class="headerlink" href="#summary" title="Permalink to this headline">¶</a></h2>
<p>In this document, we’ve covered the basic concepts in the HandleGraph
abstraction and used the odgi python library to explore some of them
interactively. Although this interface is a work in progress, it should
already provide enough material for researchers in genomics who want to
work with genome graph data structures. HandleGraph implementations like
odgi demonstrate that we can work with genome graphs even when they are
large, with dense variation. These models lift limitations on graph
structure that have been a persistent feature of other genome graph
implementations. We believe that these limitations were motivated mostly
by difficulty in managing the memory requirements of genome graphs.
Given that this issue is difficult to resolve, we hope to provide a
generic solution to it. This should provide a generic foundation for the
use of genome graphs in bioinformatics, genomics, and population
genetics.</p>
</div>
</div>


           </div>
           
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Erik Garrison. Revision v0.5.1-00d68c2.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>