

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Exploratory analysis &mdash; odgi 00d68c2 documentation</title>
  

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Detect complex regions" href="detect_complex_regions.html" />
    <link rel="prev" title="Tutorials" href="../tutorials.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home"> odgi
          

          
          </a>

          
            
            
              <div class="version">
                v0.5.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../index.html">Welcome</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../quick_start.html">Quick Start</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../tutorials.html">Tutorials</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">Exploratory analysis</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#synopsis">Synopsis</a></li>
<li class="toctree-l3"><a class="reference internal" href="#steps">Steps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#build-the-drb1-3123-graph">Build the DRB1-3123 graph</a></li>
<li class="toctree-l4"><a class="reference internal" href="#visualize-the-drb1-3123-graph">Visualize the DRB1-3123 graph</a></li>
<li class="toctree-l4"><a class="reference internal" href="#color-with-respect-to-the-node-position">Color with respect to the node position</a></li>
<li class="toctree-l4"><a class="reference internal" href="#color-with-respect-to-the-node-strandedness">Color with respect to the node strandedness</a></li>
<li class="toctree-l4"><a class="reference internal" href="#build-the-lipoprotein-a-graph">Build the Lipoprotein A graph</a></li>
<li class="toctree-l4"><a class="reference internal" href="#visualize-the-lpa-graph">Visualize the LPA graph</a></li>
<li class="toctree-l4"><a class="reference internal" href="#color-with-respect-to-the-node-depth-in-a-path">Color with respect to the node depth in a path</a></li>
<li class="toctree-l4"><a class="reference internal" href="#visualize-a-particular-region">Visualize a particular region</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="detect_complex_regions.html">Detect complex regions</a></li>
<li class="toctree-l2"><a class="reference internal" href="extract_selected_loci.html">Extract selected <cite>loci</cite></a></li>
<li class="toctree-l2"><a class="reference internal" href="sorting_layouting.html">Sorting and Layouting</a></li>
<li class="toctree-l2"><a class="reference internal" href="translate_path_positions_between_graphs.html">Translate path positions between graphs</a></li>
<li class="toctree-l2"><a class="reference internal" href="remove_artifacts_and_complex_regions.html">Remove artifacts and complex regions</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../commands.html">Commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="../binding.html">Python Binding</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">odgi</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../tutorials.html">Tutorials</a> &raquo;</li>
        
      <li>Exploratory analysis</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../../_sources/rst/tutorials/exploratory_analysis.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="exploratory-analysis">
<h1>Exploratory analysis<a class="headerlink" href="#exploratory-analysis" title="Permalink to this headline">¶</a></h1>
<div class="section" id="synopsis">
<h2>Synopsis<a class="headerlink" href="#synopsis" title="Permalink to this headline">¶</a></h2>
<p>Visualizing pangenome graphs we can gain insight into the mutual relationship between the embedded genomes and their
variation. However, complex, nonlinear graph structures are difficult to present in a convenient number of dimensions.
<a class="reference internal" href="../commands/odgi_viz.html#odgi-viz"><span class="std std-ref">odgi viz</span></a> offers a binned, linearized 1-Dimensional rendering. It is able to display gigabase scale pangenomes,
and several visualization modalities to grasp the complexity in the graphs.</p>
</div>
<div class="section" id="steps">
<h2>Steps<a class="headerlink" href="#steps" title="Permalink to this headline">¶</a></h2>
<div class="section" id="build-the-drb1-3123-graph">
<h3>Build the DRB1-3123 graph<a class="headerlink" href="#build-the-drb1-3123-graph" title="Permalink to this headline">¶</a></h3>
<p>Assuming that your current working directory is the root of the <code class="docutils literal notranslate"><span class="pre">odgi</span></code> project, to construct an <code class="docutils literal notranslate"><span class="pre">odgi</span></code> file from the
<code class="docutils literal notranslate"><span class="pre">DRB1-3123</span></code> dataset in <code class="docutils literal notranslate"><span class="pre">GFA</span></code> format, execute:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>odgi build -g test/DRB1-3123.gfa -o DRB1-3123.og
</pre></div>
</div>
<p>The command creates a file called <code class="docutils literal notranslate"><span class="pre">DRB1-3123.og</span></code>, which contains the input graph in <code class="docutils literal notranslate"><span class="pre">odgi</span></code> format. This graph contains
12 ALT sequences of the <a class="reference external" href="https://www.ncbi.nlm.nih.gov/gene/3123">HLA-DRB1 gene</a> from the GRCh38 reference genome.</p>
</div>
<div class="section" id="visualize-the-drb1-3123-graph">
<h3>Visualize the DRB1-3123 graph<a class="headerlink" href="#visualize-the-drb1-3123-graph" title="Permalink to this headline">¶</a></h3>
<p>To visualize the graph, execute:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>odgi viz -i DRB1-3123.og -o DRB1-3123.png -x <span class="m">500</span>
</pre></div>
</div>
<p>To obtain the following PNG image:</p>
<img alt="../../_images/DRB1-3123.png" src="../../_images/DRB1-3123.png" />
<p>In this 1-Dimensional visualization:</p>
<ul class="simple">
<li><p>The graph nodes are arranged from left to right, forming the <code class="docutils literal notranslate"><span class="pre">pangenome</span> <span class="pre">sequence</span></code>.</p></li>
<li><p>The colored bars represent the the paths versus the <code class="docutils literal notranslate"><span class="pre">pangenome</span> <span class="pre">sequences</span></code> in a binary matrix.</p></li>
<li><p>The path names are visualized on the left.</p></li>
<li><p>The black lines under the paths are the links, which represent the graph topology.</p></li>
</ul>
<p>See <a class="reference internal" href="../commands/odgi_viz.html#odgi-viz"><span class="std std-ref">odgi viz</span></a> documentation for more information.</p>
</div>
<div class="section" id="color-with-respect-to-the-node-position">
<h3>Color with respect to the node position<a class="headerlink" href="#color-with-respect-to-the-node-position" title="Permalink to this headline">¶</a></h3>
<p>This is a linearized visualization, but the pangenome graphs are not linear when the embedded genomes present structural
variation. However, a graph can be optimized for being better visualized in 1-Dimension by sorting its nodes properly
(see the <a class="reference internal" href="sorting_layouting.html#sorting-layouting"><span class="std std-ref">Sorting and Layouting</span></a> tutorial for more information).</p>
<p>To color the bars with respect to the node position in each path, execute:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>odgi viz -i DRB1-3123.og -o DRB1-3123.du.png -x <span class="m">500</span> -d -u
</pre></div>
</div>
<p>To obtain the following PNG image:</p>
<img alt="../../_images/DRB1-3123.du.png" src="../../_images/DRB1-3123.du.png" />
<p>For each path, the brightness goes from light (for the starting position) to black (for the ending position). A linear
genome in a well-sorted graph appears with a smooth brightness gradient in this visualization modality. <code class="docutils literal notranslate"><span class="pre">-d</span></code> changes
the color darkness based on the nucleotide position in the path. <code class="docutils literal notranslate"><span class="pre">-u</span></code> sets the color darkness range from white for
the first nucleotide position of a path to black for the last nucleotide position of a path.</p>
<p>Interestingly, the <code class="docutils literal notranslate"><span class="pre">&gt;gi|345525392:5000-18402</span></code> path has a brightness gradient which go from right to left. DNA sequence
graphs have two strands, with the node implicitly representing both strands. That gradient indicates that the path is
reversed with respect to the <code class="docutils literal notranslate"><span class="pre">pangenome</span> <span class="pre">sequence</span></code>.</p>
</div>
<div class="section" id="color-with-respect-to-the-node-strandedness">
<h3>Color with respect to the node strandedness<a class="headerlink" href="#color-with-respect-to-the-node-strandedness" title="Permalink to this headline">¶</a></h3>
<p>To color the bars with respect to the strandedness that each node has in each path, execute:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>odgi viz -i DRB1-3123.og -o DRB1-3123.z.png -x <span class="m">500</span> -z
</pre></div>
</div>
<p>to obtain the following PNG image:</p>
<img alt="../../_images/DRB1-3123.z.png" src="../../_images/DRB1-3123.z.png" />
<p><code class="docutils literal notranslate"><span class="pre">-z</span></code> changes the color palette to respect the node strandedness. Black is forward, red is reverse.
The red bar in a path indicates that that region is inverted in that path with respect to the <code class="docutils literal notranslate"><span class="pre">pangenome</span> <span class="pre">sequence</span></code>.</p>
</div>
<div class="section" id="build-the-lipoprotein-a-graph">
<h3>Build the Lipoprotein A graph<a class="headerlink" href="#build-the-lipoprotein-a-graph" title="Permalink to this headline">¶</a></h3>
<p>Assuming that your current working directory is the root of the <code class="docutils literal notranslate"><span class="pre">odgi</span></code> project, to construct an <code class="docutils literal notranslate"><span class="pre">odgi</span></code> file from the
<code class="docutils literal notranslate"><span class="pre">LPA</span></code> dataset in <code class="docutils literal notranslate"><span class="pre">GFA</span></code> format, execute:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>odgi build -g test/LPA.gfa -o LPA.og
</pre></div>
</div>
<p>The command creates a file called <code class="docutils literal notranslate"><span class="pre">LPA.og</span></code>, which contains the input graph in <code class="docutils literal notranslate"><span class="pre">odgi</span></code> format. This graph contains
13 contigs from 7 haploid human genome assemblies from 6 individuals plus the chm13 cell line. The contigs cover the
<a class="reference external" href="https://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000198670">Lipoprotein A (LPA)</a> locus, which encodes the
Apo(a) protein.</p>
</div>
<div class="section" id="visualize-the-lpa-graph">
<h3>Visualize the LPA graph<a class="headerlink" href="#visualize-the-lpa-graph" title="Permalink to this headline">¶</a></h3>
<p>To visualize the graph, execute:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>odgi viz -i LPA.og -o LPA.b.png -x <span class="m">500</span> -b
</pre></div>
</div>
<p>To obtain the following PNG image:</p>
<img alt="../../_images/LPA.b.png" src="../../_images/LPA.b.png" />
</div>
<div class="section" id="color-with-respect-to-the-node-depth-in-a-path">
<h3>Color with respect to the node depth in a path<a class="headerlink" href="#color-with-respect-to-the-node-depth-in-a-path" title="Permalink to this headline">¶</a></h3>
<p>Eukaryotic genomes are characterized by repetitive sequences. These sequences can lead to complex regions in the pangenome
graphs. To identify them, we can analyze the <strong>depth</strong> in the graph. Here we define <strong>node depth in a path</strong> as the number
of times the node is crossed by a path.</p>
<p>To color the bars with respect to the mean <cite>depth</cite>, execute:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>odgi viz -i LPA.og -o LPA.bm.png -x <span class="m">500</span> -bm
</pre></div>
</div>
<p>To obtain the following PNG image:</p>
<img alt="../../_images/LPA.bm.png" src="../../_images/LPA.bm.png" />
<p>Low depth regions are black, while high depth regions are colored green. Apo(a) proteins vary in size due to a size
polymorphism, the KIV-2 variable numbers of tandem repeats (VNTRs). The VNTR region in the LPA pangenome presents high
<strong>depth</strong>, that becomes evident as a light green stripe in the image. <code class="docutils literal notranslate"><span class="pre">-b</span></code> explicitly forces <code class="docutils literal notranslate"><span class="pre">odgi</span> <span class="pre">viz</span></code> to bin the
graph before visualizing it. <code class="docutils literal notranslate"><span class="pre">-m</span></code> changes the color palette to display the mean depth per bin as a shade of green.</p>
</div>
<div class="section" id="visualize-a-particular-region">
<h3>Visualize a particular region<a class="headerlink" href="#visualize-a-particular-region" title="Permalink to this headline">¶</a></h3>
<p>To obtain the coordinates of the VNTRs, execute:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>odgi depth -i LPA.og -r chm13__LPA__tig00000001<span class="p">|</span> <span class="se">\</span>
    bedtools makewindows -b /dev/stdin -w <span class="m">5000</span> &gt; chm13__LPA__tig00000001.w5kbps.bed

odgi depth -i LPA.og -b chm13__LPA__tig00000001.w5kbps.bed <span class="p">|</span> <span class="se">\</span>
    bedtools sort &gt; chm13__LPA__tig00000001.depth.w5kbps.bed

awk -F<span class="s2">&quot;\t&quot;</span> <span class="s1">&#39;$4 &gt; 20.0&#39;</span> chm13__LPA__tig00000001.depth.w5kbps.bed <span class="p">|</span> <span class="se">\</span>
    bedtools merge
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>chm13__LPA__tig00000001 140000  275000
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">chm13__LPA__tig00000001.w5kbps.bed</span></code> file contains 5000 bp interval windows across the chm13__LPA__tig00000001
contig. The depth is computed for each of these windows, writing the result in the
<code class="docutils literal notranslate"><span class="pre">chm13__LPA__tig00000001.depth.w5kbps.bed</span></code> file, in <code class="docutils literal notranslate"><span class="pre">BED</span></code> format. <code class="docutils literal notranslate"><span class="pre">-r</span></code> specifies the path name from which to
compute the depth from. <code class="docutils literal notranslate"><span class="pre">-b</span></code> specifies the BED ranges of which the depths should be calculated of.</p>
<p>To visualize the identified region, execute:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>odgi viz -i LPA.og -o LPA.bm.VNTRs.png -x <span class="m">500</span> -bm -r chm13__LPA__tig00000001:140000-275000
</pre></div>
</div>
<p>To obtain the following PNG image:</p>
<img alt="../../_images/LPA.bm.VNTRs.png" src="../../_images/LPA.bm.VNTRs.png" />
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="detect_complex_regions.html" class="btn btn-neutral float-right" title="Detect complex regions" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="../tutorials.html" class="btn btn-neutral float-left" title="Tutorials" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Erik Garrison. Revision v0.5.1-00d68c2.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>